// @airbnb/lunar/esm/components/Menu/index.js
import _pt from "prop-types";
import React from 'react';
import useStyles from '../../hooks/useStyles';
import Separator from './Separator';
import Item from './Item';
import Row from './Row';
import { styleSheet } from './styles';

/** An abstract menu for use within dropdowns, selects, autocompletes, and more. */
export default function Menu(_ref) {
  let {
    accessibilityLabel,
    children,
    maxHeight,
    minWidth = 200,
    overflow,
    role = 'menu'
  } = _ref;
  const scrollable = !!maxHeight && !overflow;
  const [styles, cx] = useStyles(styleSheet);
  return React.createElement("ul", {
    role: role,
    "aria-label": accessibilityLabel,
    className: cx(styles.menu, {
      maxHeight: scrollable ? maxHeight : 'auto',
      minWidth
    }, scrollable && styles.menu_scrollable)
  }, children);
}
Menu.propTypes = {
  children: _pt.node,
  accessibilityLabel: _pt.string.isRequired,
  maxHeight: _pt.number,
  minWidth: _pt.number,
  role: _pt.string,
  overflow: _pt.bool
};
export { Item, Separator, Row };
