// @airbnb/lunar/esm/components/DataTable/columns/renderExpandableColumn.js
import React from 'react';
import { Column } from 'react-virtualized';
import Spacing from '../../Spacing';
import ExpandableIcon from '../../ExpandableIcon';
import { EXPANDABLE_COLUMN_WIDTH } from '../constants';
export default function renderExpandableColumn(cx, styles, expandedRows, expandRow) {
  const cellRenderer = row => {
    const {
      children,
      originalIndex
    } = row.rowData.metadata;

    if (children && children.length > 0) {
      return React.createElement("div", {
        className: cx(styles.expand_caret),
        role: "button",
        tabIndex: 0,
        onClick: expandRow(originalIndex),
        onKeyPress: expandRow(originalIndex)
      }, React.createElement(Spacing, {
        left: 1.5
      }, React.createElement(ExpandableIcon, {
        expanded: expandedRows.has(originalIndex),
        size: "1.6em"
      })));
    }

    return React.createElement("div", {
      className: cx(styles.row)
    });
  };

  return React.createElement(Column, {
    dataKey: "expanded",
    cellRenderer: cellRenderer,
    width: EXPANDABLE_COLUMN_WIDTH
  });
}
