// @airbnb/lunar/lib/components/Rating/index.js
"use strict";

exports.__esModule = true;
exports.default = Rating;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _IconStar = _interopRequireDefault(require("@airbnb/lunar-icons/lib/interface/IconStar"));

var _IconStarFull = _interopRequireDefault(require("@airbnb/lunar-icons/lib/interface/IconStarFull"));

var _IconStarHalf = _interopRequireDefault(require("@airbnb/lunar-icons/lib/interface/IconStarHalf"));

var _useTheme = _interopRequireDefault(require("../../hooks/useTheme"));

var _Row = _interopRequireDefault(require("../Row"));

var _Text = _interopRequireDefault(require("../Text"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const NUMBER_OF_STARS = 5;
const STARS = new Array(NUMBER_OF_STARS).fill(null);

/** Display a star rating and review count. */
function Rating(_ref) {
  let {
    large,
    micro,
    rating = 0,
    reviews,
    small
  } = _ref;
  const theme = (0, _useTheme.default)();
  return _react.default.createElement(_Text.default, {
    large: large,
    micro: micro,
    small: small
  }, _react.default.createElement(_Row.default, {
    inline: true,
    middleAlign: true,
    after: reviews
  }, STARS.map((star, index) => {
    const key = "star-" + index;
    const color = rating > index || rating > index && rating < index + 1 ? theme.color.core.primary[3] : theme.color.core.neutral[3];
    let Star = _IconStar.default;

    if (rating > index && rating < index + 1) {
      Star = _IconStarHalf.default;
    } else if (rating > index) {
      Star = _IconStarFull.default;
    }

    return _react.default.createElement(Star, {
      key: key,
      decorative: true,
      color: color,
      size: "1.15em"
    });
  })));
}

Rating.propTypes = {
  large: _propTypes.default.bool,
  micro: _propTypes.default.bool,
  rating: _propTypes.default.number,
  reviews: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  small: _propTypes.default.bool
};
