// @airbnb/lunar/esm/components/Accordion/index.js
import _pt from "prop-types";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import uuid from 'uuid/v4';
import withStyles from '../../composers/withStyles';
import Item from './Item';
import { styleSheet } from './styles';
export { Item };

/** A controller for multiple accordion items. */
export class Accordion extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      id: uuid(),
      index: this.props.defaultIndex || 0
    });

    _defineProperty(this, "handleClick", index => {
      this.setState(prevState => ({
        index: index === prevState.index ? -1 : index
      }));
    });
  }

  componentDidUpdate(prevProps) {
    if (this.props.defaultIndex !== prevProps.defaultIndex) {
      this.setState({
        index: this.props.defaultIndex
      });
    }
  }

  render() {
    const {
      cx,
      bordered,
      children,
      styles
    } = this.props;
    const {
      id,
      index
    } = this.state;
    return React.createElement("div", {
      className: cx(styles.container, bordered && styles.container_bordered),
      role: "tablist"
    }, React.Children.map(children, (child, i) => {
      if (!child) {
        return null;
      }

      return React.cloneElement(child, {
        bordered,
        expanded: i === index,
        id: id + "-" + i,
        index: i,
        onClick: this.handleClick
      });
    }));
  }

}

_defineProperty(Accordion, "propTypes", {
  bordered: _pt.bool,
  children: _pt.any.isRequired,
  defaultIndex: _pt.number
});

_defineProperty(Accordion, "defaultProps", {
  bordered: false,
  defaultIndex: 0
});

export default withStyles(styleSheet)(Accordion);
