// @airbnb/lunar/lib/components/Chip/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _iconComponent = _interopRequireDefault(require("../../prop-types/iconComponent"));

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _ProfilePhoto = _interopRequireDefault(require("../ProfilePhoto"));

var _ButtonOrLink = _interopRequireDefault(require("../private/ButtonOrLink"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/** Compact component that represents a snippet of information, such as a filter. */
function Chip(_ref) {
  let {
    active,
    afterIcon,
    beforeIcon,
    children,
    compact,
    disabled,
    id,
    onClick,
    onIconClick,
    profileImageSrc,
    trackingName
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheet);
  const Component = onClick ? 'button' : 'div';
  const props = Component === 'button' ? _extends({
    disabled,
    onClick,
    type: 'button'
  }, trackingName && {
    'data-tracking-name': trackingName
  }) : {};
  const shouldRenderBefore = beforeIcon || profileImageSrc;
  return (// @ts-ignore [ts] JSX element type 'Component' does not have any construct or call signatures. [2604]
    _react.default.createElement(Component, _extends({
      className: cx(styles.chip, onClick && styles.chip_button, !shouldRenderBefore && styles.chip_noBefore, !afterIcon && styles.chip_noAfter, active && styles.chip_active, onClick && active && styles.chip_active_button, compact && styles.chip_compact, disabled && styles.chip_disabled)
    }, props, {
      id: id
    }), shouldRenderBefore && _react.default.createElement("div", {
      className: cx(styles.chipItem, styles.sideContent)
    }, _react.default.createElement("div", {
      className: cx(styles.sideContentInner, !!beforeIcon && styles.iconWrapper, !!beforeIcon && styles.iconWrapperBefore)
    }, profileImageSrc && _react.default.createElement(_ProfilePhoto.default, {
      imageSrc: profileImageSrc,
      title: "",
      size: 4
    }), beforeIcon)), _react.default.createElement("div", {
      className: cx(styles.chipItem, styles.content)
    }, children), afterIcon && _react.default.createElement("div", {
      className: cx(styles.chipItem, styles.sideContent)
    }, _react.default.createElement("div", {
      className: cx(styles.sideContentInner, styles.iconWrapper, styles.iconWrapperAfter)
    }, onIconClick ? _react.default.createElement(_ButtonOrLink.default, {
      className: cx(styles.iconButton, disabled && styles.iconButton_disabled),
      disabled: disabled,
      onClick: onIconClick
    }, afterIcon) : afterIcon)))
  );
}

Chip.propTypes = {
  trackingName: _propTypes.default.string,
  onIconClick: _propTypes.default.func,
  id: _propTypes.default.string,
  disabled: _propTypes.default.bool,
  children: _propTypes.default.any.isRequired,
  active: _propTypes.default.bool,
  afterIcon: (0, _airbnbPropTypes.requiredBy)('onIconClick', _iconComponent.default),
  beforeIcon: (0, _airbnbPropTypes.mutuallyExclusiveProps)(_propTypes.default.node, 'beforeIcon', 'profileImageSrc'),
  compact: (0, _airbnbPropTypes.mutuallyExclusiveProps)(_propTypes.default.any, 'profileImageSrc', 'compact'),
  profileImageSrc: (0, _airbnbPropTypes.mutuallyExclusiveProps)(_propTypes.default.any, 'beforeIcon', 'profileImageSrc', 'compact'),
  onClick: (0, _airbnbPropTypes.mutuallyExclusiveProps)(_propTypes.default.func, 'onIconClick')
};
var _default = Chip;
exports.default = _default;
