// @airbnb/lunar/lib/composers/withBoundary.js
"use strict";

exports.__esModule = true;
exports.default = withBoundary;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _ErrorBoundary = _interopRequireDefault(require("../components/ErrorBoundary"));

var _TrackingBoundary = _interopRequireDefault(require("../components/TrackingBoundary"));

var _finishHOC = _interopRequireDefault(require("../utils/finishHOC"));

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function withBoundary(name)
/* infer */
{
  return function withBoundaryFactory(WrappedComponent) {
    function WithBoundary(_ref) {
      let {
        trackingName
      } = _ref,
          restProps = _objectWithoutPropertiesLoose(_ref, ["trackingName"]);

      return _react.default.createElement(_TrackingBoundary.default, {
        name: trackingName || name
      }, _react.default.createElement(_ErrorBoundary.default, {
        name: trackingName || name,
        onCatch: _.default.log
      }, _react.default.createElement(WrappedComponent, restProps)));
    }

    WithBoundary.propTypes = {
      trackingName: _propTypes.default.string
    };
    return (0, _finishHOC.default)('withBoundary', WithBoundary, WrappedComponent);
  };
}
