// @airbnb/lunar/lib/components/Lightbox/Image.js
"use strict";

exports.__esModule = true;
exports.default = Image;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _ResponsiveImage = _interopRequireDefault(require("../ResponsiveImage"));

var _Loader = _interopRequireDefault(require("../Loader"));

var _ImageViewer = _interopRequireDefault(require("../ImageViewer"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Image(_ref) {
  let {
    alt,
    aside,
    hideAside,
    src,
    showZoomControls,
    showRotateControls,
    scale,
    rotation
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheetImage);
  return _react.default.createElement("div", {
    className: cx(styles.wrapper)
  }, _react.default.createElement("div", {
    className: cx(styles.layout)
  }, _react.default.createElement("figure", {
    className: cx(styles.figure)
  }, showZoomControls || showRotateControls ? _react.default.createElement(_ImageViewer.default, {
    alt: alt,
    src: src,
    scale: scale,
    rotation: rotation,
    width: "100%",
    height: "100%"
  }) : _react.default.createElement(_ResponsiveImage.default, {
    contain: true,
    noShadow: true,
    alt: alt,
    borderRadius: 0,
    src: src,
    maxWidth: "100%",
    maxHeight: "100%",
    shimmer: _react.default.createElement(_Loader.default, null)
  })), aside && !hideAside && _react.default.createElement("aside", {
    className: cx(styles.aside)
  }, aside)));
}

Image.propTypes = {
  alt: _propTypes.default.string.isRequired,
  aside: _propTypes.default.node,
  hideAside: _propTypes.default.bool,
  rotation: _propTypes.default.number,
  scale: _propTypes.default.number,
  showRotateControls: _propTypes.default.bool,
  showZoomControls: _propTypes.default.bool,
  src: _propTypes.default.string.isRequired
};
