// @airbnb/lunar/lib/components/Lightbox/styles.js
"use strict";

exports.__esModule = true;
exports.styleSheetImage = exports.styleSheetHeader = void 0;

const styleSheetHeader = (_ref) => {
  let {
    unit
  } = _ref;
  return {
    header: {
      alignItems: 'center',
      display: 'flex',
      justifyContent: 'center'
    },
    pagination: {
      flexGrow: 1,
      flexShrink: 1,
      paddingLeft: unit
    },
    infoButton: {
      flexGrow: 0,
      flexShrink: 0,
      paddingRight: unit,
      textAlign: 'right'
    }
  };
};

exports.styleSheetHeader = styleSheetHeader;

const styleSheetImage = (_ref2) => {
  let {
    color,
    unit
  } = _ref2;
  return {
    wrapper: {
      position: 'relative',
      height: '100%'
    },
    layout: {
      display: 'flex',
      position: 'absolute',
      top: -unit * 3,
      right: -unit,
      left: -unit,
      bottom: -unit
    },
    figure: {
      background: color.accent.bgHover,
      flexGrow: 1,
      margin: 0
    },
    aside: {
      width: 300,
      flexGrow: 0,
      flexShrink: 0,
      padding: unit * 3
    }
  };
};

exports.styleSheetImage = styleSheetImage;
