// @airbnb/lunar/lib/components/SteppedProgressBar/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _Step = _interopRequireDefault(require("./Step"));

exports.Step = _Step.default;

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** A progress bar separated into individual steps. */
function SteppedProgressBar(_ref) {
  let {
    children
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheet);

  const steps = _react.default.Children.count(children);

  return _react.default.createElement("div", {
    className: cx(styles.bar)
  }, _react.default.Children.map(children, (child, index) => _react.default.cloneElement(child, {
    first: index === 0,
    last: index === steps - 1
  })));
}

SteppedProgressBar.propTypes = {
  children: (0, _airbnbPropTypes.childrenOfType)(_Step.default).isRequired
};
var _default = SteppedProgressBar;
exports.default = _default;
