// @airbnb/lunar/lib/components/SteppedProgressBar/Step.js
"use strict";

exports.__esModule = true;
exports.default = Step;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _ProgressBar = _interopRequireDefault(require("../ProgressBar"));

var _Tooltip = _interopRequireDefault(require("../Tooltip"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** An individual step within a stepped progress bar. */
function Step(_ref) {
  let {
    complete,
    first,
    last,
    label
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheetStep);

  const bar = _react.default.createElement(_ProgressBar.default, {
    leading: first || !first && !last,
    trailing: last || !first && !last,
    percent: complete ? 100 : 0
  });

  return _react.default.createElement("div", {
    className: cx(styles.step)
  }, label ? _react.default.createElement(_Tooltip.default, {
    content: label
  }, bar) : bar);
}

Step.propTypes = {
  complete: _propTypes.default.bool,
  first: _propTypes.default.bool,
  label: _propTypes.default.node,
  last: _propTypes.default.bool
};
