// @airbnb/lunar/lib/components/ToggleButtonController/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _v = _interopRequireDefault(require("uuid/v4"));

var _proxyComponent = _interopRequireDefault(require("../../utils/proxyComponent"));

var _FormField = _interopRequireWildcard(require("../FormField"));

var _Button = _interopRequireDefault(require("../Button"));

var _FormInputButton = _interopRequireDefault(require("../private/FormInputButton"));

var _ButtonGroup = _interopRequireDefault(require("../ButtonGroup"));

exports.ButtonGroup = _ButtonGroup.default;

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Manage a group of buttons with the same input `name`. */
class ToggleButtonController extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      id: (0, _v.default)(),
      value: this.props.value || ''
    });

    _defineProperty(this, "handleClick", event => {
      const newValue = event.currentTarget.dataset.value;
      const {
        value
      } = this.state;

      if (newValue === value) {
        return;
      }

      this.props.onChange(newValue, event);
      this.setState({
        value: newValue
      });
    });

    _defineProperty(this, "renderButton", (0, _proxyComponent.default)(_Button.default, (_ref) => {
      let {
        children,
        value
      } = _ref,
          props = _objectWithoutPropertiesLoose(_ref, ["children", "value"]);

      const {
        compact,
        disabled,
        invalid
      } = this.props;
      const {
        id,
        value: selectedValue
      } = this.state;
      const selected = value === selectedValue;
      return _react.default.createElement(_FormInputButton.default, _extends({}, props, {
        "data-value": value,
        id: id + "-" + value,
        name: name,
        disabled: disabled,
        inverted: !selected,
        invalid: invalid,
        small: compact,
        onClick: this.handleClick
      }), children);
    }));
  }

  componentDidUpdate(prevProps) {
    if (this.props.value !== prevProps.value) {
      this.setState({
        value: this.props.value || ''
      });
    }
  }

  render() {
    const {
      children,
      fieldProps
    } = (0, _FormField.partitionFieldProps)(this.props);
    const {
      id,
      value
    } = this.state;
    return _react.default.createElement(_FormField.default, _extends({}, fieldProps, {
      id: id
    }), children(this.renderButton, value));
  }

}

exports.default = ToggleButtonController;

_defineProperty(ToggleButtonController, "propTypes", {
  children: _propTypes.default.func.isRequired,
  name: _propTypes.default.string.isRequired,
  onChange: _propTypes.default.func.isRequired,
  value: _propTypes.default.string
});

_defineProperty(ToggleButtonController, "defaultProps", {
  value: ''
});
