// @airbnb/lunar/lib/components/CountBadge/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.CountBadge = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** A badge indicator with a count. */
class CountBadge extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "badgeRef", _react.default.createRef());
  }

  componentDidUpdate(prevProps) {
    const {
      value
    } = this.props;

    if (value && prevProps.value && value !== prevProps.value) {
      this.runPopAnimation();
    }
  }

  runPopAnimation() {
    if (this.badgeRef.current) {
      this.badgeRef.current.animate( // @ts-ignore Bug: https://github.com/Microsoft/TypeScript/issues/26073
      [{
        transform: 'scale(1)'
      }, {
        transform: 'scale(1.1)',
        offset: 0.3
      }, {
        transform: 'scale(.95)',
        offset: 0.8
      }, {
        transform: 'scale(1)'
      }], 300);
    }
  }

  render() {
    const {
      cx,
      accessibilityLabel,
      styles,
      value,
      waggle
    } = this.props;

    if (!value) {
      return null;
    }

    return _react.default.createElement("div", {
      ref: this.badgeRef,
      className: cx(styles.badge, waggle ? styles.animateInAndWaggle : styles.animateIn),
      "aria-label": accessibilityLabel
    }, value.toLocaleString());
  }

}

exports.CountBadge = CountBadge;

_defineProperty(CountBadge, "propTypes", {
  accessibilityLabel: _propTypes.default.string.isRequired,
  value: _propTypes.default.number.isRequired,
  waggle: _propTypes.default.bool
});

_defineProperty(CountBadge, "defaultProps", {
  waggle: false
});

var _default = (0, _withStyles.default)(_styles.styleSheet)(CountBadge);

exports.default = _default;
