// @airbnb/lunar/lib/components/Shimmer/index.js
"use strict";

exports.__esModule = true;
exports.default = Shimmer;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** A loading indicator bar that shimmers. */
function Shimmer(_ref) {
  let {
    block,
    height = '1.25ex',
    radius = '0.125em',
    width = '60%'
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheet);
  const randomWidth = width === 'random' ? Math.round(Math.random() * (90 - 30) + 30) + "%" : '';
  return _react.default.createElement("span", {
    "aria-busy": "true",
    style: {
      borderRadius: radius,
      width: randomWidth || width,
      height
    },
    className: cx(styles.shimmer, block && styles.shimmer_block)
  });
}

Shimmer.propTypes = {
  block: _propTypes.default.bool,
  height: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  width: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  radius: _propTypes.default.string
};
