// @airbnb/lunar/lib/components/DataTable/columns/renderExpandableColumn.js
"use strict";

exports.__esModule = true;
exports.default = renderExpandableColumn;

var _react = _interopRequireDefault(require("react"));

var _reactVirtualized = require("react-virtualized");

var _Spacing = _interopRequireDefault(require("../../Spacing"));

var _ExpandableIcon = _interopRequireDefault(require("../../ExpandableIcon"));

var _constants = require("../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function renderExpandableColumn(cx, styles, expandedRows, expandRow) {
  const cellRenderer = row => {
    const {
      children,
      originalIndex
    } = row.rowData.metadata;

    if (children && children.length > 0) {
      return _react.default.createElement("div", {
        className: cx(styles.expand_caret),
        role: "button",
        tabIndex: 0,
        onClick: expandRow(originalIndex),
        onKeyPress: expandRow(originalIndex)
      }, _react.default.createElement(_Spacing.default, {
        left: 1.5
      }, _react.default.createElement(_ExpandableIcon.default, {
        expanded: expandedRows.has(originalIndex),
        size: "1.6em"
      })));
    }

    return _react.default.createElement("div", {
      className: cx(styles.row)
    });
  };

  return _react.default.createElement(_reactVirtualized.Column, {
    dataKey: "expanded",
    cellRenderer: cellRenderer,
    width: _constants.EXPANDABLE_COLUMN_WIDTH
  });
}
