// @airbnb/lunar/lib/components/HierarchyPicker/styles.js
"use strict";

exports.__esModule = true;
exports.styleSheet = exports.default = void 0;

var _buildInputStyles = _interopRequireDefault(require("../../themes/buildInputStyles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const styleSheet = theme => _extends({}, (0, _buildInputStyles.default)(theme), {
  selectlike: {
    position: 'relative'
  },
  selectlike_enabled: {
    cursor: 'pointer'
  },
  arrow: {
    position: 'absolute',
    right: 0,
    top: 0,
    bottom: 0,
    overflow: 'hidden',
    width: theme.unit * 5,
    pointerEvents: 'none',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center'
  }
});

exports.styleSheet = styleSheet;
var _default = styleSheet;
exports.default = _default;
