// @airbnb/lunar/lib/utils/getLanguageFromLocale.js
"use strict";

exports.__esModule = true;
exports.default = getLanguageFromLocale;

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let loaded = false;
let languages = {};

function getLanguageFromLocale(locale) {
  if (!loaded) {
    loaded = true;
    languages = {
      nl: _.default.translate('Dutch', {}, 'Language name within a language selector'),
      en: _.default.translate('English', {}, 'Language name within a language selector'),
      fr: _.default.translate('French', {}, 'Language name within a language selector'),
      de: _.default.translate('German', {}, 'Language name within a language selector'),
      it: _.default.translate('Italian', {}, 'Language name within a language selector'),
      ja: _.default.translate('Japanese', {}, 'Language name within a language selector'),
      ko: _.default.translate('Korean', {}, 'Language name within a language selector'),
      zh: _.default.translate('Mandarin', {}, 'Language name within a language selector'),
      ms: _.default.translate('Malay', {}, 'Language name within a language selector'),
      pt: _.default.translate('Portuguese', {}, 'Language name within a language selector'),
      ru: _.default.translate('Russian', {}, 'Language name within a language selector'),
      es: _.default.translate('Spanish', {}, 'Language name within a language selector'),
      tr: _.default.translate('Turkish', {}, 'Language name within a language selector')
    };
  }

  return languages[locale];
}
