// @vx/legend/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Legend", {
  enumerable: true,
  get: function get() {
    return _Legend.default;
  }
});
Object.defineProperty(exports, "LegendQuantile", {
  enumerable: true,
  get: function get() {
    return _Quantile.default;
  }
});
Object.defineProperty(exports, "LegendLinear", {
  enumerable: true,
  get: function get() {
    return _Linear.default;
  }
});
Object.defineProperty(exports, "LegendOrdinal", {
  enumerable: true,
  get: function get() {
    return _Ordinal.default;
  }
});
Object.defineProperty(exports, "LegendThreshold", {
  enumerable: true,
  get: function get() {
    return _Threshold.default;
  }
});
Object.defineProperty(exports, "LegendSize", {
  enumerable: true,
  get: function get() {
    return _Size.default;
  }
});
Object.defineProperty(exports, "LegendItem", {
  enumerable: true,
  get: function get() {
    return _LegendItem.default;
  }
});
Object.defineProperty(exports, "LegendLabel", {
  enumerable: true,
  get: function get() {
    return _LegendLabel.default;
  }
});
Object.defineProperty(exports, "LegendShape", {
  enumerable: true,
  get: function get() {
    return _LegendShape.default;
  }
});

var _Legend = _interopRequireDefault(require("./legends/Legend"));

var _Quantile = _interopRequireDefault(require("./legends/Quantile"));

var _Linear = _interopRequireDefault(require("./legends/Linear"));

var _Ordinal = _interopRequireDefault(require("./legends/Ordinal"));

var _Threshold = _interopRequireDefault(require("./legends/Threshold"));

var _Size = _interopRequireDefault(require("./legends/Size"));

var _LegendItem = _interopRequireDefault(require("./legends/LegendItem"));

var _LegendLabel = _interopRequireDefault(require("./legends/LegendLabel"));

var _LegendShape = _interopRequireDefault(require("./legends/LegendShape"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
