// @airbnb/lunar/lib/components/Copy/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _copyToClipboard = _interopRequireDefault(require("copy-to-clipboard"));

var _IconCopy = _interopRequireDefault(require("@airbnb/lunar-icons/lib/interface/IconCopy"));

var _Translate = _interopRequireDefault(require("../Translate"));

var _Tooltip = _interopRequireDefault(require("../Tooltip"));

var _Link = _interopRequireDefault(require("../Link"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** A component for easily copying a string of text to the clipboard. */
class Copy extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      copied: false
    });

    _defineProperty(this, "handleClick", event => {
      const {
        text,
        onCopy
      } = this.props;
      const result = (0, _copyToClipboard.default)(text);
      event.preventDefault();
      this.setState({
        copied: true
      });

      if (onCopy) {
        onCopy(text, result);
      }
    });

    _defineProperty(this, "handleMouseLeave", () => {
      window.setTimeout(() => {
        this.setState({
          copied: false
        });
      }, 500);
    });
  }

  render() {
    const {
      prompt,
      children,
      id,
      trackingName,
      underlined
    } = this.props;

    const element = children || // eslint-disable-next-line jsx-a11y/anchor-is-valid
    _react.default.createElement(_Link.default, {
      trackingName: trackingName,
      id: id
    }, _react.default.createElement(_IconCopy.default, {
      decorative: true
    }));

    return _react.default.createElement(_Tooltip.default, {
      remainOnMouseDown: true,
      content: this.state.copied ? _react.default.createElement(_Translate.default, {
        k: "lunar.copy.copied",
        phrase: "Copied!",
        context: "Text has been copied to a clipboard"
      }) : prompt || _react.default.createElement(_Translate.default, {
        k: "lunar.copy.copyToClipboard",
        phrase: "Copy to clipboard",
        context: "Message informing user to copy test to a clipboard"
      }),
      underlined: underlined
    }, _react.default.cloneElement(element, {
      onClick: this.handleClick,
      onMouseLeave: this.handleMouseLeave
    }));
  }

}

exports.default = Copy;

_defineProperty(Copy, "propTypes", {
  children: _propTypes.default.element,
  id: _propTypes.default.string,
  text: _propTypes.default.string.isRequired,
  onCopy: _propTypes.default.func,
  prompt: _propTypes.default.node,
  trackingName: _propTypes.default.string,
  underlined: _propTypes.default.bool
});
