// @airbnb/lunar/lib/components/DateTimeSelect/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.DateTimeSelect = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _v = _interopRequireDefault(require("uuid/v4"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _BaseSelect = _interopRequireDefault(require("../private/BaseSelect"));

var _FormField = _interopRequireWildcard(require("../FormField"));

var _Translate = _interopRequireDefault(require("../Translate"));

var _createRange = _interopRequireDefault(require("../../utils/createRange"));

var _createDateTime = _interopRequireDefault(require("../../utils/createDateTime"));

var _getMonths = _interopRequireDefault(require("../../utils/getMonths"));

var _styles = require("./styles");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** An uncontrolled multi-select field for date and time ranges in UTC. */
class DateTimeSelect extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "date", (0, _createDateTime.default)(this.props.value, {
      locale: this.props.locale,
      timezone: this.props.timezone
    }).set({
      minute: 0,
      second: 0
    }));

    _defineProperty(this, "state", {
      id: (0, _v.default)(),
      date: this.date,
      meridiem: this.date.get('hour') <= 11 ? 'am' : 'pm'
    });

    _defineProperty(this, "handleChange", (value, event) => {
      const {
        id
      } = event.target;
      const key = id.split('_')[1];
      this.setState(prevState => {
        let {
          date
        } = prevState;
        const meridiem = key === 'meridiem' ? value : prevState.meridiem;

        if (this.props.enable12HourClock && (key === 'hour' || key === 'meridiem')) {
          let hour = Number(key === 'hour' ? value : date.get('hour'));

          if (hour <= 12 && meridiem === 'pm') {
            hour += 12;

            if (hour === 24) {
              hour = 12;
            }
          } else if (hour >= 12 && meridiem === 'am') {
            hour -= 12;
          }

          date = date.set({
            hour
          });
        } else {
          date = date.set({
            [key]: value
          });
        }

        return {
          date,
          meridiem
        };
      }, () => {
        this.props.onChange(this.state.date.toISO(), event);
      });
    });
  }

  componentDidUpdate(prevProps) {
    const {
      value,
      locale,
      timezone
    } = this.props; // Don't set minute/second to 0 here, because when used in conjunction with the form kit,
    // the value is always passed down, causing the numbers to always reset to 0.

    if (value !== prevProps.value) {
      const date = (0, _createDateTime.default)(value, {
        locale,
        timezone
      });
      this.setState({
        date,
        meridiem: date.get('hour') <= 11 ? 'am' : 'pm'
      });
    }
  }

  getDayRange() {
    return (0, _createRange.default)(1, this.state.date.daysInMonth).map(day => ({
      label: day,
      value: String(day)
    }));
  }

  getCurrentValue(date, key) {
    let value = Number(date.get(key));

    if (key === 'meridiem') {
      value = this.state.meridiem;
    } else if (key === 'hour' && this.props.enable12HourClock) {
      if (value === 0) {
        value = 12;
      } else if (value > 12) {
        value -= 12;
      }
    }

    return String(value);
  }

  getHourRange() {
    return (this.props.enable12HourClock ? (0, _createRange.default)(1, 12) : (0, _createRange.default)(0, 23)).map(hour => ({
      label: hour,
      value: hour
    }));
  }

  getMinuteRange() {
    return (0, _createRange.default)(0, 59, this.props.minuteStep).map(minute => ({
      label: minute.padStart(2, '0'),
      value: minute
    }));
  }

  getMonthRange() {
    return (0, _getMonths.default)().map((month, i) => ({
      label: month,
      value: String(i + 1)
    }));
  }

  getYearRange() {
    const now = (0, _createDateTime.default)();
    return (0, _createRange.default)(now.year - this.props.yearPastBuffer, now.year + this.props.yearFutureBuffer).reverse().map(year => ({
      label: year,
      value: year
    }));
  }

  render() {
    const {
      fieldProps,
      inputProps
    } = (0, _FormField.partitionFieldProps)(this.props);

    const {
      cx,
      name,
      styles,
      // Omit everything
      enable12HourClock,
      hideDate,
      hideTime,
      hideYear
    } = inputProps,
          restProps = _objectWithoutPropertiesLoose(inputProps, ["cx", "name", "styles", "enable12HourClock", "hideDate", "hideTime", "hideYear", "minuteStep", "yearFutureBuffer", "yearPastBuffer"]);

    const {
      id,
      date
    } = this.state;
    return _react.default.createElement(_FormField.default, _extends({}, fieldProps, {
      id: id
    }), _react.default.createElement("div", {
      className: cx(styles.selects)
    }, !hideDate && _react.default.createElement(_react.default.Fragment, null, _react.default.createElement(_BaseSelect.default, _extends({}, restProps, {
      id: id + "_month",
      name: name + "[month]",
      value: this.getCurrentValue(date, 'month'),
      placeholder: _Translate.default.phrase('Month', {}, {
        context: 'Month dropdown in a form datetime field',
        key: 'lunar.common.month'
      }),
      onChange: this.handleChange
    }), this.getMonthRange().map(month => _react.default.createElement("option", {
      key: month.value,
      value: month.value
    }, month.label))), _react.default.createElement("div", {
      className: cx(styles.spacer)
    }), _react.default.createElement(_BaseSelect.default, _extends({}, restProps, {
      id: id + "_day",
      name: name + "[day]",
      value: this.getCurrentValue(date, 'day'),
      placeholder: _Translate.default.phrase('Day', {}, {
        context: 'Day dropdown in a form datetime field',
        key: 'lunar.common.day'
      }),
      onChange: this.handleChange
    }), this.getDayRange().map(day => _react.default.createElement("option", {
      key: day.value,
      value: day.value
    }, day.label))), _react.default.createElement("div", {
      className: cx(styles.spacer)
    }), !hideYear && _react.default.createElement(_BaseSelect.default, _extends({}, restProps, {
      id: id + "_year",
      name: name + "[year]",
      value: this.getCurrentValue(date, 'year'),
      placeholder: _Translate.default.phrase('Year', {}, {
        context: 'Year dropdown in a form datetime field',
        key: 'lunar.common.year'
      }),
      onChange: this.handleChange
    }), this.getYearRange().map(year => _react.default.createElement("option", {
      key: year.value,
      value: year.value
    }, year.label))), !hideTime && _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("div", {
      className: cx(styles.spacer)
    }), _react.default.createElement("div", {
      className: cx(styles.spacer)
    }))), !hideTime && _react.default.createElement(_react.default.Fragment, null, _react.default.createElement(_BaseSelect.default, _extends({}, restProps, {
      id: id + "_hour",
      name: name + "[hour]",
      value: this.getCurrentValue(date, 'hour'),
      placeholder: _Translate.default.phrase('Hour', {}, {
        context: 'Hour dropdown in a form datetime field',
        key: 'lunar.common.hour'
      }),
      onChange: this.handleChange
    }), this.getHourRange().map(hour => _react.default.createElement("option", {
      key: hour.value,
      value: hour.value
    }, hour.label))), _react.default.createElement("div", {
      className: cx(styles.spacer)
    }, ":"), _react.default.createElement(_BaseSelect.default, _extends({}, restProps, {
      id: id + "_minute",
      name: name + "[minute]",
      value: this.getCurrentValue(date, 'minute'),
      placeholder: _Translate.default.phrase('Minute', {}, {
        context: 'Minute dropdown in a form datetime field',
        key: 'lunar.common.minute'
      }),
      onChange: this.handleChange
    }), this.getMinuteRange().map(minute => _react.default.createElement("option", {
      key: minute.value,
      value: minute.value
    }, minute.label))), enable12HourClock && _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("div", {
      className: cx(styles.spacer)
    }), _react.default.createElement(_BaseSelect.default, _extends({}, restProps, {
      id: id + "_meridiem",
      name: name + "[meridiem]",
      value: this.getCurrentValue(date, 'meridiem'),
      placeholder: _Translate.default.phrase('Meridiem', {}, {
        context: 'Meridiem (AM/PM) dropdown in a form datetime field',
        key: 'lunar.common.meridiem'
      }),
      onChange: this.handleChange
    }), _react.default.createElement("option", {
      value: "am"
    }, _Translate.default.phrase('AM', {}, {
      context: 'Meridiem for timestamps',
      key: 'lunar.common.meridiemAM'
    })), _react.default.createElement("option", {
      value: "pm"
    }, _Translate.default.phrase('PM', {}, {
      context: 'Meridiem for timestamps',
      key: 'lunar.common.meridiemPM'
    })))))));
  }

}

exports.DateTimeSelect = DateTimeSelect;

_defineProperty(DateTimeSelect, "propTypes", {
  enable12HourClock: _propTypes.default.bool,
  hideDate: _propTypes.default.bool,
  hideTime: _propTypes.default.bool,
  hideYear: _propTypes.default.bool,
  locale: _propTypes.default.any,
  minuteStep: _propTypes.default.number,
  placeholder: _propTypes.default.string,
  onChange: _propTypes.default.func.isRequired,
  timezone: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.bool]),
  yearFutureBuffer: _propTypes.default.number,
  yearPastBuffer: _propTypes.default.number,
  value: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number, _propTypes.default.instanceOf(Date), _propTypes.default.any])
});

_defineProperty(DateTimeSelect, "defaultProps", {
  enable12HourClock: false,
  hideDate: false,
  hideTime: false,
  hideYear: false,
  minuteStep: 5,
  yearFutureBuffer: 5,
  yearPastBuffer: 80
});

var _default = (0, _withStyles.default)(_styles.styleSheet)(DateTimeSelect);

exports.default = _default;
