// @airbnb/lunar/lib/components/ErrorMessage/index.js
"use strict";

exports.__esModule = true;
exports.getErrorMessage = getErrorMessage;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Translate = _interopRequireDefault(require("../Translate"));

var _Alert = _interopRequireDefault(require("../Alert"));

var _MutedButton = _interopRequireDefault(require("../MutedButton"));

var _Spacing = _interopRequireDefault(require("../Spacing"));

var _StatusText = _interopRequireDefault(require("../StatusText"));

var _ = _interopRequireDefault(require("../.."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function getErrorMessage(error, includeCode) {
  if (includeCode === void 0) {
    includeCode = false;
  }

  if (typeof error === 'string') {
    return error;
  }

  if (error instanceof Error) {
    return error.message;
  }

  const debug = error.debug_info;
  const message = debug ? debug.error_class + " - " + (debug.error_message || debug.response_message) : error.user_message || error.error_details || error.error_message;

  if (includeCode && error.error_code) {
    return error.error_code + " - " + message;
  }

  return message || '';
} // istanbul ignore next


function createRedirectURL(id, url) {
  return () => window.open(url || _.default.settings.errorURL.replace('{{id}}', id), '_blank');
}

/** Display an error message from an `Error` instance or API endpoint. */
class ErrorMessage extends _react.default.PureComponent {
  render() {
    const {
      error,
      inline,
      title,
      subtitle,
      onClose
    } = this.props;

    if (!error) {
      return null;
    }

    const message = subtitle || getErrorMessage(error);
    const code = error instanceof Error ? null : error.error_code;
    const id = error instanceof Error ? null : error.error_id;
    const url = error instanceof Error ? '' : error.error_url;

    if (inline) {
      return _react.default.createElement(_StatusText.default, {
        danger: true
      }, message);
    }

    return _react.default.createElement(_Alert.default, {
      danger: true,
      title: title || code || _react.default.createElement(_Translate.default, {
        k: "lunar.error.unknown",
        phrase: "Unknown error",
        context: "Title for an unknown error occurred"
      }),
      onClose: onClose
    }, message, id && _react.default.createElement(_Spacing.default, {
      top: 1
    }, _react.default.createElement(_MutedButton.default, {
      inverted: true,
      onClick: createRedirectURL(id, url)
    }, _react.default.createElement(_Translate.default, {
      k: "lunar.error.viewDetails",
      phrase: "View error details",
      context: "Button label to view details about an error"
    }))));
  }

}

exports.default = ErrorMessage;

_defineProperty(ErrorMessage, "propTypes", {
  error: _propTypes.default.any,
  inline: _propTypes.default.bool,
  title: _propTypes.default.node,
  subtitle: _propTypes.default.node,
  onClose: _propTypes.default.func
});

_defineProperty(ErrorMessage, "defaultProps", {
  inline: false,
  subtitle: null,
  title: null
});
