// @airbnb/lunar/esm/components/List/Item.js
import _pt from "prop-types";
import React from 'react';
import { mutuallyExclusiveTrueProps, forbidExtraProps } from 'airbnb-prop-types';
import useStyles from '../../hooks/useStyles';
import { styleSheetItem as styleSheet } from './styles';

function ListItem(_ref) {
  let {
    bordered,
    children,
    compact,
    horizontal,
    spacious
  } = _ref;
  const [styles, cx] = useStyles(styleSheet);
  return React.createElement("li", {
    className: cx(!horizontal && bordered && styles.item_bordered, horizontal && bordered && styles.item_bordered_horizontal, !horizontal && compact && styles.item_compact, horizontal && compact && styles.item_compact_horizontal, !horizontal && spacious && styles.item_spacious, horizontal && spacious && styles.item_spacious_horizontal)
  }, children);
}

const paddingPropType = mutuallyExclusiveTrueProps('compact', 'spacious');
ListItem.propTypes = {
  horizontal: _pt.bool,
  children: _pt.any.isRequired,
  bordered: _pt.bool,
  compact: paddingPropType,
  spacious: paddingPropType
};
export default ListItem;
