// @airbnb/lunar/lib/components/Loader/index.js
"use strict";

exports.__esModule = true;
exports.default = Loader;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** A small 3-dot loading indicator. */
function Loader(_ref) {
  let {
    inline,
    inverted,
    large,
    static: isStatic
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheet);
  return _react.default.createElement("div", {
    className: cx(styles.loader, inline && styles.loader_inline, !isStatic && !inline && styles.loader_absolute)
  }, [1, 2, 3].map(no => _react.default.createElement("span", {
    key: no,
    className: cx(styles.dot, styles["dot_" + no], large && styles.dot_large, inverted && styles.dot_inverted)
  })));
}

Loader.propTypes = {
  inline: _propTypes.default.bool,
  inverted: _propTypes.default.bool,
  large: _propTypes.default.bool,
  static: _propTypes.default.bool
};
