// @airbnb/lunar/lib/components/DataTable/constants.js
"use strict";

exports.__esModule = true;
exports.SELECTABLE_COLUMN_WIDTH = exports.EXPANDABLE_COLUMN_WIDTH = exports.DEFAULT_WIDTH_PROPERTIES = exports.HEIGHT_TO_PX = exports.STATUS_OPTIONS = exports.SELECTION_OPTIONS = void 0;
const SELECTION_OPTIONS = {
  ACTIVE: 'ACTIVE',
  DISABLED: 'DISABLED',
  INACTIVE: 'INACTIVE',
  HAS_ACTIVE_CHILD: 'HAS_ACTIVE_CHILD'
};
exports.SELECTION_OPTIONS = SELECTION_OPTIONS;
const STATUS_OPTIONS = {
  ALERT: 'ALERT',
  WARNING: 'WARNING'
};
exports.STATUS_OPTIONS = STATUS_OPTIONS;
const HEIGHT_TO_PX = {
  micro: 32,
  small: 48,
  regular: 56,
  large: 64,
  jumbo: 108
};
exports.HEIGHT_TO_PX = HEIGHT_TO_PX;
const DEFAULT_WIDTH_PROPERTIES = {
  width: 50,
  flexGrow: 1,
  flexShrink: 1,
  minWidth: 0
};
exports.DEFAULT_WIDTH_PROPERTIES = DEFAULT_WIDTH_PROPERTIES;
const EXPANDABLE_COLUMN_WIDTH = 50;
exports.EXPANDABLE_COLUMN_WIDTH = EXPANDABLE_COLUMN_WIDTH;
const SELECTABLE_COLUMN_WIDTH = 50;
exports.SELECTABLE_COLUMN_WIDTH = SELECTABLE_COLUMN_WIDTH;
