// aesthetic-adapter-aphrodite/lib/AphroditeAesthetic.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _aesthetic = _interopRequireWildcard(require("aesthetic"));

var _aestheticUtils = require("aesthetic-utils");

var _aphrodite = require("aphrodite");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var AphroditeAesthetic = function (_Aesthetic) {
  _inheritsLoose(AphroditeAesthetic, _Aesthetic);

  function AphroditeAesthetic(extensions, options) {
    var _this;

    if (extensions === void 0) {
      extensions = [];
    }

    if (options === void 0) {
      options = {};
    }

    _this = _Aesthetic.call(this, options) || this;

    _defineProperty(_assertThisInitialized(_this), "aphrodite", void 0);

    _defineProperty(_assertThisInitialized(_this), "fontFaces", {});

    _defineProperty(_assertThisInitialized(_this), "keyframes", {});

    _defineProperty(_assertThisInitialized(_this), "handleCss", function (css) {
      _this.getStyleSheetManager().injectStatements(css);
    });

    _defineProperty(_assertThisInitialized(_this), "handleFontFace", function (sheet, fontFaces, fontFamily) {
      _this.fontFaces[fontFamily] = fontFaces.map(function (face) {
        return face.toObject();
      });
    });

    _defineProperty(_assertThisInitialized(_this), "handleGlobal", function (sheet, selector, ruleset) {
      var current = sheet.ruleSets.globals || sheet.createRuleset('globals');
      current.addNested("*" + selector, ruleset);
      sheet.addRuleset(current);
    });

    _defineProperty(_assertThisInitialized(_this), "handleKeyframe", function (sheet, keyframe, animationName) {
      _this.keyframes[animationName] = keyframe.toObject();
    });

    _defineProperty(_assertThisInitialized(_this), "handleMedia", function (ruleset, query, value) {
      ruleset.addNested("@media " + query, value);
    });

    _defineProperty(_assertThisInitialized(_this), "handleNested", function (ruleset, selector, value) {
      ruleset.addNested(selector, value);
    });

    _defineProperty(_assertThisInitialized(_this), "handleProperty", function (ruleset, name, value) {
      if (name === 'animationName') {
        ruleset.addCompoundProperty(name, _this.syntax.injectKeyframes(String(value), _this.keyframes));
      } else if (name === 'fontFamily') {
        if (String(value).includes(ruleset.selector)) {
          ruleset.addProperty(name, value);
        } else {
          ruleset.addCompoundProperty(name, _this.syntax.injectFontFaces(String(value), _this.fontFaces));
        }
      } else {
        ruleset.addProperty(name, value);
      }
    });

    _this.aphrodite = _aphrodite.StyleSheet.extend([].concat(extensions, [{
      selectorHandler: _this.handleHierarchySelector
    }, {
      selectorHandler: _this.handleGlobalSelector
    }]));

    _this.syntax.on('attribute', _this.handleNested).on('css', _this.handleCss).on('font-face', _this.handleFontFace).on('global', _this.handleGlobal).on('keyframe', _this.handleKeyframe).on('media', _this.handleMedia).on('property', _this.handleProperty).on('pseudo', _this.handleNested).on('selector', _this.handleNested);

    return _this;
  }

  var _proto = AphroditeAesthetic.prototype;

  _proto.flushStyles = function flushStyles() {
    (0, _aphrodite.flushToStyleTag)();
  };

  _proto.isParsedBlock = function isParsedBlock(block) {
    return Boolean(block && block._name && block._definition);
  };

  _proto.parseStyleSheet = function parseStyleSheet(styleSheet) {
    return this.aphrodite.StyleSheet.create(styleSheet);
  };

  _proto.purgeStyles = function purgeStyles(styleName) {
    (0, _aestheticUtils.purgeStyles)((0, _aestheticUtils.getStyleElements)('data-aphrodite'), styleName === _aesthetic.GLOBAL_STYLE_NAME);

    if (styleName) {
      (0, _aphrodite.resetInjectedStyle)(styleName);
    }
  };

  _proto.transformToClassName = function transformToClassName(styles) {
    var _this$aphrodite;

    return (_this$aphrodite = this.aphrodite).css.apply(_this$aphrodite, styles);
  };

  _proto.handleGlobalSelector = function handleGlobalSelector(selector, baseSelector, callback) {
    if (selector.charAt(0) !== '*') {
      return null;
    }

    return callback(selector.slice(1));
  };

  _proto.handleHierarchySelector = function handleHierarchySelector(selector, baseSelector, callback) {
    if (selector.charAt(0) === '>') {
      return callback(baseSelector + " " + selector);
    }

    if (selector.charAt(0) === '[') {
      return callback("" + baseSelector + selector);
    }

    return null;
  };

  return AphroditeAesthetic;
}(_aesthetic.default);

exports.default = AphroditeAesthetic;
