// @airbnb/lunar/esm/components/Text/index.js
import _pt from "prop-types";
import React from 'react';
import { mutuallyExclusiveTrueProps, forbidExtraProps } from 'airbnb-prop-types';
import useStyles from '../../hooks/useStyles';
import { styleSheet } from './styles';

/** Display a string of text with pre-defined sizing, emphasis, and state styling. */
function Text(_ref) {
  let {
    baseline,
    bold,
    centerAlign,
    children,
    disabled,
    endAlign,
    inline,
    inverted,
    large,
    light,
    micro,
    muted,
    noWrap,
    preserveWhitespace,
    small,
    startAlign,
    truncated,
    uppercased
  } = _ref;
  const [styles, cx] = useStyles(styleSheet);
  let Tag = 'div';

  if (inline) {
    Tag = 'span';
  } else if (micro) {
    Tag = 'small';
  } else if (large) {
    Tag = 'h4';
  }

  return React.createElement(Tag, {
    className: cx(styles.text, bold && styles.text_bold, disabled && styles.text_disabled, inline && styles.text_inline, baseline && styles.text_baseline, inverted && styles.text_inverted, large && styles.text_large, light && styles.text_light, micro && styles.text_micro, muted && styles.text_muted, preserveWhitespace && styles.text_preserveWhitespace, small && styles.text_small, truncated && styles.text_truncated, uppercased && styles.text_uppercased, micro && uppercased && styles.text_uppercased_micro, centerAlign && styles.text_center, endAlign && styles.text_end, startAlign && styles.text_start, noWrap && styles.text_noWrap)
  }, children);
}

const sizingProp = mutuallyExclusiveTrueProps('micro', 'small', 'large');
const emphasisProp = mutuallyExclusiveTrueProps('bold', 'light');
const stateProp = mutuallyExclusiveTrueProps('disabled', 'muted', 'inverted');
const alignProp = mutuallyExclusiveTrueProps('centerAlign', 'endAlign', 'startAlign');
Text.propTypes = {
  uppercased: _pt.bool,
  truncated: _pt.bool,
  preserveWhitespace: _pt.bool,
  noWrap: _pt.bool,
  inline: _pt.bool,
  children: _pt.node,
  baseline: _pt.bool,
  bold: emphasisProp,
  centerAlign: alignProp,
  disabled: stateProp,
  endAlign: alignProp,
  inverted: stateProp,
  large: sizingProp,
  light: emphasisProp,
  micro: sizingProp,
  muted: stateProp,
  small: sizingProp,
  startAlign: alignProp
};
export default Text;
