// @airbnb/lunar/lib/components/Glyph/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

// https://developer.mozilla.org/en-US/docs/Web/CSS/font-variant-numeric
const fractionProp = (0, _airbnbPropTypes.mutuallyExclusiveTrueProps)('diagonal', 'stacked'); // istanbul ignore next

function getNumericVariant(props) {
  const variants = [];

  if (props.diagonal) {
    variants.push('diagonal-fractions');
  } else if (props.stacked) {
    variants.push('stacked-fractions');
  }

  if (props.ordinal) {
    variants.push('ordinal');
  }

  if (props.slashed) {
    variants.push('slashed-zero');
  }

  if (props.tabular) {
    variants.push('tabular-nums');
  }

  return variants.length > 0 ? variants.join(' ') : 'normal';
}

/** Controls hidden and alternative glyphs within the current font. */
function Glyph(_ref) {
  let {
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, ["children"]);

  const [, cx] = (0, _useStyles.default)(() => ({}));
  return _react.default.createElement("span", {
    className: cx({
      fontVariantNumeric: getNumericVariant(props)
    })
  }, children);
}

Glyph.propTypes = {
  tabular: _propTypes.default.bool,
  slashed: _propTypes.default.bool,
  ordinal: _propTypes.default.bool,
  children: _propTypes.default.any.isRequired,
  diagonal: fractionProp,
  stacked: fractionProp
};
var _default = Glyph;
exports.default = _default;
