// @airbnb/lunar/esm/components/ErrorMessage/index.js
import _pt from "prop-types";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import T from '../Translate';
import Alert from '../Alert';
import MutedButton from '../MutedButton';
import Spacing from '../Spacing';
import StatusText from '../StatusText';
import Core from '../..';
export function getErrorMessage(error, includeCode) {
  if (includeCode === void 0) {
    includeCode = false;
  }

  if (typeof error === 'string') {
    return error;
  }

  if (error instanceof Error) {
    return error.message;
  }

  const debug = error.debug_info;
  const message = debug ? debug.error_class + " - " + (debug.error_message || debug.response_message) : error.user_message || error.error_details || error.error_message;

  if (includeCode && error.error_code) {
    return error.error_code + " - " + message;
  }

  return message || '';
} // istanbul ignore next

function createRedirectURL(id, url) {
  return () => window.open(url || Core.settings.errorURL.replace('{{id}}', id), '_blank');
}

/** Display an error message from an `Error` instance or API endpoint. */
export default class ErrorMessage extends React.PureComponent {
  render() {
    const {
      error,
      inline,
      title,
      subtitle,
      onClose
    } = this.props;

    if (!error) {
      return null;
    }

    const message = subtitle || getErrorMessage(error);
    const code = error instanceof Error ? null : error.error_code;
    const id = error instanceof Error ? null : error.error_id;
    const url = error instanceof Error ? '' : error.error_url;

    if (inline) {
      return React.createElement(StatusText, {
        danger: true
      }, message);
    }

    return React.createElement(Alert, {
      danger: true,
      title: title || code || React.createElement(T, {
        k: "lunar.error.unknown",
        phrase: "Unknown error",
        context: "Title for an unknown error occurred"
      }),
      onClose: onClose
    }, message, id && React.createElement(Spacing, {
      top: 1
    }, React.createElement(MutedButton, {
      inverted: true,
      onClick: createRedirectURL(id, url)
    }, React.createElement(T, {
      k: "lunar.error.viewDetails",
      phrase: "View error details",
      context: "Button label to view details about an error"
    }))));
  }

}

_defineProperty(ErrorMessage, "propTypes", {
  error: _pt.any,
  inline: _pt.bool,
  title: _pt.node,
  subtitle: _pt.node,
  onClose: _pt.func
});

_defineProperty(ErrorMessage, "defaultProps", {
  inline: false,
  subtitle: null,
  title: null
});
