// @airbnb/lunar/esm/components/ImageViewer/index.js
import _pt from "prop-types";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import withStyles from '../../composers/withStyles';
import ZoomControls from './ZoomControls';
import RotateControls from './RotateControls';
import ResponsiveImage from '../ResponsiveImage';
import { styleSheet } from './styles';

/** An image viewer that can zoom, drag, and rotate an image. */
export class ImageViewer extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      dragging: false,
      imageLocation: {
        x: 0,
        y: 0
      },
      lastMouseLocation: {
        x: 0,
        y: 0
      }
    });

    _defineProperty(this, "handleMouseDown", event => {
      event.preventDefault();
      this.setState({
        dragging: true,
        lastMouseLocation: {
          x: event.pageX,
          y: event.pageY
        }
      });
    });

    _defineProperty(this, "handleMouseUp", event => {
      event.preventDefault();
      this.setState({
        dragging: false,
        lastMouseLocation: {
          x: 0,
          y: 0
        }
      });
    });

    _defineProperty(this, "handleMouseMove", event => {
      event.preventDefault();

      if (this.state.dragging) {
        this.setState((_ref) => {
          let {
            dragging,
            imageLocation,
            lastMouseLocation
          } = _ref;
          const xDiff = lastMouseLocation.x - event.pageX;
          const yDiff = lastMouseLocation.y - event.pageY;
          return {
            imageLocation: {
              x: imageLocation.x - xDiff,
              y: imageLocation.y - yDiff
            },
            lastMouseLocation: {
              x: event.pageX,
              y: event.pageY
            }
          };
        });
      }
    });
  }

  componentDidMount() {
    document.addEventListener('mousemove', this.handleMouseMove, false);
  }

  componentWillUnmount() {
    document.removeEventListener('mousemove', this.handleMouseMove, false);
  }

  getTransformStyle() {
    const {
      imageLocation: {
        x,
        y
      }
    } = this.state;
    const {
      scale,
      rotation
    } = this.props; // rotation and scale have defaultProp values

    const radian = rotation / 180 * Math.PI;
    const sinRotation = Math.sin(radian);
    const cosRotation = Math.cos(radian);
    const translateX = (y * sinRotation + x * cosRotation) / scale;
    const translateY = (y * cosRotation - x * sinRotation) / scale;
    return {
      transform: "scale(" + scale + ") rotate(" + rotation + "deg) translateY(" + translateY + "px) translateX(" + translateX + "px)"
    };
  }

  render() {
    const {
      alt,
      borderless,
      height,
      src,
      width,
      cx,
      styles
    } = this.props;
    return React.createElement("div", {
      className: cx(styles.container, borderless && styles.container_borderless),
      role: "presentation",
      style: {
        width,
        height
      },
      onMouseDown: this.handleMouseDown,
      onMouseUp: this.handleMouseUp
    }, React.createElement("div", {
      className: cx(styles.image),
      style: this.getTransformStyle()
    }, React.createElement(ResponsiveImage, {
      contain: true,
      noShadow: true,
      alt: alt,
      borderRadius: 0,
      maxWidth: width,
      maxHeight: height,
      src: src
    })));
  }

}

_defineProperty(ImageViewer, "propTypes", {
  alt: _pt.string.isRequired,
  borderless: _pt.bool,
  height: _pt.oneOfType([_pt.number, _pt.string]),
  rotation: _pt.number,
  src: _pt.string.isRequired,
  scale: _pt.number,
  width: _pt.oneOfType([_pt.number, _pt.string])
});

_defineProperty(ImageViewer, "defaultProps", {
  height: 'none',
  rotation: 0,
  scale: 1,
  width: 'none'
});

export { ZoomControls, RotateControls };
export default withStyles(styleSheet)(ImageViewer);
