// @airbnb/lunar/lib/components/HierarchyPicker/Hierarchy/ItemDescription.js
"use strict";

exports.__esModule = true;
exports.default = ItemDescription;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Spacing = _interopRequireDefault(require("../../Spacing"));

var _Text = _interopRequireDefault(require("../../Text"));

var _Translate = _interopRequireDefault(require("../../Translate"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ItemDescription(_ref) {
  let {
    item
  } = _ref;
  return _react.default.createElement("div", null, _react.default.createElement(_Text.default, {
    bold: true
  }, _react.default.createElement(_Translate.default, {
    k: "lunar.common.description",
    phrase: "Description",
    context: "Description of item"
  })), _react.default.createElement(_Spacing.default, {
    top: 0.5
  }, _react.default.createElement(_Text.default, null, item.description)));
}

ItemDescription.propTypes = {
  item: _propTypes.default.any.isRequired
};
