// @airbnb/lunar/esm/components/Tabs/Tab.js
import _pt from "prop-types";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import camelCase from 'lodash/camelCase';
import upperFirst from 'lodash/upperFirst';
import withStyles from '../../composers/withStyles';
import ButtonOrLink from '../private/ButtonOrLink';
import TrackingBoundary from '../TrackingBoundary';
import { styleSheetTab as styleSheet } from './styles';

/** A single tab button. Usually rendered amongst a collection of tabs. */
export class Tab extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleClick", () => {
      if (this.props.onClick) {
        this.props.onClick(this.props.keyName);
      }
    });
  }

  componentDidUpdate(prevProps) {
    if (!prevProps.selected && this.props.selected && this.props.onSelected) {
      this.props.onSelected();
    }
  }

  render() {
    const {
      cx,
      afterIcon,
      beforeIcon,
      borderless,
      disabled,
      href,
      keyName,
      label,
      secondary,
      selected,
      small,
      stretched,
      styles
    } = this.props;
    const trackingName = upperFirst(camelCase(keyName || 'Tab'));
    const noBorder = secondary || borderless;
    const noHover = secondary || noBorder && disabled;
    return React.createElement("span", {
      className: cx(styles.tab, secondary && styles.tab_secondary, disabled && styles.tab_disabled, noBorder && styles.tab_noBorder, noHover && styles.tab_noHover, selected && !secondary && styles.tab_selected, stretched && styles.tab_stretched)
    }, React.createElement(TrackingBoundary, {
      name: trackingName
    }, React.createElement(ButtonOrLink, {
      flexAlign: true,
      "aria-selected": selected,
      afterIcon: afterIcon,
      beforeIcon: beforeIcon,
      disabled: disabled,
      href: href,
      role: "tab",
      className: cx(styles.tabButton, secondary && styles.tabButton_secondary, selected && styles.tabButton_selected, selected && secondary && styles.tabButton_secondary_selected, small && styles.tabButton_small, disabled && styles.tabButton_disabled),
      onClick: disabled ? undefined : this.handleClick
    }, label)));
  }

}

_defineProperty(Tab, "propTypes", {
  borderless: _pt.bool,
  children: _pt.node,
  keyName: _pt.string,
  label: _pt.any.isRequired,
  onClick: _pt.func,
  onSelected: _pt.func,
  secondary: _pt.bool,
  selected: _pt.bool,
  small: _pt.bool,
  stretched: _pt.bool
});

_defineProperty(Tab, "defaultProps", {
  borderless: false,
  children: null,
  secondary: false,
  selected: false,
  small: false,
  stretched: false
});

export default withStyles(styleSheet)(Tab);
