// @airbnb/lunar/lib/components/ProfilePhoto/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.ProfilePhoto = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const mutuallyExclusiveSizePropType = (0, _airbnbPropTypes.mutuallyExclusiveProps)(_propTypes.default.any, 'small', 'large', 'macro', 'size');
const namedSizePropType = (0, _airbnbPropTypes.and)([_propTypes.default.bool, mutuallyExclusiveSizePropType]);
const unitSizePropType = (0, _airbnbPropTypes.and)([_propTypes.default.number, mutuallyExclusiveSizePropType]);

/** Display a profile photo. */
class ProfilePhoto extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      src: this.props.imageSrc || ''
    });

    _defineProperty(this, "handleError", event => {
      const {
        fallbackImageSrc
      } = this.props;

      if (fallbackImageSrc) {
        this.setState({
          src: fallbackImageSrc
        });
      }
    });
  }

  componentDidUpdate(prevProps) {
    const {
      imageSrc
    } = this.props;

    if (imageSrc !== prevProps.imageSrc) {
      this.setState({
        src: imageSrc
      });
    }
  }

  render() {
    const {
      src
    } = this.state;
    const {
      cx,
      fallbackImageSrc,
      inline,
      macro,
      large,
      small,
      size,
      square,
      styles,
      title,
      theme
    } = this.props;
    const {
      unit
    } = theme;
    return _react.default.createElement("div", {
      className: cx(inline && styles.inline, styles.regular, macro && styles.macro, large && styles.large, small && styles.small, !!size && size > 0 && {
        height: size * unit,
        maxHeight: size * unit,
        maxWidth: size * unit,
        width: size * unit
      })
    }, _react.default.createElement("img", {
      className: cx(styles.image, styles.regular, !square && styles.roundedImage, macro && styles.macro, large && styles.large, small && styles.small, !!size && size > 0 && {
        height: size * unit,
        maxHeight: size * unit,
        maxWidth: size * unit,
        width: size * unit
      }),
      src: src,
      alt: title,
      title: title,
      onError: fallbackImageSrc ? this.handleError : undefined
    }));
  }

}

exports.ProfilePhoto = ProfilePhoto;

_defineProperty(ProfilePhoto, "propTypes", {
  title: _propTypes.default.string.isRequired,
  square: _propTypes.default.bool,
  inline: _propTypes.default.bool,
  imageSrc: _propTypes.default.string.isRequired,
  fallbackImageSrc: _propTypes.default.string,
  large: namedSizePropType,
  macro: namedSizePropType,
  size: unitSizePropType,
  small: namedSizePropType
});

_defineProperty(ProfilePhoto, "defaultProps", {
  inline: false,
  square: false
});

var _default = (0, _withStyles.default)(_styles.styleSheet, {
  passThemeProp: true
})(ProfilePhoto);

exports.default = _default;
