// @airbnb/lunar/esm/components/Toast/index.js
import _pt from "prop-types";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import { mutuallyExclusiveTrueProps, forbidExtraProps } from 'airbnb-prop-types';
import IconClose from '@airbnb/lunar-icons/lib/interface/IconClose';
import withStyles from '../../composers/withStyles';
import { getErrorMessage } from '../ErrorMessage';
import IconButton from '../IconButton';
import Button from '../Button';
import Text from '../Text';
import T from '../Translate';
import Spacing from '../Spacing';
import crosstab from '../../crosstab';
import { styleSheet } from './styles';
const statusPropType = mutuallyExclusiveTrueProps('danger', 'success');

/** Abstract component for displaying a toast message above the pages content. */
export class Toast extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "hideTimer", 0);

    _defineProperty(this, "state", {
      visible: false
    });

    _defineProperty(this, "showToast", () => {
      this.setState({
        visible: true
      }, () => {
        if (this.props.onOpen) {
          this.props.onOpen();
        }
      });
    });

    _defineProperty(this, "handleClosePress", () => {
      this.handleClose();

      if (this.props.crosstabClose) {
        crosstab.emit(this.crosstabCloseEvent());
      }
    });

    _defineProperty(this, "handleClose", () => {
      window.clearTimeout(this.hideTimer);
      this.setState({
        visible: false
      }, () => {
        // Wait for the transition
        window.setTimeout(() => {
          if (this.props.onClose) {
            this.props.onClose();
          }

          if (this.props.onRemove) {
            this.props.onRemove(this.props.id);
          }
        }, 150);
      });
    });
  }

  componentDidMount() {
    const {
      delay = 0,
      duration = 0,
      crosstabClose
    } = this.props;
    window.setTimeout(this.showToast, delay);

    if (duration > 0) {
      this.hideTimer = window.setTimeout(this.handleClose, delay + duration);
    }

    if (crosstabClose) {
      crosstab.on(this.crosstabCloseEvent(), this.handleClose);
    }
  }

  componentWillUnmount() {
    crosstab.off(this.crosstabCloseEvent(), this.handleClose);
  }

  /* istanbul ignore next */
  handleRefreshPress() {
    global.location.reload();
  }

  crosstabCloseEvent() {
    return "toast:crosstabClose:" + this.props.id;
  }

  render() {
    const {
      visible
    } = this.state;
    const {
      cx,
      styles,
      message,
      title,
      danger,
      success,
      refresh
    } = this.props;
    const isError = message instanceof Error;
    const failed = danger || isError;
    return React.createElement("div", {
      className: cx(styles.container, visible && styles.container_visible, failed && styles.container_danger, success && styles.container_success),
      role: "status"
    }, refresh ? React.createElement("div", null, React.createElement(Text, {
      bold: true,
      inverted: true
    }, message), React.createElement(Spacing, {
      top: 0.5
    }, React.createElement(Button, {
      small: true,
      onClick: this.handleRefreshPress
    }, React.createElement(T, {
      k: "lunar.common.refresh",
      phrase: "Refresh",
      context: "Refresh the page that was triggerd by a toast"
    })))) : React.createElement("div", null, title && React.createElement(Text, {
      bold: true,
      inverted: true
    }, title), React.createElement(Text, {
      inverted: true
    }, isError ? getErrorMessage(message, true) : message)), React.createElement("div", {
      className: cx(styles.right)
    }, React.createElement(IconButton, {
      inverted: true,
      onClick: this.handleClosePress
    }, React.createElement(IconClose, {
      size: "1.5em",
      accessibilityLabel: T.phrase('Close', {}, {
        context: 'Close a toast popup',
        key: 'lunar.common.close'
      })
    }))));
  }

}

_defineProperty(Toast, "propTypes", {
  title: _pt.node,
  refresh: _pt.bool,
  onRemove: _pt.func,
  onClose: _pt.func,
  onOpen: _pt.func,
  message: _pt.oneOfType([_pt.string, _pt.instanceOf(Error)]).isRequired,
  id: _pt.string.isRequired,
  duration: _pt.number,
  delay: _pt.number,
  crosstabClose: _pt.bool,
  danger: statusPropType,
  success: statusPropType
});

_defineProperty(Toast, "defaultProps", {
  danger: false,
  delay: 250,
  duration: 10000,
  refresh: false,
  success: false,
  title: null
});

export default withStyles(styleSheet)(Toast);
