// @airbnb/lunar/lib/components/ProfilePhotoGroup/index.js
"use strict";

exports.__esModule = true;
exports.default = ProfilePhotoGroup;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _useTheme = _interopRequireDefault(require("../../hooks/useTheme"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** Horizontally align `ProfilePhoto`s in a stacked fashion. */
function ProfilePhotoGroup(_ref) {
  let {
    children,
    max = 3,
    size = 5
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheet);
  const {
    unit
  } = (0, _useTheme.default)();
  const margin = {
    marginLeft: -(size * (unit / 2.5))
  };

  let photos = _react.default.Children.toArray(children).filter(child => !!child);

  let remainder = 0;

  if (max && photos.length > max) {
    remainder = photos.length - max;
    photos = photos.slice(0, max);
  }

  return _react.default.createElement("div", {
    className: cx(styles.group)
  }, photos.map((photo, i) => _react.default.createElement("div", {
    key: i,
    className: cx(styles.cell, margin)
  }, _react.default.cloneElement(photo, {
    size,
    inline: false,
    square: false
  }))), remainder > 0 && _react.default.createElement("div", {
    key: "remainder",
    className: cx(styles.cell, margin)
  }, _react.default.createElement("div", {
    className: cx(styles.remainder, {
      height: size * unit,
      maxHeight: size * unit,
      maxWidth: size * unit,
      width: size * unit
    })
  }, _react.default.createElement("span", {
    className: cx(styles.remainderNumber)
  }, "+" + remainder))));
}

ProfilePhotoGroup.propTypes = {
  children: _propTypes.default.any.isRequired,
  max: _propTypes.default.number,
  size: _propTypes.default.number
};
