// @airbnb/lunar/lib/components/ProfilePhotoGroup/styles.js
"use strict";

exports.__esModule = true;
exports.styleSheet = exports.default = void 0;

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const styleSheet = (_ref) => {
  let {
    color,
    font
  } = _ref;
  return {
    group: {
      display: 'flex',
      alignItems: 'stretch'
    },
    cell: {
      borderRadius: '50%',
      border: "1px solid " + color.base,
      background: color.core.neutral[6],
      ':first-of-type': {
        marginLeft: 0
      },
      ':empty': {
        display: 'none'
      }
    },
    remainder: _extends({}, font.textRegular, {
      position: 'relative'
    }),
    remainderNumber: {
      transform: 'translate(-50%, -50%)',
      position: 'absolute',
      top: '50%',
      left: '50%',
      color: color.base
    }
  };
};

exports.styleSheet = styleSheet;
var _default = styleSheet;
exports.default = _default;
