// @airbnb/lunar/esm/components/MutedButton/index.js
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from 'react';
import Button from '../Button';
export const StyledMutedButton = Button.extendStyles((_ref) => {
  let {
    color
  } = _ref;
  return {
    button: {
      backgroundColor: color.core.neutral[5],
      border: "2px solid " + color.core.neutral[5],
      '@selectors': {
        ':not([disabled]):hover': {
          backgroundColor: color.core.neutral[6],
          borderColor: color.core.neutral[6]
        }
      }
    },
    button_inverted: {
      color: color.core.neutral[5],
      backgroundColor: color.accent.bg,
      '@selectors': {
        ':not([disabled]):hover': {
          color: color.core.neutral[6],
          backgroundColor: color.accent.bgHover
        }
      }
    }
  };
});
/** A button to use for secondary actions. Supports all the same props as `Button`. */

export default function MutedButton(_ref2) {
  let {
    children
  } = _ref2,
      props = _objectWithoutPropertiesLoose(_ref2, ["children"]);

  return React.createElement(StyledMutedButton, props, children);
}
