// @airbnb/lunar/esm/components/Switch/index.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import uuid from 'uuid/v4';
import BaseSwitch from '../private/BaseSwitch';
import FormField, { partitionFieldProps } from '../FormField';

/** A controlled switch (fancy checkbox) field. */
export default class Switch extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      id: uuid()
    });
  }

  render() {
    const {
      fieldProps,
      inputProps
    } = partitionFieldProps(this.props);
    const {
      id
    } = this.state;
    return React.createElement(FormField, _extends({}, fieldProps, {
      inline: true,
      stretchLabel: true,
      id: id
    }), React.createElement(BaseSwitch, _extends({
      value: "1"
    }, inputProps, {
      id: id
    })));
  }

}

_defineProperty(Switch, "defaultProps", {
  checked: false
});
