// @airbnb/lunar/lib/components/MutedButton/index.js
"use strict";

exports.__esModule = true;
exports.default = MutedButton;
exports.StyledMutedButton = void 0;

var _react = _interopRequireDefault(require("react"));

var _Button = _interopRequireDefault(require("../Button"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

const StyledMutedButton = _Button.default.extendStyles((_ref) => {
  let {
    color
  } = _ref;
  return {
    button: {
      backgroundColor: color.core.neutral[5],
      border: "2px solid " + color.core.neutral[5],
      '@selectors': {
        ':not([disabled]):hover': {
          backgroundColor: color.core.neutral[6],
          borderColor: color.core.neutral[6]
        }
      }
    },
    button_inverted: {
      color: color.core.neutral[5],
      backgroundColor: color.accent.bg,
      '@selectors': {
        ':not([disabled]):hover': {
          color: color.core.neutral[6],
          backgroundColor: color.accent.bgHover
        }
      }
    }
  };
});
/** A button to use for secondary actions. Supports all the same props as `Button`. */


exports.StyledMutedButton = StyledMutedButton;

function MutedButton(_ref2) {
  let {
    children
  } = _ref2,
      props = _objectWithoutPropertiesLoose(_ref2, ["children"]);

  return _react.default.createElement(StyledMutedButton, props, children);
}
