// @airbnb/lunar/lib/components/SortCarets/index.js
"use strict";

exports.__esModule = true;
exports.default = SortCarets;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _IconCaretUp = _interopRequireDefault(require("@airbnb/lunar-icons/lib/interface/IconCaretUp"));

var _IconCaretDown = _interopRequireDefault(require("@airbnb/lunar-icons/lib/interface/IconCaretDown"));

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** Carets to indicate sorting on a table. */
function SortCarets(_ref) {
  let {
    down,
    enableDown,
    enableUp,
    up
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheet);
  return _react.default.createElement("span", {
    className: cx(styles.container, up && down && styles.container_full)
  }, up && _react.default.createElement("span", {
    className: cx(styles.caret, styles.caret_up, enableUp ? styles.caret_active : styles.caret_inactive)
  }, _react.default.createElement(_IconCaretUp.default, {
    decorative: true,
    size: "1.6em"
  })), down && _react.default.createElement("span", {
    className: cx(styles.caret, styles.caret_down, enableDown ? styles.caret_active : styles.caret_inactive)
  }, _react.default.createElement(_IconCaretDown.default, {
    decorative: true,
    size: "1.6em"
  })));
}

SortCarets.propTypes = {
  down: _propTypes.default.bool,
  enableDown: _propTypes.default.bool,
  enableUp: _propTypes.default.bool,
  up: _propTypes.default.bool
};
