// aesthetic/lib/testUtils.js
"use strict";

exports.__esModule = true;
exports.registerTestTheme = registerTestTheme;
exports.cleanupStyleElements = cleanupStyleElements;
exports.getFlushedStyles = getFlushedStyles;
exports.convertDirection = convertDirection;
exports.renderAndExpect = renderAndExpect;
exports.SYNTAX_RAW_CSS = exports.SYNTAX_VIEWPORT = exports.SYNTAX_SUPPORTS = exports.SYNTAX_PSEUDO = exports.SYNTAX_PROPERTIES = exports.SYNTAX_PAGE = exports.SYNTAX_NAMESPACE = exports.SYNTAX_MULTI_SELECTOR = exports.SYNTAX_MEDIA_QUERY_NESTED = exports.SYNTAX_MEDIA_QUERY = exports.SYNTAX_KEYFRAMES_INLINE = exports.SYNTAX_KEYFRAMES_MIXED = exports.SYNTAX_KEYFRAMES_PERCENT = exports.SYNTAX_KEYFRAMES = exports.SYNTAX_GLOBAL = exports.SYNTAX_FONT_FACES_INLINE = exports.SYNTAX_FONT_FACE_MIXED = exports.SYNTAX_FONT_FACE_MULTIPLE = exports.SYNTAX_FONT_FACE = exports.SYNTAX_FALLBACKS = exports.SYNTAX_IMPORT_MULTIPLE = exports.SYNTAX_IMPORT = exports.SYNTAX_DESCENDANT = exports.SYNTAX_CHARSET = exports.SYNTAX_ATTRIBUTE = exports.SYNTAX_UNIFIED_LOCAL_FULL = exports.SYNTAX_UNIFIED_GLOBAL_FULL = exports.SYNTAX_UNIFIED_LOCAL = exports.KEYFRAME_SLIDE_PERCENT = exports.KEYFRAME_FADE = exports.FONT_CIRCULAR_MULTIPLE_FLAT_SRC = exports.FONT_CIRCULAR_MULTIPLE = exports.FONT_ROBOTO_FLAT_SRC = exports.FONT_ROBOTO = exports.DIRECTIONS = exports.TEST_STATEMENT = exports.TEST_CLASS_NAMES = void 0;

var _rtlCssJs = _interopRequireDefault(require("rtl-css-js"));

var _aestheticUtils = require("aesthetic-utils");

var _TestAesthetic = _interopRequireDefault(require("./TestAesthetic"));

exports.TestAesthetic = _TestAesthetic.default;

var _constants = require("./constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function registerTestTheme(aesthetic) {
  aesthetic.registerTheme('default', {
    color: 'black',
    unit: 8
  }, function (_ref) {
    var unit = _ref.unit;
    return {
      '@global': {
        body: {
          padding: unit
        }
      }
    };
  });
  aesthetic.extendTheme('light', 'default', {});
  aesthetic.extendTheme('dark', 'default', {});
}

function cleanupStyleElements() {
  (0, _aestheticUtils.getStyleElements)().forEach(function (style) {
    style.remove();
  });
}

function getFlushedStyles(namespace) {
  return (0, _aestheticUtils.getFlushedStyles)((0, _aestheticUtils.getStyleElements)(namespace));
}

function convertDirection(value, dir) {
  if (dir !== 'rtl') {
    return value;
  }

  if (Array.isArray(value)) {
    return value.map(function (object) {
      return convertDirection(object, dir);
    });
  }

  if (!(0, _aestheticUtils.isObject)(value)) {
    return value;
  }

  var props = {};
  var nested = {};
  Object.keys(value).forEach(function (key) {
    var prop = value[key];

    if ((0, _aestheticUtils.isObject)(prop) || Array.isArray(prop)) {
      nested[key] = convertDirection(prop, dir);
    } else {
      props[key] = prop;
    }
  });
  return _extends({}, (0, _rtlCssJs.default)(props), {}, nested);
}

function renderAndExpect(aesthetic, styleSheet, expectedStyles, _ref2) {
  if (expectedStyles === void 0) {
    expectedStyles = {};
  }

  var dir = _ref2.dir,
      _ref2$global = _ref2.global,
      global = _ref2$global === void 0 ? false : _ref2$global;
  var name = global ? _constants.GLOBAL_STYLE_NAME : aesthetic.constructor.name.replace('Aesthetic', '').toLowerCase();
  var options = {
    name: name,
    dir: dir
  };
  var convertedSheet = global ? aesthetic.syntax.convertGlobalSheet(styleSheet, options).toObject() : aesthetic.syntax.convertStyleSheet(styleSheet, options).toObject();
  var parsedSheet = aesthetic.parseStyleSheet(convertedSheet, name);
  var className = aesthetic.transformStyles(Object.values(parsedSheet), options);
  aesthetic.flushStyles(name);
  expect(convertedSheet).toEqual(convertDirection(expectedStyles, dir));
  expect(getFlushedStyles()).toMatchSnapshot();
  expect(className).toMatchSnapshot();
}

var TEST_CLASS_NAMES = {
  footer: '.footer',
  header: '.header'
};
exports.TEST_CLASS_NAMES = TEST_CLASS_NAMES;
var TEST_STATEMENT = {
  footer: {
    color: 'blue'
  },
  header: {
    color: 'red'
  }
};
exports.TEST_STATEMENT = TEST_STATEMENT;
var DIRECTIONS = ['ltr', 'rtl'];
exports.DIRECTIONS = DIRECTIONS;
var FONT_ROBOTO = {
  fontFamily: 'Roboto',
  fontStyle: 'normal',
  fontWeight: 'normal',
  local: ['Robo'],
  srcPaths: ['fonts/Roboto.woff2', 'fonts/Roboto.ttf']
};
exports.FONT_ROBOTO = FONT_ROBOTO;
var FONT_ROBOTO_FLAT_SRC = {
  fontFamily: 'Roboto',
  fontStyle: 'normal',
  fontWeight: 'normal',
  src: "local('Robo'), url('fonts/Roboto.woff2') format('woff2'), url('fonts/Roboto.ttf') format('truetype')"
};
exports.FONT_ROBOTO_FLAT_SRC = FONT_ROBOTO_FLAT_SRC;
var FONT_CIRCULAR_MULTIPLE = [{
  fontFamily: 'Circular',
  fontStyle: 'normal',
  fontWeight: 'normal',
  srcPaths: ['fonts/Circular.woff2']
}, {
  fontFamily: 'Circular',
  fontStyle: 'italic',
  fontWeight: 'normal',
  srcPaths: ['fonts/Circular-Italic.woff2']
}, {
  fontFamily: 'Circular',
  fontStyle: 'normal',
  fontWeight: 300,
  srcPaths: ['fonts/Circular-Light.woff2']
}, {
  fontFamily: 'Circular',
  fontStyle: 'normal',
  fontWeight: 700,
  srcPaths: ['fonts/Circular-Bold.woff2']
}];
exports.FONT_CIRCULAR_MULTIPLE = FONT_CIRCULAR_MULTIPLE;
var FONT_CIRCULAR_MULTIPLE_FLAT_SRC = [{
  fontFamily: 'Circular',
  fontStyle: 'normal',
  fontWeight: 'normal',
  src: "url('fonts/Circular.woff2') format('woff2')"
}, {
  fontFamily: 'Circular',
  fontStyle: 'italic',
  fontWeight: 'normal',
  src: "url('fonts/Circular-Italic.woff2') format('woff2')"
}, {
  fontFamily: 'Circular',
  fontStyle: 'normal',
  fontWeight: 300,
  src: "url('fonts/Circular-Light.woff2') format('woff2')"
}, {
  fontFamily: 'Circular',
  fontStyle: 'normal',
  fontWeight: 700,
  src: "url('fonts/Circular-Bold.woff2') format('woff2')"
}];
exports.FONT_CIRCULAR_MULTIPLE_FLAT_SRC = FONT_CIRCULAR_MULTIPLE_FLAT_SRC;
var KEYFRAME_FADE = {
  from: {
    opacity: 0
  },
  to: {
    opacity: 1
  }
};
exports.KEYFRAME_FADE = KEYFRAME_FADE;
var KEYFRAME_SLIDE_PERCENT = {
  '0%': {
    left: '0%'
  },
  '50%': {
    left: '50%'
  },
  '100%': {
    left: '100%'
  }
};
exports.KEYFRAME_SLIDE_PERCENT = KEYFRAME_SLIDE_PERCENT;
var SYNTAX_UNIFIED_LOCAL = {
  button: {
    margin: 0,
    padding: '6px 12px',
    border: '1px solid #2e6da4',
    borderRadius: 4,
    display: 'inline-block',
    cursor: 'pointer',
    fontFamily: 'Roboto',
    fontWeight: 'normal',
    lineHeight: 'normal',
    whiteSpace: 'nowrap',
    textDecoration: 'none',
    textAlign: 'left',
    backgroundColor: '#337ab7',
    verticalAlign: 'middle',
    color: 'rgba(0, 0, 0, 0)',
    animationName: 'fade',
    animationDuration: '.3s',
    ':hover': {
      backgroundColor: '#286090',
      borderColor: '#204d74'
    },
    '::before': {
      content: '"★"',
      display: 'inline-block',
      verticalAlign: 'middle',
      marginRight: 5
    }
  }
};
exports.SYNTAX_UNIFIED_LOCAL = SYNTAX_UNIFIED_LOCAL;
var SYNTAX_UNIFIED_GLOBAL_FULL = {
  '@font-face': {
    Roboto: FONT_ROBOTO
  },
  '@keyframes': {
    fade: KEYFRAME_FADE
  }
};
exports.SYNTAX_UNIFIED_GLOBAL_FULL = SYNTAX_UNIFIED_GLOBAL_FULL;
var SYNTAX_UNIFIED_LOCAL_FULL = {
  button: _extends({}, SYNTAX_UNIFIED_LOCAL.button, {
    '@media': {
      '(max-width: 600px)': {
        padding: '4px 8px'
      }
    }
  })
};
exports.SYNTAX_UNIFIED_LOCAL_FULL = SYNTAX_UNIFIED_LOCAL_FULL;
var SYNTAX_ATTRIBUTE = {
  attr: {
    display: 'block',
    '[disabled]': {
      opacity: 0.5
    }
  }
};
exports.SYNTAX_ATTRIBUTE = SYNTAX_ATTRIBUTE;
var SYNTAX_CHARSET = {
  '@charset': 'utf8'
};
exports.SYNTAX_CHARSET = SYNTAX_CHARSET;
var SYNTAX_DESCENDANT = {
  list: {
    margin: 0,
    padding: 0,
    '@selectors': {
      '> li': {
        listStyle: 'bullet'
      }
    }
  }
};
exports.SYNTAX_DESCENDANT = SYNTAX_DESCENDANT;
var SYNTAX_IMPORT = {
  '@import': 'url("./some/path.css")'
};
exports.SYNTAX_IMPORT = SYNTAX_IMPORT;
var SYNTAX_IMPORT_MULTIPLE = {
  '@import': ['url("./some/path.css")', 'url("./another/path.css")']
};
exports.SYNTAX_IMPORT_MULTIPLE = SYNTAX_IMPORT_MULTIPLE;
var SYNTAX_FALLBACKS = {
  fallback: {
    background: 'linear-gradient(...)',
    display: 'flex',
    '@fallbacks': {
      background: 'red',
      display: ['block', 'inline-block'],
      color: 'blue'
    }
  }
};
exports.SYNTAX_FALLBACKS = SYNTAX_FALLBACKS;
var SYNTAX_FONT_FACE = {
  '@font-face': {
    Roboto: [FONT_ROBOTO]
  }
};
exports.SYNTAX_FONT_FACE = SYNTAX_FONT_FACE;
var SYNTAX_FONT_FACE_MULTIPLE = {
  '@font-face': {
    Circular: FONT_CIRCULAR_MULTIPLE
  }
};
exports.SYNTAX_FONT_FACE_MULTIPLE = SYNTAX_FONT_FACE_MULTIPLE;
var SYNTAX_FONT_FACE_MIXED = {
  '@font-face': {
    Roboto: [FONT_ROBOTO],
    Circular: FONT_CIRCULAR_MULTIPLE
  }
};
exports.SYNTAX_FONT_FACE_MIXED = SYNTAX_FONT_FACE_MIXED;
var SYNTAX_FONT_FACES_INLINE = {
  single: {
    fontFamily: FONT_ROBOTO
  },
  multiple: {
    fontFamily: [].concat(FONT_CIRCULAR_MULTIPLE, ['OtherFont', FONT_ROBOTO])
  }
};
exports.SYNTAX_FONT_FACES_INLINE = SYNTAX_FONT_FACES_INLINE;
var SYNTAX_GLOBAL = {
  '@global': {
    body: {
      margin: 0
    },
    html: {
      height: '100%'
    },
    a: {
      color: 'red',
      ':hover': {
        color: 'darkred'
      },
      '@selectors': {
        ':focus': {
          color: 'lightred'
        }
      }
    },
    ul: {
      margin: 0,
      '@selectors': {
        '> li': {
          margin: 0
        }
      },
      '@media': {
        '(max-width: 500px)': {
          margin: 20
        }
      }
    }
  }
};
exports.SYNTAX_GLOBAL = SYNTAX_GLOBAL;
var SYNTAX_KEYFRAMES = {
  '@keyframes': {
    fade: KEYFRAME_FADE
  }
};
exports.SYNTAX_KEYFRAMES = SYNTAX_KEYFRAMES;
var SYNTAX_KEYFRAMES_PERCENT = {
  '@keyframes': {
    slide: KEYFRAME_SLIDE_PERCENT
  }
};
exports.SYNTAX_KEYFRAMES_PERCENT = SYNTAX_KEYFRAMES_PERCENT;
var SYNTAX_KEYFRAMES_MIXED = {
  '@keyframes': {
    fade: KEYFRAME_FADE,
    slide: KEYFRAME_SLIDE_PERCENT
  }
};
exports.SYNTAX_KEYFRAMES_MIXED = SYNTAX_KEYFRAMES_MIXED;
var SYNTAX_KEYFRAMES_INLINE = {
  single: {
    animationName: _extends({}, KEYFRAME_SLIDE_PERCENT, {
      name: 'slide'
    })
  },
  multiple: {
    animationName: [_extends({}, KEYFRAME_SLIDE_PERCENT, {
      name: 'slide'
    }), 'unknown', KEYFRAME_FADE]
  }
};
exports.SYNTAX_KEYFRAMES_INLINE = SYNTAX_KEYFRAMES_INLINE;
var SYNTAX_MEDIA_QUERY = {
  media: {
    color: 'red',
    paddingLeft: 10,
    '@media': {
      '(min-width: 300px)': {
        color: 'blue',
        paddingLeft: 15
      },
      '(max-width: 1000px)': {
        color: 'green',
        paddingLeft: 20
      }
    }
  }
};
exports.SYNTAX_MEDIA_QUERY = SYNTAX_MEDIA_QUERY;
var SYNTAX_MEDIA_QUERY_NESTED = {
  media: {
    color: 'red',
    '@media': {
      '(min-width: 300px)': {
        color: 'blue',
        '@media': {
          '(max-width: 1000px)': {
            color: 'green'
          }
        }
      }
    }
  }
};
exports.SYNTAX_MEDIA_QUERY_NESTED = SYNTAX_MEDIA_QUERY_NESTED;
var SYNTAX_MULTI_SELECTOR = {
  multi: {
    cursor: 'pointer',
    '@selectors': {
      ':disabled, [disabled], > span': {
        cursor: 'default'
      }
    }
  }
};
exports.SYNTAX_MULTI_SELECTOR = SYNTAX_MULTI_SELECTOR;
var SYNTAX_NAMESPACE = {
  '@namespace': 'url(http://www.w3.org/1999/xhtml)'
};
exports.SYNTAX_NAMESPACE = SYNTAX_NAMESPACE;
var SYNTAX_PAGE = {
  '@page': {
    margin: '1cm',
    ':left': {
      margin: 0
    }
  }
};
exports.SYNTAX_PAGE = SYNTAX_PAGE;
var SYNTAX_PROPERTIES = {
  props: {
    color: 'black',
    display: 'inline',
    marginRight: 10,
    padding: 0
  }
};
exports.SYNTAX_PROPERTIES = SYNTAX_PROPERTIES;
var SYNTAX_PSEUDO = {
  pseudo: {
    position: 'fixed',
    ':hover': {
      position: 'static'
    },
    '::before': {
      position: 'absolute'
    }
  }
};
exports.SYNTAX_PSEUDO = SYNTAX_PSEUDO;
var SYNTAX_SUPPORTS = {
  sup: {
    display: 'block',
    '@supports': {
      '(display: flex)': {
        display: 'flex'
      },
      'not (display: flex)': {
        float: 'left'
      }
    }
  }
};
exports.SYNTAX_SUPPORTS = SYNTAX_SUPPORTS;
var SYNTAX_VIEWPORT = {
  '@viewport': {
    width: 'device-width',
    orientation: 'landscape'
  }
};
exports.SYNTAX_VIEWPORT = SYNTAX_VIEWPORT;
var SYNTAX_RAW_CSS = {
  button: "\n    display: 'block';\n    font-size: 16px;\n    text-align: 'left';\n\n    &:hover {\n      color: 'red';\n    }\n\n    & {\n      vertical-align: 'middle';\n    }\n\n    @media (max-width: 600px) {\n      display: 'none';\n    }\n\n    @supports (display: flex) {\n      display: 'flex';\n    }\n  "
};
exports.SYNTAX_RAW_CSS = SYNTAX_RAW_CSS;
