// @airbnb/lunar/lib/components/Price/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("../../constants");

var _formatPrice = _interopRequireDefault(require("../../utils/formatPrice"));

var _Empty = _interopRequireDefault(require("../Empty"));

var _ = _interopRequireDefault(require("../.."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Display a formatted and localized currency or price amount. */
class Price extends _react.default.PureComponent {
  render() {
    const {
      amount: baseAmount,
      currency: baseCurrency,
      micros: baseMicros,
      display,
      locale,
      precision,
      round,
      trimTrailingZeros
    } = this.props;
    let amount = 0;
    let currency = baseCurrency;
    let micros = baseMicros;

    if (baseAmount === undefined || baseAmount === null) {
      return _react.default.createElement(_Empty.default, null);
    }

    if (typeof baseAmount === 'object') {
      currency = baseAmount.currency;
      micros = baseAmount.is_micros_accuracy;
      amount = micros ? baseAmount.amount_micros : baseAmount.amount;
    } else if (typeof baseAmount === 'number') {
      amount = baseAmount;
    }

    if (micros) {
      amount /= _constants.MICROS;
    }

    return _react.default.createElement("span", null, (0, _formatPrice.default)(amount, currency, {
      display,
      precision,
      round,
      trimTrailingZeros,
      locale: locale || _.default.locale()
    }));
  }

}

exports.default = Price;

_defineProperty(Price, "propTypes", {
  currency: _propTypes.default.any,
  trimTrailingZeros: _propTypes.default.bool,
  round: _propTypes.default.bool,
  precision: _propTypes.default.number,
  micros: _propTypes.default.bool,
  locale: _propTypes.default.any,
  amount: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.number]),
  display: _propTypes.default.oneOf(['symbol', 'code', 'name'])
});

_defineProperty(Price, "defaultProps", {
  currency: _constants.DEFAULT_CURRENCY,
  display: 'symbol',
  locale: undefined,
  micros: false,
  precision: 0,
  round: false,
  trimTrailingZeros: false
});
