// @airbnb/lunar/esm/components/ResponsiveImage/index.js
import _pt from "prop-types";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import { mutuallyExclusiveTrueProps, forbidExtraProps } from 'airbnb-prop-types';
import withStyles from '../../composers/withStyles';
import { styleSheet } from './styles';
export const DEFAULT_BORDER_RADIUS = 6;
const objectFitPropType = mutuallyExclusiveTrueProps('contain', 'cover');

/** An image that is constrained proportionally in one or both dimensions. */
export class ResponsiveImage extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "image", void 0);

    _defineProperty(this, "state", {
      imageLoaded: false
    });
  }

  componentDidMount() {
    this.createAsyncImage();
  }

  componentDidUpdate(_ref) {
    let {
      src: previousSrc
    } = _ref;
    const {
      src: currentSrc
    } = this.props;

    if (currentSrc !== previousSrc) {
      this.removeImage();
      this.setState({
        imageLoaded: false
      });
      this.createAsyncImage();
    }
  }

  componentWillUnmount() {
    this.removeImage();
  }

  createAsyncImage() {
    this.image = new Image();
    this.image.addEventListener('load', () => {
      this.handleAsyncImageLoad();
    });
    this.image.addEventListener('error', () => {
      this.handleAsyncImageLoad();
    });
    this.image.src = this.props.src;
  }

  removeImage() {
    if (this.image) {
      // Prevent these callbacks from being fired if they haven't fired yet.
      delete this.image.onload;
      delete this.image.onerror;
    }
  }

  handleAsyncImageLoad() {
    // We no longer need the Image object, so let's remove it so it can be
    // garbage collected.
    delete this.image;
    this.setState({
      imageLoaded: true
    });
  }

  render() {
    const {
      cx,
      alt,
      contain,
      cover,
      noShadow,
      maxWidth,
      maxHeight,
      src,
      borderRadius,
      shimmer,
      styles
    } = this.props;
    const {
      imageLoaded
    } = this.state;

    if (!imageLoaded && shimmer) {
      return shimmer;
    }

    return React.createElement("img", {
      className: cx(styles.image, contain && styles.image_contain, cover && styles.image_cover, noShadow && styles.image_noShadow, {
        borderRadius,
        maxWidth,
        maxHeight
      }),
      src: src,
      width: "100%",
      height: "auto",
      alt: alt
    });
  }

}

_defineProperty(ResponsiveImage, "defaultProps", {
  borderRadius: DEFAULT_BORDER_RADIUS,
  contain: false,
  cover: false,
  maxHeight: 'none',
  maxWidth: 'none',
  noShadow: false,
  shimmer: null
});

_defineProperty(ResponsiveImage, "propTypes", {
  shimmer: _pt.node,
  src: _pt.string.isRequired,
  maxHeight: _pt.oneOfType([_pt.string, _pt.number]),
  maxWidth: _pt.oneOfType([_pt.string, _pt.number]),
  noShadow: _pt.bool,
  borderRadius: _pt.number,
  alt: _pt.string.isRequired,
  contain: objectFitPropType,
  cover: objectFitPropType
});

export default withStyles(styleSheet)(ResponsiveImage);
