// interweave-autolink/esm/EmailMatcher.js
function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

import React from 'react';
import { Matcher } from 'interweave';
import Email from './Email';
import { EMAIL_PATTERN } from './constants';

var EmailMatcher = function (_Matcher) {
  _inheritsLoose(EmailMatcher, _Matcher);

  function EmailMatcher() {
    return _Matcher.apply(this, arguments) || this;
  }

  var _proto = EmailMatcher.prototype;

  _proto.replaceWith = function replaceWith(children, props) {
    return React.createElement(Email, props, children);
  };

  _proto.asTag = function asTag() {
    return 'a';
  };

  _proto.match = function match(string) {
    return this.doMatch(string, EMAIL_PATTERN, function (matches) {
      return {
        email: matches[0],
        emailParts: {
          host: matches[2],
          username: matches[1]
        }
      };
    });
  };

  return EmailMatcher;
}(Matcher);

export { EmailMatcher as default };
