// @airbnb/lunar/esm/components/Lightbox/Image.js
import _pt from "prop-types";
import React from 'react';
import useStyles from '../../hooks/useStyles';
import ResponsiveImage from '../ResponsiveImage';
import Loader from '../Loader';
import ImageViewer from '../ImageViewer';
import { styleSheetImage as styleSheet } from './styles';
export default function Image(_ref) {
  let {
    alt,
    aside,
    hideAside,
    src,
    showZoomControls,
    showRotateControls,
    scale,
    rotation
  } = _ref;
  const [styles, cx] = useStyles(styleSheet);
  return React.createElement("div", {
    className: cx(styles.wrapper)
  }, React.createElement("div", {
    className: cx(styles.layout)
  }, React.createElement("figure", {
    className: cx(styles.figure)
  }, showZoomControls || showRotateControls ? React.createElement(ImageViewer, {
    alt: alt,
    src: src,
    scale: scale,
    rotation: rotation,
    width: "100%",
    height: "100%"
  }) : React.createElement(ResponsiveImage, {
    contain: true,
    noShadow: true,
    alt: alt,
    borderRadius: 0,
    src: src,
    maxWidth: "100%",
    maxHeight: "100%",
    shimmer: React.createElement(Loader, null)
  })), aside && !hideAside && React.createElement("aside", {
    className: cx(styles.aside)
  }, aside)));
}
Image.propTypes = {
  alt: _pt.string.isRequired,
  aside: _pt.node,
  hideAside: _pt.bool,
  rotation: _pt.number,
  scale: _pt.number,
  showRotateControls: _pt.bool,
  showZoomControls: _pt.bool,
  src: _pt.string.isRequired
};
