// @airbnb/lunar/esm/components/HierarchyPicker/Hierarchy/index.js
import _pt from "prop-types";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import debounce from 'lodash/debounce';
import HierarchyList from './HierarchyList';
import { readonlyReducer } from './helpers';
export default class Hierarchy extends React.Component {
  constructor() {
    var _this;

    super(...arguments);
    _this = this;

    _defineProperty(this, "state", {
      focusDef: [],
      filteredItems: []
    });

    _defineProperty(this, "setFocusDebounced", debounce(this.setFocus, 100));

    _defineProperty(this, "handleSubtree", function (focusDef, callback, immediate) {
      if (immediate === void 0) {
        immediate = false;
      }

      if (immediate) {
        _this.setFocus(focusDef, callback);
      } else {
        _this.setFocusDebounced(focusDef, callback);
      }
    });

    _defineProperty(this, "handleItemPicked", chosen => {
      this.props.onItemPicked(chosen, {
        origin: 'Hierarchy'
      });
    });
  }

  componentDidMount() {
    this.filterItems();
  }

  componentDidUpdate(prevProps) {
    if (this.props.items !== prevProps.items) {
      this.filterItems();
    }
  }

  filterItems() {
    this.setState({
      filteredItems: (this.props.items || []).reduce(readonlyReducer, [])
    });
  }

  setFocus(focusDef, callback) {
    this.setState({
      focusDef
    }, callback);
  }

  render() {
    const {
      filteredItems,
      focusDef
    } = this.state;
    const {
      chosen,
      renderItem,
      hierarchyMaxHeight,
      hierarchyWidth,
      verticallyAlign
    } = this.props;
    return React.createElement(HierarchyList, {
      chosen: chosen,
      focus: focusDef,
      items: filteredItems,
      maxHeight: hierarchyMaxHeight,
      renderItem: renderItem,
      width: hierarchyWidth,
      verticallyAlign: verticallyAlign,
      onItemPicked: this.handleItemPicked,
      onSubtree: this.handleSubtree
    });
  }

}

_defineProperty(Hierarchy, "propTypes", {
  chosen: _pt.any,
  formatter: _pt.any.isRequired,
  hierarchyMaxHeight: _pt.number,
  hierarchyWidth: _pt.number,
  items: _pt.arrayOf(_pt.any),
  onItemPicked: _pt.func.isRequired,
  renderItem: _pt.any,
  verticallyAlign: _pt.bool
});

_defineProperty(Hierarchy, "defaultProps", {
  chosen: [],
  items: []
});
