// @airbnb/lunar/lib/components/PriceComparison/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("../../constants");

var _PriceGroup = _interopRequireDefault(require("../PriceGroup"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Display a comparison between USD and native amounts. Supports all props in `Price`. */
class PriceComparison extends _react.default.PureComponent {
  render() {
    const _this$props = this.props,
          {
      amount,
      amountUSD,
      currency
    } = _this$props,
          restProps = _objectWithoutPropertiesLoose(_this$props, ["amount", "amountUSD", "currency"]);

    const amounts = {};
    const hasPrice = amount !== null && amount !== undefined;
    const hasPriceUSD = amountUSD !== null && amountUSD !== undefined;
    const targetCurrency = amount && typeof amount === 'object' ? amount.currency : currency; // Sometimes the native price could be empty

    if (hasPrice) {
      amounts[targetCurrency] = amount;
    } // Show the USD price if the native price is empty, or if the currencies don't match


    if (hasPriceUSD && (!hasPrice || targetCurrency !== _constants.DEFAULT_CURRENCY)) {
      amounts[_constants.DEFAULT_CURRENCY] = amountUSD;
    }

    return _react.default.createElement(_PriceGroup.default, _extends({}, restProps, {
      amounts: amounts
    }));
  }

}

exports.default = PriceComparison;

_defineProperty(PriceComparison, "propTypes", {
  currency: _propTypes.default.string,
  amount: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.number]),
  amountUSD: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.number])
});

_defineProperty(PriceComparison, "defaultProps", {
  amount: null,
  amountUSD: null,
  currency: _constants.DEFAULT_CURRENCY
});
