// @vx/axis/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Axis", {
  enumerable: true,
  get: function get() {
    return _Axis.default;
  }
});
Object.defineProperty(exports, "AxisLeft", {
  enumerable: true,
  get: function get() {
    return _AxisLeft.default;
  }
});
Object.defineProperty(exports, "AxisRight", {
  enumerable: true,
  get: function get() {
    return _AxisRight.default;
  }
});
Object.defineProperty(exports, "AxisTop", {
  enumerable: true,
  get: function get() {
    return _AxisTop.default;
  }
});
Object.defineProperty(exports, "AxisBottom", {
  enumerable: true,
  get: function get() {
    return _AxisBottom.default;
  }
});
Object.defineProperty(exports, "Orientation", {
  enumerable: true,
  get: function get() {
    return _orientation.default;
  }
});

var _Axis = _interopRequireDefault(require("./axis/Axis"));

var _AxisLeft = _interopRequireDefault(require("./axis/AxisLeft"));

var _AxisRight = _interopRequireDefault(require("./axis/AxisRight"));

var _AxisTop = _interopRequireDefault(require("./axis/AxisTop"));

var _AxisBottom = _interopRequireDefault(require("./axis/AxisBottom"));

var _orientation = _interopRequireDefault(require("./constants/orientation"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
