// @airbnb/lunar/esm/components/BrowserNotification/index.js
import _pt from "prop-types";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';

/** Display a browser notification through a declarative component. */
export default class BrowserNotification extends React.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "closeTimeout", void 0);

    _defineProperty(this, "notification", null);
  }

  componentDidMount() {
    this.showNotification();
  }

  componentDidUpdate() {
    this.showNotification();
  }

  componentWillUnmount() {
    if (this.closeTimeout) {
      window.clearTimeout(this.closeTimeout);
    }

    if (this.notification) {
      this.notification.close();
    }
  }

  showNotification() {
    Notification.requestPermission().then(permission => {
      if (permission === 'granted') {
        const {
          title,
          tag,
          body,
          icon,
          timeout,
          onClick
        } = this.props;
        this.notification = new Notification(title, {
          tag,
          body,
          icon
        });

        if (onClick) {
          this.notification.addEventListener('click', onClick);
        }

        if (timeout) {
          this.closeTimeout = window.setTimeout(() => {
            this.closeTimeout = 0;

            if (this.notification) {
              this.notification.close();
              this.notification = null;
            }
          }, timeout);
        }
      }
      /* istanbul ignore next */


      if ("production" !== process.env.NODE_ENV && permission !== 'granted') {
        // eslint-disable-next-line no-console
        console.warn('Attempted to show a notification, but was denied permission.');
      }
    });
  }

  render() {
    return null;
  }

}

_defineProperty(BrowserNotification, "propTypes", {
  title: _pt.string.isRequired,
  tag: _pt.string,
  body: _pt.string,
  icon: _pt.string,
  timeout: _pt.number,
  onClick: _pt.func
});

_defineProperty(BrowserNotification, "defaultProps", {
  onClick: null,
  timeout: null
});
