// @airbnb/lunar/lib/components/Spacing/index.js
"use strict";

exports.__esModule = true;
exports.default = Spacing;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** Common component for arbitray layout structure and spacing. */
function Spacing(_ref) {
  let {
    all = 0,
    bottom = 0,
    children,
    horizontal = 0,
    inline,
    inner,
    left = 0,
    right = 0,
    tag: TagProp = 'div',
    top = 0,
    vertical = 0
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheet);
  const type = inner ? 'inner' : 'outer';
  const classes = [];

  if (all) {
    const cleaned = (0, _styles.cleanClassName)(all);
    classes.push(styles[type + "Top_" + cleaned], styles[type + "Right_" + cleaned], styles[type + "Bottom_" + cleaned], styles[type + "Left_" + cleaned]);
  } else {
    if (vertical) {
      const cleaned = (0, _styles.cleanClassName)(vertical);
      classes.push(styles[type + "Top_" + cleaned], styles[type + "Bottom_" + cleaned]);
    } else {
      if (top) {
        const cleaned = (0, _styles.cleanClassName)(top);
        classes.push(styles[type + "Top_" + cleaned]);
      }

      if (bottom) {
        const cleaned = (0, _styles.cleanClassName)(bottom);
        classes.push(styles[type + "Bottom_" + cleaned]);
      }
    }

    if (horizontal) {
      const cleaned = (0, _styles.cleanClassName)(horizontal);
      classes.push(styles[type + "Left_" + cleaned], styles[type + "Right_" + cleaned]);
    } else {
      if (left) {
        const cleaned = (0, _styles.cleanClassName)(left);
        classes.push(styles[type + "Left_" + cleaned]);
      }

      if (right) {
        const cleaned = (0, _styles.cleanClassName)(right);
        classes.push(styles[type + "Right_" + cleaned]);
      }
    }
  }

  return _react.default.createElement(TagProp, {
    className: cx(...classes, inline && styles.spacing_inline)
  }, children);
}

Spacing.propTypes = {
  all: _propTypes.default.oneOf([0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5, 5.5, 6, 6.5, 7, 7.5, 8, 8.5, 9, 9.5, 10, 10.5, 11, 11.5, 12]),
  bottom: _propTypes.default.oneOf([0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5, 5.5, 6, 6.5, 7, 7.5, 8, 8.5, 9, 9.5, 10, 10.5, 11, 11.5, 12]),
  children: _propTypes.default.any.isRequired,
  horizontal: _propTypes.default.oneOf([0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5, 5.5, 6, 6.5, 7, 7.5, 8, 8.5, 9, 9.5, 10, 10.5, 11, 11.5, 12]),
  inline: _propTypes.default.bool,
  inner: _propTypes.default.bool,
  left: _propTypes.default.oneOf([0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5, 5.5, 6, 6.5, 7, 7.5, 8, 8.5, 9, 9.5, 10, 10.5, 11, 11.5, 12]),
  right: _propTypes.default.oneOf([0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5, 5.5, 6, 6.5, 7, 7.5, 8, 8.5, 9, 9.5, 10, 10.5, 11, 11.5, 12]),
  tag: _propTypes.default.oneOf(['article', 'div', 'footer', 'header', 'section']),
  top: _propTypes.default.oneOf([0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5, 5.5, 6, 6.5, 7, 7.5, 8, 8.5, 9, 9.5, 10, 10.5, 11, 11.5, 12]),
  vertical: _propTypes.default.oneOf([0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5, 5.5, 6, 6.5, 7, 7.5, 8, 8.5, 9, 9.5, 10, 10.5, 11, 11.5, 12])
};
