// @airbnb/lunar/lib/themes/global.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _default = fontFaces => (_ref) => {
  let {
    color,
    font
  } = _ref;
  return {
    '@global': {
      '*': {
        boxSizing: 'border-box',
        '@selectors': {
          ':focus': {
            outlineColor: color.core.primary[3]
          },
          ':focus:not(:focus-visible)': {
            outline: 'none'
          }
        }
      },
      'html, body': {
        height: '100%',
        margin: 0,
        padding: 0
      },
      body: {
        color: color.accent.text,
        backgroundColor: color.base,
        fontFamily: font.fontFamily,
        fontSize: font.textRegular.fontSize,
        lineHeight: font.textRegular.lineHeight,
        '-ms-text-size-adjust': '100%',
        '-webkit-text-size-adjust': '100%',
        '-webkit-font-smoothing': 'antialiased',
        '-moz-osx-font-smoothing': 'grayscale'
      },
      'a, button, input, select, textarea': {
        color: 'inherit',
        backgroundColor: 'inherit',
        font: 'inherit'
      },
      img: {
        display: 'inline-block',
        verticalAlign: 'middle'
      },
      'tracking-boundary': {
        display: 'block'
      }
    },
    '@font-face': fontFaces
  };
};

exports.default = _default;
