// @airbnb/lunar/esm/components/RadioButtonController/index.js
import _pt from "prop-types";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import uuid from 'uuid/v4';
import proxyComponent from '../../utils/proxyComponent';
import FormField, { partitionFieldProps } from '../FormField';
import RadioButton from '../RadioButton';

/** Manage multiple radio buttons with the same input `name`. */
export default class RadioButtonController extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      id: uuid(),
      value: this.props.value || ''
    });

    _defineProperty(this, "handleChange", (checked, value, event) => {
      if (value !== this.state.value) {
        this.props.onChange(value, event);
        this.setState({
          value
        });
      }
    });

    _defineProperty(this, "renderRadioButton", proxyComponent(RadioButton, (_ref) => {
      let {
        value
      } = _ref,
          props = _objectWithoutPropertiesLoose(_ref, ["value"]);

      const {
        inputProps
      } = partitionFieldProps(this.props);
      const {
        id,
        value: currentValue
      } = this.state;
      return React.createElement(RadioButton, _extends({
        compactSpacing: true
      }, props, inputProps, {
        hideOptionalLabel: true,
        id: id + "-" + value,
        value: value,
        checked: value === currentValue,
        onChange: this.handleChange
      }));
    }));
  }

  componentDidUpdate(prevProps) {
    if (this.props.value !== prevProps.value) {
      this.setState({
        value: this.props.value || ''
      });
    }
  }

  render() {
    const {
      children,
      fieldProps
    } = partitionFieldProps(this.props);
    const {
      id,
      value
    } = this.state;
    return React.createElement(FormField, _extends({}, fieldProps, {
      id: id
    }), children(this.renderRadioButton, value, id));
  }

}

_defineProperty(RadioButtonController, "propTypes", {
  children: _pt.func.isRequired,
  name: _pt.string.isRequired,
  onChange: _pt.func.isRequired,
  value: _pt.string
});

_defineProperty(RadioButtonController, "defaultProps", {
  value: ''
});
