// @airbnb/lunar/lib/components/Multicomplete/private/Chip.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _IconCloseAlt = _interopRequireDefault(require("@airbnb/lunar-icons/lib/interface/IconCloseAlt"));

var _Chip = _interopRequireDefault(require("../../Chip"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class MulticompleteChip extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleClick", event => {
      this.props.onClick(this.props.value, event);
    });
  }

  render() {
    const {
      value
    } = this.props;
    return _react.default.createElement(_Chip.default, {
      afterIcon: _react.default.createElement(_IconCloseAlt.default, {
        decorative: true,
        size: "2em"
      }),
      onIconClick: this.handleClick
    }, value);
  }

}

exports.default = MulticompleteChip;

_defineProperty(MulticompleteChip, "propTypes", {
  value: _propTypes.default.string.isRequired,
  onClick: _propTypes.default.func.isRequired
});
