// @airbnb/lunar/lib/components/HierarchyPicker/Search/SearchResult.js
"use strict";

exports.__esModule = true;
exports.default = SearchResult;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _groupBy = _interopRequireDefault(require("lodash/groupBy"));

var _useStyles = _interopRequireDefault(require("../../../hooks/useStyles"));

var _Text = _interopRequireDefault(require("../../Text"));

var _Highlight = _interopRequireDefault(require("./Highlight"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function SearchResult(_ref) {
  let {
    item,
    formattedParents,
    matches = [],
    query = ''
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheetSearchResult);
  const {
    description,
    label,
    name
  } = item;
  const mbk = (0, _groupBy.default)(matches, 'key');
  const [labelMatch = null] = mbk.label || [];
  const [descMatch = null] = mbk.description || [];
  const [keywMatch = null] = mbk.keywords || [];
  const [longest] = query.split(/\s{1,}/).sort((a, b) => b.length - a.length);
  return _react.default.createElement("div", {
    className: cx(styles.resultItem)
  }, _react.default.createElement(_Text.default, {
    bold: true
  }, formattedParents, _react.default.createElement(_Highlight.default, {
    word: longest,
    match: labelMatch,
    fallback: label || name
  })), description && _react.default.createElement(_Text.default, null, _react.default.createElement(_Highlight.default, {
    word: longest,
    match: descMatch,
    fallback: description
  })), keywMatch && _react.default.createElement(_Text.default, null, _react.default.createElement(_Highlight.default, {
    word: longest,
    match: keywMatch
  })));
}

SearchResult.propTypes = {
  item: _propTypes.default.any.isRequired,
  formattedParents: _propTypes.default.string.isRequired,
  matches: _propTypes.default.arrayOf(_propTypes.default.any),
  query: _propTypes.default.string
};
