// @airbnb/lunar/lib/components/HierarchyPicker/Search/styles.js
"use strict";

exports.__esModule = true;
exports.styleSheetSearchResult = exports.styleSheetHighlight = exports.styleSheet = void 0;

const styleSheet = (_ref) => {
  let {
    unit
  } = _ref;
  return {
    container: {
      padding: unit
    }
  };
};

exports.styleSheet = styleSheet;

const styleSheetHighlight = (_ref2) => {
  let {
    color,
    ui
  } = _ref2;
  return {
    highlight: {
      borderRadius: ui.borderRadius,
      backgroundColor: color.core.warning[0],
      '@selectors': {
        '> mark': {
          backgroundColor: 'transparent',
          color: 'inherit',
          position: 'relative',
          whiteSpace: 'nowrap'
        }
      }
    },
    highlight_dark: {
      backgroundColor: color.core.warning[3]
    }
  };
};

exports.styleSheetHighlight = styleSheetHighlight;

const styleSheetSearchResult = (_ref3) => {
  let {
    unit
  } = _ref3;
  return {
    resultItem: {
      padding: unit,
      wordBreak: 'break-word'
    }
  };
};

exports.styleSheetSearchResult = styleSheetSearchResult;
