// @airbnb/lunar/lib/components/GradientScroller/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.GradientScroller = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _debounce = _interopRequireDefault(require("lodash/debounce"));

var _throttle = _interopRequireDefault(require("lodash/throttle"));

var _IconChevronLeft = _interopRequireDefault(require("@airbnb/lunar-icons/lib/interface/IconChevronLeft"));

var _IconChevronRight = _interopRequireDefault(require("@airbnb/lunar-icons/lib/interface/IconChevronRight"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _DirectionalIcon = _interopRequireDefault(require("../DirectionalIcon"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const INCREMENT = 25;

/** A horizontal scroller with gradients on each side. */
class GradientScroller extends _react.default.Component {
  constructor(props) {
    super(props); // Register resize observer before mounting

    _defineProperty(this, "contentsRef", null);

    _defineProperty(this, "observer", void 0);

    _defineProperty(this, "scrollInterval", 0);

    _defineProperty(this, "scrollerRef", null);

    _defineProperty(this, "state", {
      showStartGradient: false,
      showEndGradient: true
    });

    _defineProperty(this, "handleContentsRef", ref => {
      if (ref) {
        this.observer.observe(ref);
      } else if (this.contentsRef) {
        this.observer.unobserve(this.contentsRef);
      }

      this.contentsRef = ref;
    });

    _defineProperty(this, "handleScrollerRef", ref => {
      if (ref) {
        this.observer.observe(ref);
      } else if (this.scrollerRef) {
        this.observer.unobserve(this.scrollerRef);
      }

      this.scrollerRef = ref;
    });

    _defineProperty(this, "handleObserver", entries => {
      let calculate = false;
      entries.forEach((_ref) => {
        let {
          contentRect,
          target
        } = _ref;
        const width = target.dataset.prevWidth;
        const nextWidth = String(contentRect.width); // Dataset stores as strings

        if (!width || nextWidth !== width) {
          calculate = true; // eslint-disable-next-line no-param-reassign

          target.dataset.prevWidth = nextWidth;
        }
      });

      if (calculate) {
        this.handleResizeDebounced();
      }
    });

    _defineProperty(this, "handleResize", () => {
      this.calculate();
    });

    _defineProperty(this, "handleResizeDebounced", (0, _debounce.default)(this.handleResize, 150));

    _defineProperty(this, "handleScroll", () => {
      const target = this.scrollerRef;
      const {
        showStartGradient,
        showEndGradient
      } = this.state;

      if (!target) {
        return;
      }

      if (target.scrollLeft > 0 && !showStartGradient) {
        this.setState({
          showStartGradient: true
        });
      } else if (target.scrollLeft === 0 && showStartGradient) {
        this.setState({
          showStartGradient: false
        });
      } // Get the largest width child to calculate against


      const scrolledWidth = target.scrollLeft + target.clientWidth;
      const maxChildWidth = this.calculateMaxChildWidth(target.children[0]);

      if (scrolledWidth < maxChildWidth && !showEndGradient) {
        this.setState({
          showEndGradient: true
        });
      } else if (scrolledWidth >= maxChildWidth && showEndGradient) {
        this.setState({
          showEndGradient: false
        });
      }
    });

    _defineProperty(this, "handleScrollThrottled", (0, _throttle.default)(this.handleScroll, 100));

    _defineProperty(this, "handleScrollLeft", event => {
      this.doScroll(-this.calculateAutoScrollIncrement(), event.type !== 'click');
    });

    _defineProperty(this, "handleScrollRight", event => {
      this.doScroll(this.calculateAutoScrollIncrement(), event.type !== 'click');
    });

    _defineProperty(this, "handleScrollStop", () => {
      if (this.scrollInterval) {
        window.clearInterval(this.scrollInterval);
      }
    });

    this.observer = new window.ResizeObserver(this.handleObserver);
  }

  componentWillUnmount() {
    this.observer.disconnect();
  }

  calculate() {
    const scroller = this.scrollerRef;
    const contents = this.contentsRef;

    if (!scroller || !contents) {
      return;
    }

    const scrollerWidth = scroller.clientWidth;
    const maxChildWidth = this.calculateMaxChildWidth(contents);
    this.setState({
      showStartGradient: scroller.scrollLeft > 0,
      showEndGradient: maxChildWidth > scrollerWidth
    });
  }

  calculateAutoScrollIncrement() {
    let increment = this.props.autoScrollIncrement; // 15% of the width

    if (!increment && this.scrollerRef) {
      increment = this.scrollerRef.offsetWidth * 0.15;
    }

    if (!increment) {
      increment = INCREMENT;
    }

    return increment;
  }

  calculateMaxChildWidth(contents) {
    return Array.from(contents.children).reduce((width, child) => Math.max(child.clientWidth, width), contents.clientWidth);
  }

  doScroll(amount, interval) {
    if (interval === void 0) {
      interval = true;
    }

    // eslint-disable-next-line unicorn/consistent-function-scoping
    const callback = () => {
      if (this.scrollerRef) {
        this.scrollerRef.scrollLeft += amount;
        this.handleScrollThrottled();
      }
    };

    this.handleScrollStop();

    if (interval) {
      this.scrollInterval = window.setInterval(callback, 100);
    } else {
      callback();
    }
  }

  render() {
    const {
      cx,
      children,
      styles,
      hideScrollbar,
      showArrows
    } = this.props;
    const {
      showStartGradient,
      showEndGradient
    } = this.state;
    return _react.default.createElement("div", {
      className: cx(styles.container)
    }, _react.default.createElement("div", {
      className: cx(styles.leftGradient, showStartGradient && styles.gradient_reveal)
    }, showArrows ? _react.default.createElement("button", {
      className: cx(styles.leftArrow, hideScrollbar && styles.arrow_hideScrollbar),
      type: "button",
      onClick: this.handleScrollLeft
    }, _react.default.createElement(_DirectionalIcon.default, {
      decorative: true,
      direction: "left",
      left: _IconChevronLeft.default,
      right: _IconChevronRight.default,
      size: "2em"
    })) : _react.default.createElement("span", {
      className: cx(styles.scrollTrigger),
      onMouseEnter: this.handleScrollLeft,
      onMouseLeave: this.handleScrollStop
    })), _react.default.createElement("div", {
      className: cx(styles.rightGradient, showEndGradient && styles.gradient_reveal)
    }, showArrows ? _react.default.createElement("button", {
      className: cx(styles.rightArrow, hideScrollbar && styles.arrow_hideScrollbar),
      type: "button",
      onClick: this.handleScrollRight
    }, _react.default.createElement(_DirectionalIcon.default, {
      decorative: true,
      direction: "right",
      left: _IconChevronLeft.default,
      right: _IconChevronRight.default,
      size: "2em"
    })) : _react.default.createElement("span", {
      className: cx(styles.scrollTrigger),
      onMouseEnter: this.handleScrollRight,
      onMouseLeave: this.handleScrollStop
    })), _react.default.createElement("div", {
      ref: this.handleScrollerRef,
      className: cx(styles.scroller, hideScrollbar && styles.scroller_hideScrollbar),
      onScroll: this.handleScrollThrottled
    }, _react.default.createElement("div", {
      ref: this.handleContentsRef,
      className: cx(styles.contents)
    }, children)));
  }

}

exports.GradientScroller = GradientScroller;

_defineProperty(GradientScroller, "propTypes", {
  autoScrollIncrement: _propTypes.default.number,
  children: _propTypes.default.node,
  hideScrollbar: _propTypes.default.bool,
  showArrows: _propTypes.default.bool
});

_defineProperty(GradientScroller, "defaultProps", {
  children: null,
  hideScrollbar: false,
  showArrows: false
});

var _default = (0, _withStyles.default)(_styles.styleSheet)(GradientScroller);

exports.default = _default;
