// @airbnb/lunar/esm/components/SortCarets/index.js
import _pt from "prop-types";
import React from 'react';
import IconCaretUp from '@airbnb/lunar-icons/lib/interface/IconCaretUp';
import IconCaretDown from '@airbnb/lunar-icons/lib/interface/IconCaretDown';
import useStyles from '../../hooks/useStyles';
import { styleSheet } from './styles';

/** Carets to indicate sorting on a table. */
export default function SortCarets(_ref) {
  let {
    down,
    enableDown,
    enableUp,
    up
  } = _ref;
  const [styles, cx] = useStyles(styleSheet);
  return React.createElement("span", {
    className: cx(styles.container, up && down && styles.container_full)
  }, up && React.createElement("span", {
    className: cx(styles.caret, styles.caret_up, enableUp ? styles.caret_active : styles.caret_inactive)
  }, React.createElement(IconCaretUp, {
    decorative: true,
    size: "1.6em"
  })), down && React.createElement("span", {
    className: cx(styles.caret, styles.caret_down, enableDown ? styles.caret_active : styles.caret_inactive)
  }, React.createElement(IconCaretDown, {
    decorative: true,
    size: "1.6em"
  })));
}
SortCarets.propTypes = {
  down: _pt.bool,
  enableDown: _pt.bool,
  enableUp: _pt.bool,
  up: _pt.bool
};
