// @airbnb/lunar/lib/components/NotchedBox/styles.js
"use strict";

exports.__esModule = true;
exports.createPosition = createPosition;
exports.styleSheet = exports.default = exports.NOTCH_SPACING = exports.NOTCH_SIZE = void 0;
const NOTCH_SIZE = 1.5;
exports.NOTCH_SIZE = NOTCH_SIZE;
const NOTCH_SPACING = 1.5;
exports.NOTCH_SPACING = NOTCH_SPACING;

function createPosition(offset) {
  if (typeof offset === 'number' && offset < 0) {
    return {
      right: Math.abs(offset)
    };
  }

  return {
    left: offset
  };
}

const styleSheet = (_ref) => {
  let {
    ui,
    color,
    unit
  } = _ref;
  const {
    border,
    borderRadius
  } = ui;
  const notchSide = NOTCH_SIZE * unit;
  const offset = -notchSide / 2;
  const left = unit + notchSide / Math.SQRT2;
  return {
    box: {
      position: 'relative',
      borderRadius
    },
    box_inline: {
      display: 'inline-block'
    },
    notch: {
      width: notchSide,
      height: notchSide,
      position: 'absolute',
      backgroundColor: color.accent.bg,
      transform: "translate(" + offset + "px, " + offset + "px) rotate(-45deg)"
    },
    notch_position: createPosition(left),
    notch_below: {
      bottom: -notchSide
    },
    border: {
      border
    },
    inverted: {
      borderColor: color.clear,
      backgroundColor: color.core.neutral[6]
    },
    content: {
      padding: NOTCH_SPACING * unit,
      position: 'relative',
      backgroundColor: color.accent.bg,
      borderRadius
    }
  };
};

exports.styleSheet = styleSheet;
var _default = styleSheet;
exports.default = _default;
