// interweave-autolink/lib/EmailMatcher.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _interweave = require("interweave");

var _Email = _interopRequireDefault(require("./Email"));

var _constants = require("./constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var EmailMatcher = function (_Matcher) {
  _inheritsLoose(EmailMatcher, _Matcher);

  function EmailMatcher() {
    return _Matcher.apply(this, arguments) || this;
  }

  var _proto = EmailMatcher.prototype;

  _proto.replaceWith = function replaceWith(children, props) {
    return _react.default.createElement(_Email.default, props, children);
  };

  _proto.asTag = function asTag() {
    return 'a';
  };

  _proto.match = function match(string) {
    return this.doMatch(string, _constants.EMAIL_PATTERN, function (matches) {
      return {
        email: matches[0],
        emailParts: {
          host: matches[2],
          username: matches[1]
        }
      };
    });
  };

  return EmailMatcher;
}(_interweave.Matcher);

exports.default = EmailMatcher;
