// interweave-autolink/lib/Url.js
"use strict";

exports.__esModule = true;
exports.default = Url;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Link = _interopRequireDefault(require("./Link"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function Url(_ref) {
  var children = _ref.children,
      url = _ref.url,
      urlParts = _ref.urlParts,
      props = _objectWithoutPropertiesLoose(_ref, ["children", "url", "urlParts"]);

  var href = url;

  if (!href.match(/^https?:\/\//)) {
    href = "http://" + href;
  }

  return _react.default.createElement(_Link.default, _extends({}, props, {
    href: href
  }), children);
}

Url.propTypes = {
  children: _propTypes.default.any.isRequired,
  url: _propTypes.default.string.isRequired,
  urlParts: _propTypes.default.shape({
    auth: _propTypes.default.string.isRequired,
    fragment: _propTypes.default.string.isRequired,
    host: _propTypes.default.string.isRequired,
    path: _propTypes.default.string.isRequired,
    port: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]).isRequired,
    query: _propTypes.default.string.isRequired,
    scheme: _propTypes.default.string.isRequired
  }).isRequired
};
