// @airbnb/lunar/esm/components/Title/index.js
import _pt from "prop-types";
import React from 'react';
import { mutuallyExclusiveTrueProps, forbidExtraProps } from 'airbnb-prop-types';
import useStyles from '../../hooks/useStyles';
import { styleSheet } from './styles';

/** Display a string of text as a heading and or section title. */
function Title(_ref) {
  let {
    centerAlign,
    children,
    endAlign,
    inline,
    inverted,
    level,
    muted,
    primary
  } = _ref;
  const [styles, cx] = useStyles(styleSheet);
  const Tag = "h" + level;
  return React.createElement(Tag, {
    className: cx(styles.title, level === 1 && styles.title_level1, level === 2 && styles.title_level2, level === 3 && styles.title_level3, inline && styles.title_inline, inverted && styles.title_inverted, muted && styles.title_muted, primary && styles.title_primary, centerAlign && styles.title_center, endAlign && styles.title_right)
  }, children);
}

const stateProp = mutuallyExclusiveTrueProps('muted', 'inverted', 'primary');
const alignProp = mutuallyExclusiveTrueProps('centerAlign', 'endAlign');
Title.propTypes = {
  level: _pt.oneOf([1, 2, 3]).isRequired,
  inline: _pt.bool,
  children: _pt.node,
  centerAlign: alignProp,
  endAlign: alignProp,
  inverted: stateProp,
  muted: stateProp,
  primary: stateProp
};
export default Title;
