// @airbnb/lunar/lib/components/Modal/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.Modal = void 0;

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _Portal = _interopRequireDefault(require("../Portal"));

var _Inner = _interopRequireDefault(require("./private/Inner"));

var _keys = require("../../keys");

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** A modal component with a backdrop and a standardized layout. */
class Modal extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleClose", event => {
      this.props.onClose(event);
    });

    _defineProperty(this, "handleKeyUp", event => {
      if (event.key === _keys.ESCAPE) {
        this.handleClose(event);
      }
    });
  }

  componentDidMount() {
    document.body.style.overflow = 'hidden';
  }

  componentWillUnmount() {
    document.body.style.overflow = '';
  }

  render() {
    const _this$props = this.props,
          {
      cx,
      styles
    } = _this$props,
          otherProps = _objectWithoutPropertiesLoose(_this$props, ["cx", "styles"]);

    return _react.default.createElement(_Portal.default, null, _react.default.createElement("div", {
      className: cx(styles.container)
    }, _react.default.createElement("div", {
      role: "presentation",
      className: cx(styles.wrapper),
      onKeyUp: this.handleKeyUp
    }, _react.default.createElement(_Inner.default, otherProps))));
  }

}

exports.Modal = Modal;

var _default = (0, _withStyles.default)(_styles.styleSheet)(Modal);

exports.default = _default;
