// @airbnb/lunar/lib/components/Modal/styles.js
"use strict";

exports.__esModule = true;
exports.styleSheetInnerContent = exports.styleSheetInner = exports.styleSheetImageLayout = exports.styleSheet = exports.MODAL_MAX_WIDTH_LARGE = exports.MODAL_MAX_WIDTH_MEDIUM = exports.MODAL_MAX_WIDTH_SMALL = void 0;

var _constants = require("../../constants");

var _toRGBA = _interopRequireDefault(require("../../utils/toRGBA"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const MODAL_MAX_WIDTH_SMALL = 400;
exports.MODAL_MAX_WIDTH_SMALL = MODAL_MAX_WIDTH_SMALL;
const MODAL_MAX_WIDTH_MEDIUM = 600;
exports.MODAL_MAX_WIDTH_MEDIUM = MODAL_MAX_WIDTH_MEDIUM;
const MODAL_MAX_WIDTH_LARGE = 800;
exports.MODAL_MAX_WIDTH_LARGE = MODAL_MAX_WIDTH_LARGE;

const styleSheet = (_ref) => {
  let {
    color,
    unit
  } = _ref;
  return {
    container: {
      bottom: 0,
      left: 0,
      overflowY: 'auto',
      position: 'fixed',
      right: 0,
      top: 0,
      zIndex: _constants.Z_INDEX_MODAL
    },
    wrapper: {
      alignItems: 'center',
      backgroundColor: (0, _toRGBA.default)(color.core.neutral[6], 75),
      display: 'flex',
      justifyContent: 'center',
      minHeight: '100%',
      padding: unit * 2,
      width: '100%'
    }
  };
};

exports.styleSheet = styleSheet;

const styleSheetImageLayout = (_ref2) => {
  let {
    responsive,
    ui,
    unit
  } = _ref2;
  return {
    splitContent: {
      display: 'flex'
    },
    splitContentPane: {
      flex: '1'
    },
    splitContentImagePane: {
      borderLeft: ui.border,
      alignItems: 'center',
      display: 'flex',
      justifyContent: 'center',
      overflow: 'hidden',
      position: 'relative',
      '@media': {
        [responsive.xsmall]: {
          display: 'none'
        }
      }
    },
    image: {
      display: 'block',
      objectFit: 'contain',
      maxWidth: '100%',
      maxHeight: '100%',
      margin: unit * 3
    },
    imageCover: {
      objectFit: 'cover',
      height: '100%',
      width: '100%',
      margin: 0
    }
  };
};

exports.styleSheetImageLayout = styleSheetImageLayout;

const styleSheetInner = (_ref3) => {
  let {
    color,
    responsive,
    ui,
    unit
  } = _ref3;
  return {
    content: {
      backgroundColor: color.accent.bg,
      borderRadius: ui.borderRadius,
      boxShadow: ui.boxShadowLarge,
      maxWidth: MODAL_MAX_WIDTH_MEDIUM,
      position: 'relative',
      width: '100%',
      ':focus': {
        outline: 'none'
      }
    },
    content_small: {
      maxWidth: MODAL_MAX_WIDTH_SMALL
    },
    content_large: {
      maxWidth: MODAL_MAX_WIDTH_LARGE
    },
    content_fluid: {
      maxWidth: '70%',
      '@media': {
        [responsive.small]: {
          maxWidth: '85%'
        }
      }
    }
  };
};

exports.styleSheetInner = styleSheetInner;

const styleSheetInnerContent = (_ref4) => {
  let {
    color,
    ui,
    unit
  } = _ref4;
  return {
    wrapper: {
      position: 'relative'
    },
    header: {
      padding: unit * 3 + "px " + unit * 4 + "px " + unit * 3 + "px " + unit * 3 + "px"
    },
    header_scrollable: {
      position: 'relative',
      ':after': {
        content: '" "',
        position: 'absolute',
        top: '100%',
        left: 0,
        right: 0,
        height: unit / 2,
        background: "linear-gradient(" + (0, _toRGBA.default)(color.core.neutral[6], 15) + ", " + (0, _toRGBA.default)(color.core.neutral[6], 0) + ")"
      }
    },
    close: {
      position: 'absolute',
      top: unit * 2,
      right: unit * 2,
      zIndex: _constants.Z_INDEX_MODAL
    },
    close_float: {
      float: 'right',
      position: 'relative',
      top: 0,
      right: 0,
      margin: unit * 2 + "px " + unit * 2 + "px " + unit / 2 + "px " + unit / 2 + "px"
    },
    body: {
      padding: "0 " + unit * 3 + "px"
    },
    body_paddingBottom: {
      paddingBottom: unit * 3
    },
    body_paddingTop: {
      paddingTop: unit * 3
    },
    body_scrollable: {
      paddingBottom: unit * 3,
      maxHeight: 300,
      overflow: 'auto',
      ':before': {
        content: '" "',
        position: 'sticky',
        display: 'block',
        width: "calc(100% + " + unit * 6 + "px)",
        marginLeft: -unit * 3,
        height: unit / 2,
        background: color.accent.bg
      }
    },
    body_scrollableSmall: {
      maxHeight: 160
    },
    body_scrollableLarge: {
      maxHeight: 500
    },
    footer: {
      padding: unit * 3
    },
    footer_scrollable: {
      borderTop: ui.border
    }
  };
};

exports.styleSheetInnerContent = styleSheetInnerContent;
