// @airbnb/lunar/lib/components/Pagination/styles.js
"use strict";

exports.__esModule = true;
exports.styleSheet = exports.default = void 0;

const styleSheet = (_ref) => {
  let {
    unit
  } = _ref;
  return {
    wrapper: {
      display: 'grid',
      gridTemplateAreas: '"previous page next"',
      gridTemplateColumns: 'auto 1fr auto',
      gridColumnGap: unit * 2,
      alignItems: 'center',
      justifyItems: 'center'
    },
    wrapper_endAlign: {
      gridTemplateAreas: '"page previous next"',
      gridTemplateColumns: 'auto',
      justifyContent: 'end'
    },
    wrapper_centerAlign: {
      gridTemplateColumns: 'auto',
      justifyContent: 'center'
    },
    wrapper_startAlign: {
      gridTemplateAreas: '"previous next page"',
      gridTemplateColumns: 'auto',
      justifyContent: 'start'
    },
    page: {
      gridArea: 'page'
    },
    previous: {
      gridArea: 'previous'
    },
    next: {
      gridArea: 'next'
    }
  };
};

exports.styleSheet = styleSheet;
var _default = styleSheet;
exports.default = _default;
