// @airbnb/lunar/lib/components/TextArea/Proofreader/Mark.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("../../../composers/withStyles"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Mark extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "ref", _react.default.createRef());

    _defineProperty(this, "handleSelect", () => {
      const {
        ref
      } = this;

      if (this.props.selected && ref.current) {
        this.props.onSelect(ref.current.offsetTop + ref.current.offsetHeight, ref.current.offsetLeft);
      }
    });
  }

  componentDidMount() {
    this.handleSelect();
  }

  componentDidUpdate(prevProps) {
    if (this.props.selected !== prevProps.selected) {
      this.handleSelect();
    }
  }

  render() {
    const {
      cx,
      children,
      selected,
      alwaysHighlight,
      styles
    } = this.props;
    const highlight = selected || alwaysHighlight;
    return _react.default.createElement("mark", {
      ref: this.ref,
      className: cx(styles.mark, highlight && styles.mark_highlight)
    }, children);
  }

}

_defineProperty(Mark, "propTypes", {
  children: _propTypes.default.any.isRequired,
  selected: _propTypes.default.bool.isRequired,
  onSelect: _propTypes.default.func.isRequired,
  alwaysHighlight: _propTypes.default.bool
});

_defineProperty(Mark, "defaultProps", {
  alwaysHighlight: false
});

var _default = (0, _withStyles.default)(_styles.styleSheetMark, {
  extendable: true
})(Mark);

exports.default = _default;
