// @airbnb/lunar/lib/components/Card/Content.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _Image = _interopRequireDefault(require("../Image"));

var _Row = _interopRequireDefault(require("../Row"));

var _Spacing = _interopRequireDefault(require("../Spacing"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function getSideImageWidth(_ref) {
  let {
    large,
    small
  } = _ref;

  if (small) {
    return 80;
  }

  if (large) {
    return 195;
  }

  return 152;
}

/** Content block within a card. */
function CardContent(_ref2) {
  let {
    after,
    afterImageSrc,
    before,
    beforeImageSrc,
    children,
    compact,
    large,
    maxHeight,
    middleAlign,
    minHeight,
    onClick,
    small,
    topImageSrc,
    truncated
  } = _ref2;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheetContent);
  const ContainerTag = onClick ? 'button' : 'div';
  const props = onClick ? {
    type: 'button',
    onClick
  } : {};
  const spacing = compact ? 1.5 : 3;
  const horizontalSpacing = compact ? 0 : 1;
  let afterContent = after ? _react.default.createElement("div", {
    className: cx(styles.side, compact && styles.side_compact, styles.after, compact && styles.after_compact)
  }, after) : null;

  if (!afterContent && afterImageSrc) {
    afterContent = _react.default.createElement(_Image.default, {
      background: true,
      cover: true,
      alt: "",
      height: "100%",
      width: getSideImageWidth({
        large,
        small
      }),
      src: afterImageSrc
    });
  }

  let beforeContent = before ? _react.default.createElement("div", {
    className: cx(styles.side, compact && styles.side_compact, styles.before, compact && styles.before_compact)
  }, before) : null;

  if (!beforeContent && beforeImageSrc) {
    beforeContent = _react.default.createElement(_Image.default, {
      background: true,
      cover: true,
      alt: "",
      height: "100%",
      width: getSideImageWidth({
        large,
        small
      }),
      src: beforeImageSrc
    });
  }

  return (// @ts-ignore [ts] JSX element type 'ContainerTag' does not have any construct or call signatures. [2604]
    _react.default.createElement(ContainerTag, _extends({}, props, {
      className: cx(styles.container, onClick && styles.container_button)
    }), topImageSrc && _react.default.createElement("div", {
      className: cx(styles.topImage, large && styles.topImage_large)
    }, _react.default.createElement("img", {
      className: cx(styles.image),
      alt: "",
      height: "100%",
      src: topImageSrc,
      width: "100%"
    })), _react.default.createElement(_Row.default, {
      after: afterContent,
      before: beforeContent,
      maxHeight: maxHeight,
      middleAlign: middleAlign,
      minHeight: minHeight,
      truncated: truncated
    }, _react.default.createElement(_Spacing.default, {
      inner: true,
      left: beforeContent ? horizontalSpacing : spacing,
      right: afterContent ? horizontalSpacing : spacing,
      vertical: spacing
    }, children)))
  );
}

const imageUrlTypePropType = (0, _airbnbPropTypes.mutuallyExclusiveProps)(_propTypes.default.string, 'beforeImageSrc', 'afterImageSrc', 'topImageSrc');
const afterPropType = (0, _airbnbPropTypes.mutuallyExclusiveProps)(_propTypes.default.oneOfType([_propTypes.default.node, imageUrlTypePropType]), 'after', 'afterImageSrc');
const beforePropType = (0, _airbnbPropTypes.mutuallyExclusiveProps)(_propTypes.default.oneOfType([_propTypes.default.node, imageUrlTypePropType]), 'before', 'beforeImageSrc');
const imageSizePropType = (0, _airbnbPropTypes.mutuallyExclusiveProps)(_propTypes.default.bool, 'small', 'large');
CardContent.propTypes = {
  onClick: _propTypes.default.func,
  truncated: _propTypes.default.bool,
  minHeight: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  middleAlign: _propTypes.default.bool,
  maxHeight: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  compact: _propTypes.default.bool,
  children: _propTypes.default.any.isRequired,
  after: afterPropType,
  afterImageSrc: imageUrlTypePropType,
  before: beforePropType,
  beforeImageSrc: imageUrlTypePropType,
  large: imageSizePropType,
  small: imageSizePropType,
  topImageSrc: imageUrlTypePropType
};
var _default = CardContent;
exports.default = _default;
