// @airbnb/lunar/lib/keys.js
"use strict";

exports.__esModule = true;
exports.DELETE = exports.ARROW_DOWN = exports.ARROW_RIGHT = exports.ARROW_UP = exports.ARROW_LEFT = exports.HOME = exports.END = exports.PAGEDOWN = exports.PAGEUP = exports.SPACE = exports.ESCAPE = exports.CAPSLOCK = exports.ALT = exports.CONTROL = exports.SHIFT = exports.ENTER = exports.TAB = exports.BACKSPACE = void 0;
// KEYCODES
const BACKSPACE = 'Backspace';
exports.BACKSPACE = BACKSPACE;
const TAB = 'Tab';
exports.TAB = TAB;
const ENTER = 'Enter';
exports.ENTER = ENTER;
const SHIFT = 'Shift';
exports.SHIFT = SHIFT;
const CONTROL = 'Control';
exports.CONTROL = CONTROL;
const ALT = 'Alt';
exports.ALT = ALT;
const CAPSLOCK = 'CapsLock';
exports.CAPSLOCK = CAPSLOCK;
const ESCAPE = 'Escape';
exports.ESCAPE = ESCAPE;
const SPACE = ' ';
exports.SPACE = SPACE;
const PAGEUP = 'PageUp';
exports.PAGEUP = PAGEUP;
const PAGEDOWN = 'PageDown';
exports.PAGEDOWN = PAGEDOWN;
const END = 'End';
exports.END = END;
const HOME = 'Home';
exports.HOME = HOME;
const ARROW_LEFT = 'ArrowLeft';
exports.ARROW_LEFT = ARROW_LEFT;
const ARROW_UP = 'ArrowUp';
exports.ARROW_UP = ARROW_UP;
const ARROW_RIGHT = 'ArrowRight';
exports.ARROW_RIGHT = ARROW_RIGHT;
const ARROW_DOWN = 'ArrowDown';
exports.ARROW_DOWN = ARROW_DOWN;
const DELETE = 'Delete';
exports.DELETE = DELETE;
