// @airbnb/lunar/esm/components/Translate/index.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import Core from '../..';

/** Translate a phrase with a key, informational context, and dynamic params. */
export default class Translate extends React.PureComponent {
  static phrase(phrase, params, options) {
    return Core.translate(phrase, params, options);
  }

  render() {
    const {
      translatorComponent: Translator
    } = Core.settings;

    const _this$props = this.props,
          {
      k: key,
      phrase,
      context,
      html
    } = _this$props,
          params = _objectWithoutPropertiesLoose(_this$props, ["children", "k", "phrase", "context", "html"]);

    const options = {
      context,
      html,
      key
    };

    if (!Translator) {
      return React.createElement("span", null, Core.translate(phrase, params, options));
    }

    return React.createElement(Translator, _extends({
      k: key,
      phrase: phrase,
      context: context,
      html: html
    }, params));
  }

}

_defineProperty(Translate, "defaultProps", {
  html: false
});
