// @airbnb/lunar/esm/components/Link/index.js
import _pt from "prop-types";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import { mutuallyExclusiveTrueProps, forbidExtraProps } from 'airbnb-prop-types';
import withStyles from '../../composers/withStyles';
import ButtonOrLink from '../private/ButtonOrLink';
import Text from '../Text';
import { styleSheet } from './styles';
const sizingProp = mutuallyExclusiveTrueProps('small', 'large');
const stateProp = mutuallyExclusiveTrueProps('disabled', 'muted', 'inverted');

/** A standard link for... linking to things. */
export class Link extends React.Component {
  render() {
    const _this$props = this.props,
          {
      cx,
      block,
      baseline,
      children,
      disabled,
      inverted,
      large,
      muted,
      small,
      bold,
      styles
    } = _this$props,
          restProps = _objectWithoutPropertiesLoose(_this$props, ["cx", "block", "baseline", "children", "disabled", "inverted", "large", "muted", "small", "bold", "styles"]);

    return React.createElement(Text, {
      inline: !block,
      baseline: baseline,
      small: small,
      large: large,
      bold: bold
    }, React.createElement(ButtonOrLink, _extends({}, restProps, {
      disabled: disabled,
      className: cx(styles.link, inverted && styles.link_inverted, muted && styles.link_muted, disabled && styles.link_disabled, block && styles.link_block, baseline && styles.link_baseline)
    }), children));
  }

}

_defineProperty(Link, "propTypes", {
  bold: _pt.bool,
  baseline: _pt.bool,
  block: _pt.bool,
  disabled: stateProp,
  inverted: stateProp,
  large: sizingProp,
  muted: stateProp,
  small: sizingProp
});

_defineProperty(Link, "defaultProps", {
  baseline: false,
  block: false,
  bold: false,
  disabled: false,
  inverted: false,
  large: false,
  muted: false,
  small: false
});

export default withStyles(styleSheet, {
  extendable: true
})(Link);
