// @airbnb/lunar/esm/components/Dropdown/index.js
import _pt from "prop-types";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import withStyles from '../../composers/withStyles';

/** An abstract component for displaing menus and overlays over content. */
class Dropdown extends React.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "ref", React.createRef());

    _defineProperty(this, "handleClick", event => {
      const {
        current
      } = this.ref;

      if (current && current.contains(event.target)) {
        if (this.props.onClickInside) {
          this.props.onClickInside(event);
        }

        return;
      }

      if (this.props.onClickOutside) {
        this.props.onClickOutside(event);
      }
    });
  }

  componentDidMount() {
    if (this.props.visible) {
      document.addEventListener('click', this.handleClick, true);
    }
  }

  componentDidUpdate(prevProps) {
    if (prevProps.visible !== this.props.visible) {
      if (this.props.visible) {
        document.addEventListener('click', this.handleClick, true);
      } else {
        document.removeEventListener('click', this.handleClick, true);
      }
    }
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleClick, true);
  }

  render() {
    const _this$props = this.props,
          {
      cx,
      children,
      fixed,
      onBlur,
      onFocus,
      tabIndex,
      zIndex
    } = _this$props,
          props = _objectWithoutPropertiesLoose(_this$props, ["cx", "children", "fixed", "onBlur", "onFocus", "tabIndex", "zIndex", "visible", "onClickOutside", "styles"]);

    const style = _extends({
      position: fixed ? 'fixed' : 'absolute',
      zIndex: zIndex || 'auto'
    }, props); // Set top by default if neither are defined


    if (!('bottom' in props) && !('top' in props)) {
      style.top = '100%';
    } // Set left by default if neither are defined


    if (!('left' in props) && !('right' in props)) {
      style.left = 0;
    }

    return React.createElement("div", {
      ref: this.ref,
      className: cx(style),
      tabIndex: tabIndex,
      onBlur: onBlur,
      onFocus: onFocus
    }, children);
  }

}

_defineProperty(Dropdown, "propTypes", {
  bottom: _pt.oneOfType([_pt.number, _pt.string]),
  children: _pt.any.isRequired,
  fixed: _pt.bool,
  left: _pt.oneOfType([_pt.number, _pt.string]),
  onBlur: _pt.func,
  onClickInside: _pt.func,
  onClickOutside: _pt.func,
  onFocus: _pt.func,
  right: _pt.oneOfType([_pt.number, _pt.string]),
  tabIndex: _pt.number,
  top: _pt.oneOfType([_pt.number, _pt.string]),
  visible: _pt.bool,
  zIndex: _pt.oneOfType([_pt.number, _pt.oneOf(['auto'])])
});

_defineProperty(Dropdown, "defaultProps", {
  fixed: false,
  visible: false
});

export default withStyles(() => ({}))(Dropdown);
