// @airbnb/lunar/lib/components/TrackingBoundary/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _componentName = _interopRequireDefault(require("../../prop-types/componentName"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Wraps a component and provides a context stack for click and keydown tracking purposes. */
class TrackingBoundary extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleTrackContext", (_ref) => {
      let {
        nativeEvent
      } = _ref;
      const {
        name
      } = this.props;

      if (!name) {
        return;
      }

      if (nativeEvent.trackingContext) {
        nativeEvent.trackingContext.push(name);
      } else {
        nativeEvent.trackingContext = [name];
      }
    });
  }

  render() {
    const {
      children,
      name
    } = this.props;

    if (!name) {
      return children;
    }

    return _react.default.createElement("tracking-boundary", {
      "data-tracking-name": name,
      onClick: this.handleTrackContext,
      onKeyDown: this.handleTrackContext
    }, children);
  }

}

exports.default = TrackingBoundary;

_defineProperty(TrackingBoundary, "propTypes", {
  children: _propTypes.default.any.isRequired,
  name: _componentName.default
});

_defineProperty(TrackingBoundary, "defaultProps", {
  name: ''
});
