// @airbnb/lunar/esm/components/Rating/index.js
import _pt from "prop-types";
import React from 'react';
import IconStar from '@airbnb/lunar-icons/lib/interface/IconStar';
import IconStarFull from '@airbnb/lunar-icons/lib/interface/IconStarFull';
import IconStarHalf from '@airbnb/lunar-icons/lib/interface/IconStarHalf';
import useTheme from '../../hooks/useTheme';
import Row from '../Row';
import Text from '../Text';
const NUMBER_OF_STARS = 5;
const STARS = new Array(NUMBER_OF_STARS).fill(null);

/** Display a star rating and review count. */
export default function Rating(_ref) {
  let {
    large,
    micro,
    rating = 0,
    reviews,
    small
  } = _ref;
  const theme = useTheme();
  return React.createElement(Text, {
    large: large,
    micro: micro,
    small: small
  }, React.createElement(Row, {
    inline: true,
    middleAlign: true,
    after: reviews
  }, STARS.map((star, index) => {
    const key = "star-" + index;
    const color = rating > index || rating > index && rating < index + 1 ? theme.color.core.primary[3] : theme.color.core.neutral[3];
    let Star = IconStar;

    if (rating > index && rating < index + 1) {
      Star = IconStarHalf;
    } else if (rating > index) {
      Star = IconStarFull;
    }

    return React.createElement(Star, {
      key: key,
      decorative: true,
      color: color,
      size: "1.15em"
    });
  })));
}
Rating.propTypes = {
  large: _pt.bool,
  micro: _pt.bool,
  rating: _pt.number,
  reviews: _pt.oneOfType([_pt.string, _pt.number]),
  small: _pt.bool
};
