// @airbnb/lunar/esm/components/DataTable/ColumnLabels.js
import React from 'react';
import { SortDirection } from 'react-virtualized';
import SortCarets from '../SortCarets';
import Spacing from '../Spacing';
import Text from '../Text';
import { caseColumnLabel, getHeight } from './helpers';

/** See https://github.com/bvaughn/react-virtualized/blob/master/source/Table/defaultHeaderRowRenderer.js.
    In order to overwrite the existing labels and carets in defaultHeaderRowRenderer,
    we clone them from props (children[0] = label, children[1] = carets), build around their data. */
export default function ColumnLabels(_ref) {
  let {
    cx,
    styles,
    columnToLabel = {},
    showColumnDividers,
    rowHeight,
    columnHeaderHeight,
    expandable,
    selectable,
    columnMetadata,
    columnLabelCase
  } = _ref;
  return (_ref2) => {
    let {
      className,
      columns,
      style
    } = _ref2;
    const leftmostIdx = Number(expandable) + Number(selectable);
    const heightStyle = {
      height: getHeight(rowHeight, columnHeaderHeight)
    };
    const rightAlignmentStyle = {
      justifyContent: 'flex-end',
      width: '100%'
    };
    const newColumns = columns.map((col, idx) => {
      if (!React.isValidElement(col)) {
        return col;
      }

      const {
        children
      } = col.props;
      const key = children[0].props.children;
      const label = columnToLabel[key] ? columnToLabel[key] : key && caseColumnLabel(key, columnLabelCase);
      const sort = children[1] && children[1].props.sortDirection;
      const isLeftmost = idx === leftmostIdx;
      const isRightmost = idx === columns.length - 1;
      const indent = !((expandable || selectable) && isLeftmost);
      const showDivider = showColumnDividers && !!label && !isRightmost;
      const sortable = !columnMetadata || !columnMetadata[key] || columnMetadata[key].disableSorting !== 1;
      const newHeader = React.createElement(Spacing, {
        left: indent ? 2 : 0
      }, React.createElement("div", {
        style: heightStyle,
        className: cx(showDivider && styles && styles.column_divider)
      }, React.createElement("div", {
        style: columnMetadata && columnMetadata[key] && columnMetadata[key].rightAlign ? rightAlignmentStyle : {},
        className: cx(styles && styles.headerRow)
      }, React.createElement(Text, {
        micro: true,
        muted: true
      }, label), label && sortable && React.createElement(Spacing, {
        inline: true,
        left: 0.5
      }, React.createElement(SortCarets, {
        down: true,
        up: true,
        enableDown: sort === SortDirection.DESC,
        enableUp: sort === SortDirection.ASC
      })))));
      return React.cloneElement(col, col.props, newHeader);
    });
    return React.createElement("div", {
      role: "row",
      style: style,
      className: cx(className, styles && styles.column_header)
    }, newColumns);
  };
}
