// @airbnb/lunar/lib/components/Title/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** Display a string of text as a heading and or section title. */
function Title(_ref) {
  let {
    centerAlign,
    children,
    endAlign,
    inline,
    inverted,
    level,
    muted,
    primary
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheet);
  const Tag = "h" + level;
  return _react.default.createElement(Tag, {
    className: cx(styles.title, level === 1 && styles.title_level1, level === 2 && styles.title_level2, level === 3 && styles.title_level3, inline && styles.title_inline, inverted && styles.title_inverted, muted && styles.title_muted, primary && styles.title_primary, centerAlign && styles.title_center, endAlign && styles.title_right)
  }, children);
}

const stateProp = (0, _airbnbPropTypes.mutuallyExclusiveTrueProps)('muted', 'inverted', 'primary');
const alignProp = (0, _airbnbPropTypes.mutuallyExclusiveTrueProps)('centerAlign', 'endAlign');
Title.propTypes = {
  level: _propTypes.default.oneOf([1, 2, 3]).isRequired,
  inline: _propTypes.default.bool,
  children: _propTypes.default.node,
  centerAlign: alignProp,
  endAlign: alignProp,
  inverted: stateProp,
  muted: stateProp,
  primary: stateProp
};
var _default = Title;
exports.default = _default;
