// @airbnb/lunar/lib/components/Translate/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _ = _interopRequireDefault(require("../.."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Translate a phrase with a key, informational context, and dynamic params. */
class Translate extends _react.default.PureComponent {
  static phrase(phrase, params, options) {
    return _.default.translate(phrase, params, options);
  }

  render() {
    const {
      translatorComponent: Translator
    } = _.default.settings;

    const _this$props = this.props,
          {
      k: key,
      phrase,
      context,
      html
    } = _this$props,
          params = _objectWithoutPropertiesLoose(_this$props, ["children", "k", "phrase", "context", "html"]);

    const options = {
      context,
      html,
      key
    };

    if (!Translator) {
      return _react.default.createElement("span", null, _.default.translate(phrase, params, options));
    }

    return _react.default.createElement(Translator, _extends({
      k: key,
      phrase: phrase,
      context: context,
      html: html
    }, params));
  }

}

exports.default = Translate;

_defineProperty(Translate, "defaultProps", {
  html: false
});
