// @airbnb/lunar/lib/components/Toast/styles.js
"use strict";

exports.__esModule = true;
exports.styleSheet = exports.default = void 0;

const styleSheet = (_ref) => {
  let {
    color,
    ui,
    unit
  } = _ref;
  return {
    container: {
      marginTop: unit * 1.5,
      padding: unit * 2,
      backgroundColor: color.core.neutral[6],
      boxShadow: ui.boxShadowLarge,
      willChange: 'transform, opacity',
      transform: 'translateY(-100%)',
      transition: 'transform .3s, opacity .3s',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'space-between',
      opacity: 0,
      lineHeight: 1,
      borderRadius: ui.borderRadius
    },
    container_visible: {
      transform: 'translateY(0)',
      opacity: 1
    },
    container_success: {
      backgroundColor: color.core.success[3]
    },
    container_danger: {
      backgroundColor: color.core.danger[3]
    },
    right: {
      textAlign: 'right',
      margin: -unit,
      marginLeft: unit
    }
  };
};

exports.styleSheet = styleSheet;
var _default = styleSheet;
exports.default = _default;
