// @airbnb/lunar/lib/components/HierarchyPicker/Picker/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.Picker = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Hierarchy = _interopRequireDefault(require("../Hierarchy"));

var _Search = _interopRequireDefault(require("../Search"));

var _withStyles = _interopRequireDefault(require("../../../composers/withStyles"));

var _keys = require("../../../keys");

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Picker extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      searchQuery: ''
    });

    _defineProperty(this, "ref", _react.default.createRef());

    _defineProperty(this, "handleKeyDown", event => {
      const {
        searchQuery
      } = this.state;
      if (searchQuery) return;

      switch (event.key) {
        case _keys.ARROW_DOWN:
        case _keys.ARROW_UP:
          event.preventDefault();
          this.focusNext(event.key === _keys.ARROW_DOWN);
          break;

        default:
          break;
      }
    });

    _defineProperty(this, "handleSearch", searchQuery => {
      this.setState({
        searchQuery
      });
    });

    _defineProperty(this, "handleItemPicked", (def, details) => {
      if (def) {
        // null if item.readonly
        this.props.onItemPicked(def, details);
        this.props.onClose();
      }
    });
  }

  componentDidMount() {
    const [focusable] = this.getFocusables();
    const {
      scrollX,
      scrollY
    } = window; // We want to focus the first focusable item,
    // but that could cause a scroll of the document
    // depending on where the dropdown is positioned.
    // This would break initial alignment of the dropdown with the trigger.
    // so we get scroll position before focus, then set it back.

    if (focusable) {
      focusable.focus();
      window.scrollTo(scrollX, scrollY);
    }
  }

  getFocusables() {
    const el = this.ref.current;
    return el ? Array.from(el.querySelectorAll('input,[tabindex]')) : [];
  }

  focusNext(forward) {
    if (forward === void 0) {
      forward = true;
    }

    const f = this.getFocusables();
    const index = f.findIndex(el => el === document.activeElement);

    if (index >= 0) {
      if (forward) {
        f[index < f.length - 1 ? index + 1 : 0].focus();
      } else {
        f[(index || f.length) - 1].focus();
      }
    }
  }

  render() {
    const {
      cx,
      chosen,
      hierarchyMaxHeight,
      hierarchyWidth,
      indexParentPath,
      items,
      formatter,
      noResultsLabel,
      renderItem,
      searchMaxHeight,
      searchPlaceholder,
      searchWidth,
      styles,
      verticallyAlign
    } = this.props;
    const {
      searchQuery
    } = this.state;
    return (
      /* eslint-disable-next-line jsx-a11y/no-static-element-interactions */
      _react.default.createElement("div", {
        ref: this.ref,
        className: cx(styles.pane),
        onKeyDown: this.handleKeyDown
      }, _react.default.createElement(_Search.default, {
        formatter: formatter,
        indexParentPath: indexParentPath,
        items: items,
        maxHeight: searchMaxHeight,
        noResultsLabel: noResultsLabel,
        placeholder: searchPlaceholder,
        query: searchQuery,
        width: searchWidth,
        onSearch: this.handleSearch,
        onItemPicked: this.handleItemPicked
      }), !searchQuery && _react.default.createElement("div", {
        role: "listbox",
        className: cx(styles.hierarchy)
      }, _react.default.createElement(_Hierarchy.default, {
        chosen: chosen,
        items: items,
        formatter: formatter,
        hierarchyMaxHeight: hierarchyMaxHeight,
        hierarchyWidth: hierarchyWidth,
        renderItem: renderItem,
        verticallyAlign: verticallyAlign,
        onItemPicked: this.handleItemPicked
      })))
    );
  }

}

exports.Picker = Picker;

_defineProperty(Picker, "propTypes", {
  chosen: _propTypes.default.any,
  hierarchyMaxHeight: _propTypes.default.number,
  hierarchyWidth: _propTypes.default.number,
  indexParentPath: _propTypes.default.bool,
  items: _propTypes.default.arrayOf(_propTypes.default.any).isRequired,
  formatter: _propTypes.default.any.isRequired,
  noResultsLabel: _propTypes.default.string.isRequired,
  onClose: _propTypes.default.func.isRequired,
  onItemPicked: _propTypes.default.func.isRequired,
  renderItem: _propTypes.default.any,
  searchMaxHeight: _propTypes.default.number,
  searchPlaceholder: _propTypes.default.string.isRequired,
  searchWidth: _propTypes.default.number,
  verticallyAlign: _propTypes.default.bool
});

_defineProperty(Picker, "defaultProps", {
  searchWidth: 300
});

var _default = (0, _withStyles.default)(_styles.styleSheet)(Picker);

exports.default = _default;
