// @airbnb/lunar/lib/components/Image/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const backgroundAlignPropType = (0, _airbnbPropTypes.mutuallyExclusiveTrueProps)('alignBottom', 'alignTop');
const objectFitPropType = (0, _airbnbPropTypes.mutuallyExclusiveTrueProps)('contain', 'cover');

const styleSheet = (_ref) => {
  let {
    color,
    ui
  } = _ref;
  return {
    background: {
      backgroundPosition: '50% 50%',
      backgroundRepeat: 'no-repeat'
    },
    backgroundSize_cover: {
      backgroundColor: color.accent.bgHover,
      backgroundSize: 'cover'
    },
    backgroundSize_contain: {
      backgroundColor: color.accent.bg,
      backgroundSize: 'contain'
    },
    backgroundPosition_top: {
      backgroundPosition: 'top'
    },
    backgroundPosition_bottom: {
      backgroundPosition: 'bottom'
    },
    borderRadius: {
      borderRadius: ui.borderRadius
    },
    container: {
      position: 'relative'
    },
    fadeIn: {
      animationName: {
        from: {
          opacity: 0.1
        },
        to: {
          opacity: 1
        }
      },
      animationDuration: ui.transitionTime,
      animationTimingFunction: 'ease-out'
    },
    fitContainer: {
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center'
    },
    fitImage: {
      maxHeight: '100%',
      maxWidth: '100%',
      height: 'auto',
      width: 'auto'
    },
    image: {
      position: 'absolute'
    }
  };
};

/** Component that displays an image */
function Image(_ref2) {
  let {
    alignBottom = false,
    alignTop = false,
    alt,
    background = false,
    borderRadius,
    contain = false,
    cover = false,
    crossOrigin = false,
    height = 'auto',
    src,
    srcSet,
    title,
    width = '100%'
  } = _ref2;
  const [styles, cx] = (0, _useStyles.default)(styleSheet);
  let backgroundAriaLabel = {};

  if (alt) {
    backgroundAriaLabel = {
      role: 'img',
      'aria-label': alt
    };
  }

  return _react.default.createElement("div", {
    className: cx(contain && styles.fitContainer, styles.container, {
      height,
      width
    })
  }, background ? _react.default.createElement("div", _extends({
    className: cx(styles.image, styles.background, styles.fadeIn, cover && styles.backgroundSize_cover, contain && styles.backgroundSize_contain, alignTop && styles.backgroundPosition_top, alignBottom && styles.backgroundPosition_bottom, {
      width,
      height,
      backgroundImage: "url(" + src + ")"
    })
  }, backgroundAriaLabel)) : _react.default.createElement("img", {
    className: cx(borderRadius === undefined ? styles.borderRadius : {
      borderRadius
    }, contain && styles.fitImage, styles.fadeIn, styles.image),
    alt: alt,
    crossOrigin: crossOrigin ? 'anonymous' : undefined,
    height: height,
    src: src,
    srcSet: srcSet,
    title: title,
    width: width
  }));
}

Image.propTypes = {
  width: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  title: _propTypes.default.string,
  srcSet: _propTypes.default.string,
  src: _propTypes.default.string.isRequired,
  height: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  crossOrigin: _propTypes.default.bool,
  borderRadius: _propTypes.default.number,
  background: _propTypes.default.bool,
  alt: _propTypes.default.string.isRequired,
  alignBottom: backgroundAlignPropType,
  alignTop: backgroundAlignPropType,
  contain: objectFitPropType,
  cover: objectFitPropType
};
var _default = Image;
exports.default = _default;
