// @airbnb/lunar/lib/components/Grid/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _styles = require("./styles");

var _Col = _interopRequireDefault(require("./Col"));

exports.Col = _Col.default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** A grid to contain columns. */
function Grid(_ref) {
  let {
    bottomAlign,
    centerAlign,
    children,
    endAlign,
    middleAlign,
    reversed,
    startAlign,
    topAlign
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheet);
  return _react.default.createElement("section", {
    className: cx(styles.grid, reversed && styles.grid_reversed, bottomAlign && styles.grid_bottom, middleAlign && styles.grid_middle, topAlign && styles.grid_top, startAlign && styles.grid_start, endAlign && styles.grid_end, centerAlign && styles.grid_center)
  }, children);
}

const horizontalAlignProp = (0, _airbnbPropTypes.mutuallyExclusiveTrueProps)('centerAlign', 'startAlign', 'endAlign');
const verticalAlignProp = (0, _airbnbPropTypes.mutuallyExclusiveTrueProps)('bottomAlign', 'middleAlign', 'topAlign');
Grid.propTypes = {
  reversed: _propTypes.default.bool,
  bottomAlign: verticalAlignProp,
  centerAlign: horizontalAlignProp,
  children: (0, _airbnbPropTypes.childrenOfType)(_Col.default).isRequired,
  endAlign: horizontalAlignProp,
  middleAlign: verticalAlignProp,
  startAlign: horizontalAlignProp,
  topAlign: verticalAlignProp
};
var _default = Grid;
exports.default = _default;
