// @airbnb/lunar/esm/components/Table/index.js
import _pt from "prop-types";
import React from 'react';
import useStyles from '../../hooks/useStyles';
import withBoundary from '../../composers/withBoundary';
import Cell from './Cell';
import Row from './Row';
import { styleSheet } from './styles';

/** A responsive table for displaying tabular data. */
function Table(_ref) {
  let {
    bordered,
    children,
    compact,
    fixed,
    horizontal,
    loading,
    middleAlign,
    noWrap,
    striped,
    transparent,
    vertical
  } = _ref;
  const [styles, cx] = useStyles(styleSheet);
  return React.createElement("div", {
    className: cx(!noWrap && styles.responsive_wrapper)
  }, React.createElement("table", {
    className: cx(styles.table, middleAlign && styles.content_middle_align, fixed && styles.table_fixed, bordered && styles.table_bordered, horizontal && styles.table_horizontal, vertical && styles.table_vertical, compact && styles.table_compact, striped && styles.table_striped, loading && styles.table_loading, transparent && styles.table_transparent)
  }, children));
}

Table.propTypes = {
  bordered: _pt.bool,
  children: _pt.any.isRequired,
  compact: _pt.bool,
  fixed: _pt.bool,
  horizontal: _pt.bool,
  loading: _pt.bool,
  middleAlign: _pt.bool,
  noWrap: _pt.bool,
  striped: _pt.bool,
  trackingName: _pt.string,
  transparent: _pt.bool,
  vertical: _pt.bool
};
export { Cell, Row };
export default withBoundary('Table')(Table);
