// @airbnb/lunar/esm/components/HierarchyPicker/index.js
import _pt from "prop-types";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import IconCaretDown from '@airbnb/lunar-icons/lib/interface/IconCaretDown';
import Overlay from '../Overlay';
import { SPACE, ENTER } from '../../keys';
import T from '../Translate';
import Text from '../Text';
import Picker from './Picker';
import withStyles from '../../composers/withStyles';
import { styleSheet } from './styles';
import defaultFormatter from './defaultFormatter';
export class HierarchyPicker extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      open: false
    });

    _defineProperty(this, "ref", React.createRef());

    _defineProperty(this, "getLabel", chosen => {
      const label = chosen.reduce((items, path, i) => {
        if (Array.isArray(items) && items.length > 0) {
          const item = items.find((_ref) => {
            let {
              name
            } = _ref;
            return path === name;
          });

          if (!item) {
            return '';
          }

          return i < chosen.length - 1 ? item.items || [] : item.label || item.name;
        }

        return '';
      }, this.props.items);
      return typeof label === 'string' ? label : '';
    });

    _defineProperty(this, "boundFormatter", chosen => this.props.formatter(chosen, this.getLabel));

    _defineProperty(this, "toggle", () => {
      if (this.props.disabled) {
        return;
      }

      this.setState(state => ({
        open: !state.open
      }), () => {
        const {
          open
        } = this.state;

        if (open) {
          this.props.onPickerToggle(true);
        } else {
          this.handleClose();
        }
      });
    });

    _defineProperty(this, "handleClose", () => {
      this.props.onPickerToggle(false);
      this.setState({
        open: false
      });
      const el = this.ref.current;

      if (el) {
        el.focus();
      }
    });

    _defineProperty(this, "handleClick", () => {
      this.toggle();
    });

    _defineProperty(this, "handleKeyDown", event => {
      if (event.target !== this.ref.current) {
        return;
      }

      switch (event.key) {
        case ENTER:
        case SPACE:
          event.preventDefault();
          this.toggle();
          break;

        default:
          break;
      }
    });
  }

  render() {
    const _this$props = this.props,
          {
      cx,
      children,
      disabled,
      invalid,
      noResultsLabel,
      searchPlaceholder,
      styles
    } = _this$props,
          passThruProps = _objectWithoutPropertiesLoose(_this$props, ["cx", "children", "disabled", "invalid", "noResultsLabel", "searchPlaceholder", "styles"]);

    const {
      chosen
    } = passThruProps;
    const {
      open
    } = this.state;
    return React.createElement("div", null, React.createElement("div", {
      ref: this.ref,
      className: cx(styles.selectlike, !disabled && styles.selectlike_enabled, styles.input, styles.select, invalid && styles.input_invalid, disabled && styles.input_disabled),
      tabIndex: disabled ? -1 : 0,
      role: "button",
      onKeyDown: this.handleKeyDown,
      onClick: this.handleClick
    }, children || React.createElement(Text, null, this.boundFormatter(chosen || [])), React.createElement("span", {
      className: cx(styles.arrow)
    }, React.createElement(IconCaretDown, {
      decorative: true,
      size: "1.5em"
    }))), React.createElement(Overlay, {
      open: open,
      onClose: this.handleClose
    }, React.createElement(Picker, _extends({}, passThruProps, {
      formatter: this.boundFormatter,
      searchPlaceholder: searchPlaceholder || T.phrase('Search', {}, {
        context: 'Search for a topic within the hierarchy picker',
        key: 'lunar.common.search'
      }),
      noResultsLabel: noResultsLabel || T.phrase('No results', {}, {
        context: 'Label to display when no results are found for hierarchy picker search',
        key: 'lunar.picker.noResults'
      }),
      onClose: this.handleClose
    }))));
  }

}

_defineProperty(HierarchyPicker, "propTypes", {
  children: _pt.node,
  chosen: _pt.any,
  disabled: _pt.bool,
  formatter: _pt.func,
  fuseOptions: _pt.any,
  hierarchyMaxHeight: _pt.number,
  hierarchyWidth: _pt.number,
  indexParentPath: _pt.bool,
  invalid: _pt.bool,
  items: _pt.arrayOf(_pt.any).isRequired,
  noResultsLabel: _pt.string,
  onItemPicked: _pt.func.isRequired,
  onPickerToggle: _pt.func,
  renderItem: _pt.any,
  searchMaxHeight: _pt.number,
  searchPlaceholder: _pt.string,
  searchWidth: _pt.number,
  verticallyAlign: _pt.bool
});

_defineProperty(HierarchyPicker, "defaultProps", {
  chosen: [],
  disabled: false,
  formatter: defaultFormatter,
  hierarchyMaxHeight: 400,
  hierarchyWidth: 225,
  invalid: false,
  onPickerToggle: () => {},
  searchMaxHeight: 400,
  searchWidth: 300,
  verticallyAlign: false
});

export default withStyles(styleSheet)(HierarchyPicker);
