// @airbnb/lunar/esm/components/Grid/index.js
import _pt from "prop-types";
import React from 'react';
import { childrenOfType, mutuallyExclusiveTrueProps, forbidExtraProps } from 'airbnb-prop-types';
import useStyles from '../../hooks/useStyles';
import { styleSheet } from './styles';
import Col from './Col';

/** A grid to contain columns. */
function Grid(_ref) {
  let {
    bottomAlign,
    centerAlign,
    children,
    endAlign,
    middleAlign,
    reversed,
    startAlign,
    topAlign
  } = _ref;
  const [styles, cx] = useStyles(styleSheet);
  return React.createElement("section", {
    className: cx(styles.grid, reversed && styles.grid_reversed, bottomAlign && styles.grid_bottom, middleAlign && styles.grid_middle, topAlign && styles.grid_top, startAlign && styles.grid_start, endAlign && styles.grid_end, centerAlign && styles.grid_center)
  }, children);
}

export { Col };
const horizontalAlignProp = mutuallyExclusiveTrueProps('centerAlign', 'startAlign', 'endAlign');
const verticalAlignProp = mutuallyExclusiveTrueProps('bottomAlign', 'middleAlign', 'topAlign');
Grid.propTypes = {
  reversed: _pt.bool,
  bottomAlign: verticalAlignProp,
  centerAlign: horizontalAlignProp,
  children: childrenOfType(Col).isRequired,
  endAlign: horizontalAlignProp,
  middleAlign: verticalAlignProp,
  startAlign: horizontalAlignProp,
  topAlign: verticalAlignProp
};
export default Grid;
