// @airbnb/lunar/lib/components/ProgressBar/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** A bar to represent the progress to completion. */
function ProgressBar(_ref) {
  let {
    percent,
    leading,
    trailing
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheet);
  return _react.default.createElement("div", {
    className: cx(styles.wrapper)
  }, _react.default.createElement("div", {
    className: cx(styles.bar, !leading && styles.bar_leading, !trailing && styles.bar_trailing)
  }, _react.default.createElement("div", {
    className: cx(styles.bar, !leading && styles.bar_leading, !trailing && styles.bar_trailing, styles.progress),
    style: {
      width: percent + "%"
    }
  })));
}

ProgressBar.propTypes = {
  trailing: _propTypes.default.bool,
  leading: _propTypes.default.bool,
  percent: (0, _airbnbPropTypes.between)({
    gte: 0,
    lte: 100
  }).isRequired
};
var _default = ProgressBar;
exports.default = _default;
