// @airbnb/lunar/lib/components/Table/Cell.js
"use strict";

exports.__esModule = true;
exports.default = TableCell;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/** An individual table cell. */
function TableCell(_ref) {
  let {
    centerAlign,
    children,
    header,
    startAlign,
    endAlign,
    truncate,
    wrap
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, ["centerAlign", "children", "header", "startAlign", "endAlign", "truncate", "wrap"]);

  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheetCell);
  const Tag = header ? 'th' : 'td';
  return _react.default.createElement(Tag, _extends({}, props, {
    className: cx(truncate && styles.cell_truncate, startAlign && styles.cell_left, centerAlign && styles.cell_center, endAlign && styles.cell_right, wrap && styles.cell_wrap)
  }), children);
}

TableCell.propTypes = {
  centerAlign: _propTypes.default.bool,
  children: _propTypes.default.node,
  colSpan: _propTypes.default.number,
  endAlign: _propTypes.default.bool,
  header: _propTypes.default.bool,
  startAlign: _propTypes.default.bool,
  truncate: _propTypes.default.bool,
  wrap: _propTypes.default.bool
};
