// @airbnb/lunar/lib/components/Table/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _withBoundary = _interopRequireDefault(require("../../composers/withBoundary"));

var _Cell = _interopRequireDefault(require("./Cell"));

exports.Cell = _Cell.default;

var _Row = _interopRequireDefault(require("./Row"));

exports.Row = _Row.default;

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** A responsive table for displaying tabular data. */
function Table(_ref) {
  let {
    bordered,
    children,
    compact,
    fixed,
    horizontal,
    loading,
    middleAlign,
    noWrap,
    striped,
    transparent,
    vertical
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheet);
  return _react.default.createElement("div", {
    className: cx(!noWrap && styles.responsive_wrapper)
  }, _react.default.createElement("table", {
    className: cx(styles.table, middleAlign && styles.content_middle_align, fixed && styles.table_fixed, bordered && styles.table_bordered, horizontal && styles.table_horizontal, vertical && styles.table_vertical, compact && styles.table_compact, striped && styles.table_striped, loading && styles.table_loading, transparent && styles.table_transparent)
  }, children));
}

Table.propTypes = {
  bordered: _propTypes.default.bool,
  children: _propTypes.default.any.isRequired,
  compact: _propTypes.default.bool,
  fixed: _propTypes.default.bool,
  horizontal: _propTypes.default.bool,
  loading: _propTypes.default.bool,
  middleAlign: _propTypes.default.bool,
  noWrap: _propTypes.default.bool,
  striped: _propTypes.default.bool,
  trackingName: _propTypes.default.string,
  transparent: _propTypes.default.bool,
  vertical: _propTypes.default.bool
};

var _default = (0, _withBoundary.default)('Table')(Table);

exports.default = _default;
