// @airbnb/lunar/lib/components/Table/Row.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _constants = require("../../constants");

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/** A table row that contains multiple cells. */
function TableRow(_ref) {
  let {
    children,
    danger,
    info,
    muted,
    notice,
    success,
    warning
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, ["children", "danger", "info", "muted", "notice", "success", "warning"]);

  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheetRow);
  return _react.default.createElement("tr", _extends({}, props, {
    className: cx(danger && styles.row_danger, info && styles.row_info, muted && styles.row_muted, notice && styles.row_notice, success && styles.row_success, warning && styles.row_warning)
  }), children);
}

const statusPropType = (0, _airbnbPropTypes.mutuallyExclusiveTrueProps)(..._constants.STATUSES);
TableRow.propTypes = {
  children: _propTypes.default.any.isRequired,
  danger: statusPropType,
  info: statusPropType,
  muted: statusPropType,
  notice: statusPropType,
  success: statusPropType,
  warning: statusPropType
};
var _default = TableRow;
exports.default = _default;
