// @airbnb/lunar/lib/components/Menu/index.js
"use strict";

exports.__esModule = true;
exports.default = Menu;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _Separator = _interopRequireDefault(require("./Separator"));

exports.Separator = _Separator.default;

var _Item = _interopRequireDefault(require("./Item"));

exports.Item = _Item.default;

var _Row = _interopRequireDefault(require("./Row"));

exports.Row = _Row.default;

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** An abstract menu for use within dropdowns, selects, autocompletes, and more. */
function Menu(_ref) {
  let {
    accessibilityLabel,
    children,
    maxHeight,
    minWidth = 200,
    overflow,
    role = 'menu'
  } = _ref;
  const scrollable = !!maxHeight && !overflow;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheet);
  return _react.default.createElement("ul", {
    role: role,
    "aria-label": accessibilityLabel,
    className: cx(styles.menu, {
      maxHeight: scrollable ? maxHeight : 'auto',
      minWidth
    }, scrollable && styles.menu_scrollable)
  }, children);
}

Menu.propTypes = {
  children: _propTypes.default.node,
  accessibilityLabel: _propTypes.default.string.isRequired,
  maxHeight: _propTypes.default.number,
  minWidth: _propTypes.default.number,
  role: _propTypes.default.string,
  overflow: _propTypes.default.bool
};
