// @airbnb/lunar/lib/components/FormErrorMessage/index.js
"use strict";

exports.__esModule = true;
exports.default = FormErrorMessage;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _StatusText = _interopRequireDefault(require("../StatusText"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** Display a simple error message for use within forms. */
function FormErrorMessage(_ref) {
  let {
    id,
    error
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheet);

  if (!error) {
    return null;
  }

  return _react.default.createElement("div", {
    id: id + "-error",
    role: "alert",
    "aria-live": "polite",
    className: cx(styles.error)
  }, _react.default.createElement(_StatusText.default, {
    danger: true
  }, error));
}

FormErrorMessage.propTypes = {
  id: _propTypes.default.string.isRequired,
  error: _propTypes.default.string
};
