// @airbnb/lunar/lib/components/Accordion/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.Accordion = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _v = _interopRequireDefault(require("uuid/v4"));

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _Item = _interopRequireDefault(require("./Item"));

exports.Item = _Item.default;

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** A controller for multiple accordion items. */
class Accordion extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      id: (0, _v.default)(),
      index: this.props.defaultIndex || 0
    });

    _defineProperty(this, "handleClick", index => {
      this.setState(prevState => ({
        index: index === prevState.index ? -1 : index
      }));
    });
  }

  componentDidUpdate(prevProps) {
    if (this.props.defaultIndex !== prevProps.defaultIndex) {
      this.setState({
        index: this.props.defaultIndex
      });
    }
  }

  render() {
    const {
      cx,
      bordered,
      children,
      styles
    } = this.props;
    const {
      id,
      index
    } = this.state;
    return _react.default.createElement("div", {
      className: cx(styles.container, bordered && styles.container_bordered),
      role: "tablist"
    }, _react.default.Children.map(children, (child, i) => {
      if (!child) {
        return null;
      }

      return _react.default.cloneElement(child, {
        bordered,
        expanded: i === index,
        id: id + "-" + i,
        index: i,
        onClick: this.handleClick
      });
    }));
  }

}

exports.Accordion = Accordion;

_defineProperty(Accordion, "propTypes", {
  bordered: _propTypes.default.bool,
  children: _propTypes.default.any.isRequired,
  defaultIndex: _propTypes.default.number
});

_defineProperty(Accordion, "defaultProps", {
  bordered: false,
  defaultIndex: 0
});

var _default = (0, _withStyles.default)(_styles.styleSheet)(Accordion);

exports.default = _default;
