// @airbnb/lunar/lib/components/Accordion/styles.js
"use strict";

exports.__esModule = true;
exports.styleSheetItem = exports.styleSheet = void 0;

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const styleSheet = (_ref) => {
  let {
    color
  } = _ref;
  return {
    container: {
      borderBottom: '1px solid transparent',
      borderTop: '1px solid transparent',
      ':empty': {
        display: 'none'
      }
    },
    container_bordered: {
      borderBottomColor: color.accent.border
    }
  };
};

exports.styleSheet = styleSheet;

const styleSheetItem = (_ref2) => {
  let {
    pattern,
    ui,
    unit
  } = _ref2;
  return {
    body: {
      display: 'none',
      padding: unit + "px " + unit * 2 + "px " + unit * 2 + "px"
    },
    body_noSpacing: {
      padding: "0 0 " + unit * 2 + "px"
    },
    body_expanded: {
      display: 'block'
    },
    item_bordered: {
      borderTop: ui.border
    },
    title: _extends({}, pattern.resetButton, {
      display: 'flex',
      alignItems: 'center',
      padding: unit * 2,
      textAlign: 'left',
      width: '100%'
    }),
    title_noSpacing: {
      paddingLeft: 0,
      paddingRight: 0
    },
    titleText: {
      flex: '1'
    }
  };
};

exports.styleSheetItem = styleSheetItem;
