// @airbnb/lunar/lib/components/Multicomplete/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _shallowequal = _interopRequireDefault(require("shallowequal"));

var _Autocomplete = _interopRequireDefault(require("../Autocomplete"));

var _Spacing = _interopRequireDefault(require("../Spacing"));

var _Chip = _interopRequireDefault(require("./private/Chip"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** An uncontrolled input field for selecting multiple values via an autocomplete. */
class Multicomplete extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      values: new Set(this.props.value)
    });

    _defineProperty(this, "isItemSelected", (item, value) => this.state.values.has(value));

    _defineProperty(this, "handleChange", (value, event) => {
      this.props.onChange(Array.from(this.state.values), event);
    });

    _defineProperty(this, "handleSelectItem", (value, item, event) => {
      if (this.props.onSelectItem) {
        this.props.onSelectItem(value, item, event);
      }

      if (!value || !item) {
        return;
      }

      this.setState(prevState => ({
        values: new Set(prevState.values).add(value)
      }), () => {
        this.props.onChange(Array.from(this.state.values), event);
      });
    });

    _defineProperty(this, "handleChipClick", (value, event) => {
      this.setState(prevState => {
        const values = new Set(prevState.values);
        values.delete(value);
        return {
          values
        };
      }, () => {
        this.props.onChange(Array.from(this.state.values), event);
      });
    });
  }

  componentDidUpdate(prevProps) {
    if (!(0, _shallowequal.default)(this.props.value, prevProps.value)) {
      this.setState({
        values: new Set(this.props.value)
      });
    }
  }

  render() {
    const _this$props = this.props,
          props = _objectWithoutPropertiesLoose(_this$props, ["onChange"]);

    const selected = Array.from(this.state.values);
    return _react.default.createElement(_Autocomplete.default, _extends({
      clearOnSelect: true
    }, props, {
      isItemSelected: this.isItemSelected,
      value: "",
      onChange: this.handleChange,
      onSelectItem: this.handleSelectItem
    }), selected.length > 0 && _react.default.createElement("div", null, selected.map(item => _react.default.createElement(_Spacing.default, {
      key: item,
      inline: true,
      right: 1,
      top: 1
    }, _react.default.createElement(_Chip.default, {
      value: item,
      onClick: this.handleChipClick
    })))));
  }

}

exports.default = Multicomplete;

_defineProperty(Multicomplete, "propTypes", {
  onChange: _propTypes.default.func.isRequired,
  value: _propTypes.default.arrayOf(_propTypes.default.string)
});

_defineProperty(Multicomplete, "defaultProps", {
  value: []
});
