// @airbnb/lunar/lib/components/Modal/private/ImageLayout.js
"use strict";

exports.__esModule = true;
exports.default = ModalImageLayout;
exports.MAX_HEIGHT_IMAGE_SMALL = exports.MAX_HEIGHT_IMAGE = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _useStyles = _interopRequireDefault(require("../../../hooks/useStyles"));

var _styles = require("../styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const MAX_HEIGHT_IMAGE = 720;
exports.MAX_HEIGHT_IMAGE = MAX_HEIGHT_IMAGE;
const MAX_HEIGHT_IMAGE_SMALL = 420;
exports.MAX_HEIGHT_IMAGE_SMALL = MAX_HEIGHT_IMAGE_SMALL;

function ModalImageLayout(_ref) {
  let {
    children,
    sizes,
    srcSet,
    type,
    url
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheetImageLayout);
  return _react.default.createElement("div", {
    className: cx(styles.splitContent)
  }, _react.default.createElement("div", {
    className: cx(styles.splitContentPane)
  }, children), _react.default.createElement("div", {
    className: cx(styles.splitContentPane, styles.splitContentImagePane)
  }, type === 'center' && _react.default.createElement("img", {
    className: cx(styles.image),
    src: url,
    srcSet: srcSet && srcSet.join(','),
    sizes: sizes && sizes.join(','),
    alt: ""
  }), type === 'cover' && _react.default.createElement("img", {
    className: cx(styles.image, styles.imageCover),
    src: url,
    srcSet: srcSet && srcSet.join(','),
    alt: ""
  })));
}

ModalImageLayout.propTypes = {
  sizes: _propTypes.default.arrayOf(_propTypes.default.string),
  srcSet: _propTypes.default.arrayOf(_propTypes.default.string),
  type: _propTypes.default.oneOf(['center', 'cover']).isRequired,
  url: _propTypes.default.string.isRequired,
  children: _propTypes.default.any.isRequired
};
