// @airbnb/lunar/lib/components/Button/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.Button = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _ButtonOrLink = _interopRequireDefault(require("../private/ButtonOrLink"));

var _Loader = _interopRequireDefault(require("../Loader"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const sizingProp = (0, _airbnbPropTypes.mutuallyExclusiveTrueProps)('small', 'large');

/** A standard button and or link for common UI interactions. */
class Button extends _react.default.Component {
  render() {
    const _this$props = this.props,
          {
      cx,
      block,
      borderless,
      children,
      disabled,
      invalid,
      inverted,
      large,
      loading,
      small,
      styles
    } = _this$props,
          restProps = _objectWithoutPropertiesLoose(_this$props, ["cx", "block", "borderless", "children", "disabled", "invalid", "inverted", "large", "loading", "small", "styles"]);

    return _react.default.createElement(_ButtonOrLink.default, _extends({}, restProps, {
      "aria-busy": loading,
      disabled: disabled,
      loading: loading,
      className: cx(styles.button, large && styles.button_large, small && styles.button_small, !large && !small && styles.button_regular, block && styles.button_block, (disabled || loading) && styles.button_disabled, (borderless || inverted) && styles.button_inverted, invalid && styles.button_invalid, borderless && styles.button_borderless, loading && styles.button_loading)
    }), loading ? _react.default.createElement(_Loader.default, {
      inline: true,
      inverted: !inverted
    }) : children);
  }

}

exports.Button = Button;

_defineProperty(Button, "propTypes", {
  inverted: _propTypes.default.bool,
  invalid: _propTypes.default.bool,
  borderless: _propTypes.default.bool,
  block: _propTypes.default.bool,
  large: sizingProp,
  small: sizingProp
});

_defineProperty(Button, "defaultProps", {
  block: false,
  borderless: false,
  disabled: false,
  invalid: false,
  inverted: false,
  large: false,
  loading: false,
  small: false
});

var _default = (0, _withStyles.default)(_styles.styleSheet, {
  extendable: true
})(Button);

exports.default = _default;
