// @airbnb/lunar/esm/components/Tooltip/styles.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const styleSheet = (_ref) => {
  let {
    unit,
    color,
    pattern,
    ui
  } = _ref;
  return {
    container: {
      display: 'inline-block'
    },
    offscreen: _extends({}, pattern.offscreen),
    underlined: {
      borderBottom: "1px dotted " + color.core.primary[3],
      cursor: 'help'
    },
    tooltip: {
      animationDuration: '200ms',
      animationTimingFunction: 'ease-out'
    },
    tooltip_above: {
      animationName: {
        name: 'fadeDown',
        from: {
          opacity: 0,
          transform: "translateY(" + unit * 1.5 + "px)"
        },
        to: {
          opacity: 1
        }
      }
    },
    tooltip_below: {
      animationName: {
        name: 'fadeUp',
        from: {
          opacity: 0,
          transform: "translateY(-" + unit * 1.5 + "px)"
        },
        to: {
          opacity: 1
        }
      }
    },
    shadow: {
      display: 'inline-block',
      boxShadow: ui.boxShadowLarge,
      borderRadius: ui.borderRadius
    }
  };
};

export default styleSheet;
export { styleSheet };
