// @airbnb/lunar/lib/components/private/BaseAffix.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _airbnbPropTypes = require("airbnb-prop-types");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const dirProp = (0, _airbnbPropTypes.mutuallyExclusiveTrueProps)('after', 'before');

class BaseAffix extends _react.default.PureComponent {
  render() {
    const {
      cx,
      after,
      before,
      children,
      compact,
      disabled,
      flex,
      styles
    } = this.props;
    return _react.default.createElement("div", {
      className: cx(styles.affix, compact && styles.affix_compact, before && styles.affix_before, after && styles.affix_after, disabled && styles.affix_disabled, flex && styles.affix_flex)
    }, children);
  }

}

exports.default = BaseAffix;

_defineProperty(BaseAffix, "propTypes", {
  flex: _propTypes.default.bool,
  disabled: _propTypes.default.bool,
  compact: _propTypes.default.bool,
  children: _propTypes.default.any.isRequired,
  after: dirProp,
  before: dirProp
});

_defineProperty(BaseAffix, "defaultProps", {
  after: false,
  before: false,
  compact: false,
  disabled: false,
  flex: false
});
