// @airbnb/lunar/lib/components/private/FieldAffix.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _withStyles = _interopRequireDefault(require("../../composers/withStyles"));

var _BaseAffix = _interopRequireDefault(require("./BaseAffix"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var _default = (0, _withStyles.default)((_ref) => {
  let {
    color,
    font,
    unit,
    ui,
    pattern
  } = _ref;
  return {
    affix: _extends({}, font.textRegular, {
      height: '100%',
      display: 'flex',
      alignItems: 'center',
      border: ui.borderThick,
      borderRadius: ui.borderRadius,
      background: color.accent.bgHover,
      paddingLeft: unit * 1.5,
      paddingRight: unit * 1.5
    }),
    affix_compact: _extends({}, font.textSmall, {
      paddingLeft: unit,
      paddingRight: unit
    }),
    affix_after: {
      borderLeft: 0,
      borderTopLeftRadius: 0,
      borderBottomLeftRadius: 0
    },
    affix_before: {
      borderRight: 0,
      borderTopRightRadius: 0,
      borderBottomRightRadius: 0
    },
    affix_disabled: _extends({}, pattern.disabled)
  };
})(_BaseAffix.default);

exports.default = _default;
