// @data-ui/sparkline/build/annotation/Label.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultProps = exports.propTypes = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _color = require('@data-ui/theme/build/color');

var _color2 = _interopRequireDefault(_color);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var propTypes = exports.propTypes = {
  anchor: _propTypes2.default.oneOf(['start', 'middle', 'end']),
  dominantBaseline: _propTypes2.default.string,
  dx: _propTypes2.default.number,
  dy: _propTypes2.default.number,
  fill: _propTypes2.default.string,
  label: _propTypes2.default.node,
  paintOrder: _propTypes2.default.string,
  stroke: _propTypes2.default.string,
  strokeWidth: _propTypes2.default.number,
  x: _propTypes2.default.number,
  y: _propTypes2.default.number
};

var defaultProps = exports.defaultProps = {
  anchor: 'middle',
  dominantBaseline: 'middle',
  dx: 0,
  dy: 0,
  fill: _color2.default.text,
  label: null,
  paintOrder: 'stroke',
  stroke: '#fff',
  strokeWidth: 2,
  x: 0,
  y: 0
};

function Label(_ref) {
  var anchor = _ref.anchor,
      dominantBaseline = _ref.dominantBaseline,
      dx = _ref.dx,
      dy = _ref.dy,
      fill = _ref.fill,
      label = _ref.label,
      paintOrder = _ref.paintOrder,
      stroke = _ref.stroke,
      strokeWidth = _ref.strokeWidth,
      x = _ref.x,
      y = _ref.y,
      rest = _objectWithoutProperties(_ref, ['anchor', 'dominantBaseline', 'dx', 'dy', 'fill', 'label', 'paintOrder', 'stroke', 'strokeWidth', 'x', 'y']);

  return _react2.default.createElement(
    'text',
    _extends({
      x: x,
      y: y,
      dx: dx,
      dy: dy,
      fill: fill,
      dominantBaseline: dominantBaseline,
      paintOrder: paintOrder,
      stroke: stroke,
      strokeWidth: strokeWidth,
      textAnchor: anchor
    }, rest),
    label
  );
}

Label.propTypes = propTypes;
Label.defaultProps = defaultProps;
Label.displayName = 'Label';

exports.default = Label;
