// @airbnb/lunar/esm/components/DatePickerInput/Input/index.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import { DayModifiers } from 'react-day-picker';
import DayPickerInput from 'react-day-picker/DayPickerInput';
import FormInput from '../../private/FormInput';
import Dropdown from '../../Dropdown';
import DatePicker from '../../DatePicker'; // The original `DayPickerInput` has very restrictive and problematic
// markup being rendered, that doesn't interop with our form layer very well.
// Since we can't hook into it using render props, we need to extend
// the component and override the render methods.

export default class PrivatePickerInput extends DayPickerInput {
  constructor() {
    super(...arguments);

    _defineProperty(this, "props", void 0);

    _defineProperty(this, "daypicker", null);

    _defineProperty(this, "input", null);

    _defineProperty(this, "loadInputRef", ref => {
      // @ts-ignore Ignore typings
      const {
        propagateRef
      } = this.props.inputProps;
      this.input = ref;

      if (propagateRef) {
        propagateRef.current = ref;
      }
    });

    _defineProperty(this, "loadPickerRef", ref => {
      this.daypicker = ref;
    });

    _defineProperty(this, "handleResetClick", () => {
      // @ts-ignore Ignore typings
      this.updateState(new Date(), '', this.hideAfterDayClick);
    });

    _defineProperty(this, "handleTodayButtonClick", () => {
      // @ts-ignore Ignore typings
      this.updateState(new Date(), this.props.formatDate(new Date()), this.hideAfterDayClick);
    });

    _defineProperty(this, "renderOverlay", () => {
      const {
        dayPickerProps = {},
        dropdownProps = {}
      } = this.props;
      const {
        selectedDays,
        value
      } = this.state;
      let selectedDay;

      if (!selectedDays && value) {
        // @ts-ignore Ignore typings
        const day = this.props.parseDate(value);

        if (day) {
          selectedDay = day;
        }
      } else if (selectedDays) {
        selectedDay = selectedDays;
      }

      return React.createElement(Dropdown, _extends({
        top: "100%",
        tabIndex: 0,
        zIndex: 100
      }, dropdownProps, {
        onFocus: this.handleOverlayFocus,
        onBlur: this.handleOverlayBlur
      }), React.createElement(DatePicker, _extends({}, dayPickerProps, {
        month: this.state.month,
        selectedDays: selectedDay,
        pickerRef: this.loadPickerRef,
        onDayClick: this.handleDayClick,
        onMonthChange: this.handleMonthChange,
        onResetClick: this.handleResetClick,
        onTodayButtonClick: this.handleTodayButtonClick
      })));
    });
  }

  render() {
    const {
      inputProps
    } = this.props;
    return React.createElement(React.Fragment, null, React.createElement(FormInput, _extends({
      type: "text",
      tagName: "input",
      placeholder: this.props.placeholder
    }, inputProps, {
      value: this.state.typedValue || this.state.value,
      propagateRef: this.loadInputRef,
      onChange: this.handleInputChange,
      onFocus: this.handleInputFocus,
      onBlur: this.handleInputBlur,
      onKeyDown: this.handleInputKeyDown,
      onKeyUp: this.handleInputKeyUp,
      onClick: this.handleInputClick
    })), this.state.showOverlay && this.renderOverlay());
  }

}
