// @airbnb/lunar/lib/components/List/index.js
"use strict";

exports.__esModule = true;
exports.default = List;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _Item = _interopRequireDefault(require("./Item"));

exports.Item = _Item.default;

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function List(_ref) {
  let {
    children,
    gutter,
    horizontal,
    ordered,
    wrap
  } = _ref;
  const Tag = ordered ? 'ol' : 'ul';
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheet);
  return _react.default.createElement(Tag, {
    className: cx(styles.list, !horizontal && gutter && styles.list_gutter, horizontal && styles.list_horizontal, horizontal && gutter && styles.list_gutter_horizontal, horizontal && wrap && styles.list_horizontal_wrap)
  }, _react.default.Children.map(children, child => {
    if (!child) {
      return null;
    }

    if (horizontal) {
      return _react.default.cloneElement(child, {
        horizontal
      });
    }

    return child;
  }));
}

List.propTypes = {
  children: _propTypes.default.any.isRequired,
  gutter: _propTypes.default.bool,
  horizontal: _propTypes.default.bool,
  ordered: _propTypes.default.bool,
  wrap: _propTypes.default.bool
};
