// @airbnb/lunar/lib/components/List/Item.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _airbnbPropTypes = require("airbnb-prop-types");

var _useStyles = _interopRequireDefault(require("../../hooks/useStyles"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ListItem(_ref) {
  let {
    bordered,
    children,
    compact,
    horizontal,
    spacious
  } = _ref;
  const [styles, cx] = (0, _useStyles.default)(_styles.styleSheetItem);
  return _react.default.createElement("li", {
    className: cx(!horizontal && bordered && styles.item_bordered, horizontal && bordered && styles.item_bordered_horizontal, !horizontal && compact && styles.item_compact, horizontal && compact && styles.item_compact_horizontal, !horizontal && spacious && styles.item_spacious, horizontal && spacious && styles.item_spacious_horizontal)
  }, children);
}

const paddingPropType = (0, _airbnbPropTypes.mutuallyExclusiveTrueProps)('compact', 'spacious');
ListItem.propTypes = {
  horizontal: _propTypes.default.bool,
  children: _propTypes.default.any.isRequired,
  bordered: _propTypes.default.bool,
  compact: paddingPropType,
  spacious: paddingPropType
};
var _default = ListItem;
exports.default = _default;
