// @airbnb/lunar/lib/components/DateTimeRange/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _messages = require("../../messages");

var _DateTime = _interopRequireDefault(require("../DateTime"));

var _Empty = _interopRequireDefault(require("../Empty"));

var _createDateTime = _interopRequireDefault(require("../../utils/createDateTime"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Display a range between 2 timestamps. */
class DateTimeRange extends _react.default.PureComponent {
  render() {
    const {
      from,
      locale,
      separator,
      timezone,
      to
    } = this.props;

    if (!from || !to) {
      return _react.default.createElement(_Empty.default, null);
    }

    const fromTimeStamp = (0, _createDateTime.default)(from, {
      locale,
      timezone
    });
    const toTimeStamp = (0, _createDateTime.default)(to, {
      locale,
      timezone
    });

    if ("production" !== process.env.NODE_ENV) {
      if (!fromTimeStamp.isValid || !toTimeStamp.isValid) {
        throw new Error('Invalid timestamps passed to `DateTimeRange`.');
      }

      if (toTimeStamp < fromTimeStamp) {
        throw new Error('Invalid chronological order of timestamps passed to `DateTimeRange`.');
      }
    }

    const props = {
      locale,
      timezone
    };

    let fromFormat = _messages.rangeFromDayBundle.get(locale);

    let toFormat;

    if (fromTimeStamp.year !== toTimeStamp.year) {
      fromFormat = _messages.dateMediumBundle.get(locale);
      toFormat = _messages.dateMediumBundle.get(locale);
    } else if (fromTimeStamp.month !== toTimeStamp.month) {
      toFormat = _messages.dateMediumBundle.get(locale);
    } else if (fromTimeStamp.day !== toTimeStamp.day) {
      toFormat = _messages.rangeToDayBundle.get(locale);
    } else {
      return _react.default.createElement(_DateTime.default, _extends({}, props, {
        medium: true,
        noTime: true,
        noTimezone: true,
        at: toTimeStamp
      }));
    }

    return _react.default.createElement("span", null, _react.default.createElement(_DateTime.default, _extends({}, props, {
      at: fromTimeStamp,
      format: fromFormat
    })), separator, _react.default.createElement(_DateTime.default, _extends({}, props, {
      at: toTimeStamp,
      format: toFormat
    })));
  }

}

exports.default = DateTimeRange;

_defineProperty(DateTimeRange, "propTypes", {
  from: _propTypes.default.any,
  locale: _propTypes.default.any,
  separator: _propTypes.default.string,
  timezone: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.bool]),
  to: _propTypes.default.any
});

_defineProperty(DateTimeRange, "defaultProps", {
  from: null,
  separator: ' – ',
  to: null
});
